<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • ideal</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.19.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/federicomarini/ideal" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/federicomarini/ideal/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="1.18.0" id="ideal-1180">ideal 1.18.0<a class="anchor" aria-label="anchor" href="#ideal-1180"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-18-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-18-0"></a></h3>
<ul><li>Fixed the unexpected behaviour when decorating the signature heatmap with colData elements</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-18-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-18-0"></a></h3>
<ul><li>Some alignments of the UI elements in the <code><a href="../reference/ideal.html">ideal()</a></code> app, to harmonize the content of the page</li>
<li>Updated the icons in the user interface to match recent changes in the names from FontAwesome</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.16.0" id="ideal-1160">ideal 1.16.0<a class="anchor" aria-label="anchor" href="#ideal-1160"></a></h2>
<div class="section level3">
<h3 id="new-features-1-16-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-16-0"></a></h3>
<ul><li>It is now possible to export the processed data (count data, results, functional enrichment table) into a combined list object, which can be seamlessly fed into <code>GeneTonic</code> (<a href="http://bioconductor.org/packages/release/bioc/html/GeneTonic.html" class="external-link uri">http://bioconductor.org/packages/release/bioc/html/GeneTonic.html</a>).</li>
<li>Adjusted the behavior for the modeling with LRT, with extra notifications to inform the user on how to make the most out of the functionality.</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-16-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-16-0"></a></h3>
<ul><li>The manuscript of <code>ideal</code> is now published in BMC Bioinformatics! The citation file has been updated accordingly.</li>
<li>A full round of <code>styler</code> has been applied to the codebase.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.14.0" id="ideal-1140">ideal 1.14.0<a class="anchor" aria-label="anchor" href="#ideal-1140"></a></h2>
<div class="section level3">
<h3 id="other-notes-1-14-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-14-0"></a></h3>
<ul><li>Replaced dependency from <code>d3heatmap</code> with the functionality of <code>heatmaply</code>
</li>
<li>Additional notifications are returned when closing the app while saving the state to the R environment - both in the console as a <code>message</code>, and in the UI via <code>showNotification</code>
</li>
<li>The vignette and the tour contain a section about the format required to run <code>ideal</code> correctly</li>
<li>The vignette also indicates some exemplary steps after exporting the values to the R environment, showcasing how to perform additional functional enrichment analyses e.g. via <code>fgsea</code> or <code>gprofiler2</code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.12.0" id="ideal-1120">ideal 1.12.0<a class="anchor" aria-label="anchor" href="#ideal-1120"></a></h2>
<div class="section level3">
<h3 id="new-features-1-12-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-12-0"></a></h3>
<ul><li>Implemented a series of gentle fail mechanisms to avoid abrupt crashing of the app if the expected objects or formats are not provided</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-12-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-12-0"></a></h3>
<ul><li>Alignments of UI elements have been adjusted</li>
<li>Updated citation info - the preprint for <code>ideal</code>, “ideal: an R/Bioconductor package for Interactive Differential Expression Analysis”, is now published at bioRxiv, <a href="https://doi.org/10.1101/2020.01.10.901652" class="external-link uri">https://doi.org/10.1101/2020.01.10.901652</a>
</li>
<li>Data transformation (for table viewing and visualizations, where homoscedastic data is expected) is now taken care of with the variance stabilizing transformation (<code>vst</code>), instead of the <code>rlog</code>
</li>
<li>The installation information in the <code>README.md</code> includes a troubleshooting section to solve some common issues one might encounter</li>
<li>Added a collapsible element in the Data Setup panel, to remind users to perform proper Exploratory Data Analysis (e.g. with <code>pcaExplorer</code>) before testing for Differential Expression</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.0" id="ideal-1100">ideal 1.10.0<a class="anchor" aria-label="anchor" href="#ideal-1100"></a></h2>
<div class="section level3">
<h3 id="new-features-1-10-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-10-0"></a></h3>
<ul><li>
<code>ggplotCounts</code> gains a new parameter, <code>labels_repel</code>, to control the placement of the different labels in the plot - this can be useful when a large number of samples is available</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-10-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-10-0"></a></h3>
<ul><li>Fixed an error in the initialization of the app due to a new behavior introduced by <code>shinyAce</code> in version &gt;= 0.4.0 - occurred in the same way as for <code>pcaExplorer</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-10-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-10-0"></a></h3>
<ul><li>Better class checks via <code>is(...)</code> as per <code>BiocCheck</code> suggestion</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.0" id="ideal-180">ideal 1.8.0<a class="anchor" aria-label="anchor" href="#ideal-180"></a></h2>
<div class="section level3">
<h3 id="new-features-1-8-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-8-0"></a></h3>
<ul><li>
<code>plot_ma</code> gains an additional parameter, <code>labels_repel</code>, for better placing the labels on the features to mark</li>
<li>The pairwise correlation plots can now use logarithmic scale in the axes, use smaller subsets of the data for quicker inspection, and resizes the correlation info proportionally to its intensity</li>
<li>The id types can now be chosen among the <code>keytypes</code> of the corresponding annotation packages (which still need to be installed when <code>ideal</code> is launched). Other input fields that specify id types also behave in a similar manner (e.g. in the Signature Explorer tab). This caused a problem for scenarios where common id types such as <code>ENSEMBL</code> are not available, like in Arabidopsis (where ids are provided often as <code>TAIR</code>) - thanks to Marc Galland for picking this up in <a href="https://github.com/federicomarini/ideal/issues/1" class="external-link uri">https://github.com/federicomarini/ideal/issues/1</a>
</li>
<li>The Signatures Explorer tab also has a fully fledged tour for first-time users, together with a collapsible help panel to describe its functionality in brief</li>
<li>The zoomed MA-plot gains a new widget to control the labels size for the names of the genes</li>
<li>It is possible to export the input data together with the results in a combined <code>SummarizedExperiment</code> object, which can be seamlessly fed into <code>iSEE</code> (<a href="http://bioconductor.org/packages/release/bioc/html/iSEE.html" class="external-link uri">http://bioconductor.org/packages/release/bioc/html/iSEE.html</a>). This leverages a new function, <code>wrapup_for_iSEE</code>, which is available and exported from <code>ideal</code>.</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-8-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-8-0"></a></h3>
<ul><li>An information box has been added to provide detailed information on the required input formats</li>
<li>Added notification to specify how to install the airway package for demonstration purposes if not already available</li>
<li>The Signatures Explorer tab is now displayed in a conditional panel, i.e. only when the required inputs are provided</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.0" id="ideal-160">ideal 1.6.0<a class="anchor" aria-label="anchor" href="#ideal-160"></a></h2>
<div class="section level3">
<h3 id="new-features-1-6-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-6-0"></a></h3>
<ul><li>Added extra diagnostic to results exploration (stratified pvalues histograms, schweder-spjotvoll plot)</li>
<li>Added functionality for the tab Signatures Explorer: read in gmt files, and explore signatures for all/DE genes visually as heatmaps. The signatures can also be provided in the main call of the function as an argument, on top of uploading at runtime</li>
<li>Different handling of the selection of ids/gene names, now not requiring anymore that the result is computed, but based on the dds object (and on the presence or not of the annotation object)</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-6-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-6-0"></a></h3>
<ul><li>Built project website via pkgdown, with customized reference structure</li>
<li>Added a <code>NEWS.md</code> file to track changes to the package</li>
<li>Updated the report template as well for including new functionality, and also updated vignette entry</li>
<li>Correctly adding the resources to shinyBS, loaded via <code>.onLoad</code>
</li>
<li>Replaced the first tour structure and call</li>
<li>New skin for the app, also with logo in the title header</li>
<li>Instructions start collapsed for more compact main page</li>
<li>Edited link in button to provide feedback, with subject specified</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-6-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-6-0"></a></h3>
<ul><li>Fixed behavior of rendering inline the content of the report - did not work properly for server deployed instances</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.0" id="ideal-140">ideal 1.4.0<a class="anchor" aria-label="anchor" href="#ideal-140"></a></h2>
<div class="section level3">
<h3 id="new-features-1-4-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-4-0"></a></h3>
<ul><li>Specified single go term selection for generating heatmaps of gene signatures</li>
<li>Added support for logFC shrinkage, following the latest devels of DESeq2</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-4-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-4-0"></a></h3>
<ul><li>Corrected output for the vignette, as html_document2 is now deprecated</li>
<li>Menus are back in the expanded form</li>
<li>Fixed the behavior with addMLE</li>
</ul></div>
<div class="section level3">
<h3 id="other-notes-1-4-0">Other notes<a class="anchor" aria-label="anchor" href="#other-notes-1-4-0"></a></h3>
<ul><li>Added further progress indicators to give feedback during lengthy steps</li>
<li>Improved ggplotCounts for better scale display, using exact arg matching, defaulting to the transformed counts</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.0" id="ideal-0990">ideal 0.99.0<a class="anchor" aria-label="anchor" href="#ideal-0990"></a></h2>
<div class="section level3">
<h3 id="new-features-0-99-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-99-0"></a></h3>
<ul><li>Ready for Bioc submission</li>
<li>Completed the news</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.1" id="ideal-091">ideal 0.9.1<a class="anchor" aria-label="anchor" href="#ideal-091"></a></h2>
<div class="section level3">
<h3 id="new-features-0-9-1">New features<a class="anchor" aria-label="anchor" href="#new-features-0-9-1"></a></h3>
<ul><li>Added Instructions fully from rendered version of the vignette to have available at runtime</li>
<li>Added support for downloading all plots and tables</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.0" id="ideal-090">ideal 0.9.0<a class="anchor" aria-label="anchor" href="#ideal-090"></a></h2>
<div class="section level3">
<h3 id="new-features-0-9-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-9-0"></a></h3>
<ul><li>Interactive tours are covering now all tabs, with extensive walkthroughs for the user</li>
<li>Added all screenshots to vignette</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.6.2" id="ideal-062">ideal 0.6.2<a class="anchor" aria-label="anchor" href="#ideal-062"></a></h2>
<div class="section level3">
<h3 id="new-features-0-6-2">New features<a class="anchor" aria-label="anchor" href="#new-features-0-6-2"></a></h3>
<ul><li>Interactive tours are now available, coded in external files</li>
<li>Travis-CI is now supported</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.6.0" id="ideal-060">ideal 0.6.0<a class="anchor" aria-label="anchor" href="#ideal-060"></a></h2>
<div class="section level3">
<h3 id="new-features-0-6-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-6-0"></a></h3>
<ul><li>Added MA plot with extra custom list to avoid manual selection of many genes</li>
<li>MA plot function now automatically supports subset of gene to be extra plotted</li>
<li>Added documentation with roxygen to all functions</li>
<li>Heatmap functions for genes annotated to a GO term as signature</li>
<li>Template report also provided</li>
<li>Full draft of vignette now available, working towards bioc submission</li>
<li>Added textual help to all sections, with collapsible element</li>
<li>Added proof of principle to have interactive tours based on rintrojs</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.4.0" id="ideal-040">ideal 0.4.0<a class="anchor" aria-label="anchor" href="#ideal-040"></a></h2>
<div class="section level3">
<h3 id="new-features-0-4-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-4-0"></a></h3>
<ul><li>Gene box info added, based on rentrez</li>
<li>New look for MA plots and volcano plots</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.0" id="ideal-030">ideal 0.3.0<a class="anchor" aria-label="anchor" href="#ideal-030"></a></h2>
<div class="section level3">
<h3 id="new-features-0-3-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-3-0"></a></h3>
<ul><li>Restructuring of the folders done, package can be correctly installed, loaded - namespace, description are set up</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.0" id="ideal-020">ideal 0.2.0<a class="anchor" aria-label="anchor" href="#ideal-020"></a></h2>
<div class="section level3">
<h3 id="new-features-0-2-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-2-0"></a></h3>
<ul><li>Correct structure of the package</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.0" id="ideal-010">ideal 0.1.0<a class="anchor" aria-label="anchor" href="#ideal-010"></a></h2>
<div class="section level3">
<h3 id="new-features-0-1-0">New features<a class="anchor" aria-label="anchor" href="#new-features-0-1-0"></a></h3>
<ul><li>Package created!</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://federicomarini.github.io" class="external-link">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

