<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate a tidy table with the DE genes from the results of DESeq — deseqresult2DEgenes • ideal</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a tidy table with the DE genes from the results of DESeq — deseqresult2DEgenes"><meta property="og:description" content="Generate a tidy table with the DE genes from the results of DESeq"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.19.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/federicomarini/ideal" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate a tidy table with the DE genes from the results of DESeq</h1>
    <small class="dont-index">Source: <a href="https://github.com/federicomarini/ideal/blob/HEAD/R/res2tbl.R" class="external-link"><code>R/res2tbl.R</code></a></small>
    <div class="hidden name"><code>deseqresult2DEgenes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a tidy table with the DE genes from the results of DESeq</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">deseqresult2DEgenes</span><span class="op">(</span><span class="va">deseqresult</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>deseqresult</dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html" class="external-link">DESeqResults</a></code> object</p></dd>
<dt>FDR</dt>
<dd><p>Numeric value, the significance level for thresholding adjusted p-values</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A "tidy" data.frame with only genes marked as differentially expressed</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="co"># with simulated data...</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-in"><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/makeExampleDESeqDataSet.html" class="external-link">makeExampleDESeqDataSet</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, m <span class="op">=</span> <span class="fl">8</span>, betaSD <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span class="r-in"><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-in"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">deseqresult2DEgenes</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            id   baseMean log2FoldChange     lfcSE      stat       pvalue</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene72 gene72 450.195711      2.8726496 0.3039964  9.449618 3.400704e-21</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene73 gene73 390.764308      4.1571243 0.4369197  9.514619 1.823804e-21</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene84 gene84 186.010722      2.9690842 0.3856856  7.698199 1.379973e-14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene12 gene12  63.618725     -3.5076988 0.4618335 -7.595159 3.074136e-14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene28 gene28  48.449407     -3.2207571 0.4954035 -6.501280 7.963925e-11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene50 gene50  41.038476     -3.4085868 0.5469753 -6.231702 4.613954e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene90 gene90  93.952569      2.9129864 0.4682538  6.220957 4.941323e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene26 gene26  64.055099     -2.9967959 0.4841817 -6.189403 6.039240e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene65 gene65  70.410741      3.3832153 0.5473923  6.180605 6.385654e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene29 gene29  23.479540     -4.6968674 0.7909313 -5.938401 2.878154e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene80 gene80 403.167831      2.0581544 0.3456874  5.953802 2.619841e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene11 gene11  19.224838     -5.2532456 0.8924150 -5.886550 3.943401e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6   gene6  14.557136     -5.1869180 0.9830491 -5.276357 1.317774e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene18 gene18  30.775077     -3.1266181 0.6038368 -5.177920 2.243740e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene58 gene58 115.563100      2.0757737 0.4014103  5.171202 2.325926e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene86 gene86  41.678654      3.3919218 0.6817206  4.975531 6.506910e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene24 gene24   6.757583     -6.5501442 1.3833432 -4.735010 2.190442e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene33 gene33  10.498930     -4.6597222 1.0345739 -4.504001 6.668574e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene92 gene92 136.849120      4.0578331 0.9218432  4.401869 1.073223e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene55 gene55   9.595056     -4.5438177 1.0722111 -4.237801 2.257194e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene51 gene51  65.312786      2.0961135 0.5226604  4.010470 6.059806e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene30 gene30  38.466708      2.0882741 0.5261362  3.969075 7.215218e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene83 gene83  32.130914      2.2438232 0.5714145  3.926787 8.608811e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene99 gene99  13.215602     -2.8805990 0.7382424 -3.901969 9.541335e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7   gene7   9.037552     -3.8558639 0.9946321 -3.876673 1.058944e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene63 gene63  10.646118     -4.3948026 1.1332874 -3.877924 1.053515e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene40 gene40  30.274109     -2.3262181 0.6185099 -3.761004 1.692331e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene42 gene42  36.094886      2.1812231 0.5995681  3.637990 2.747736e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene47 gene47  87.481917      1.7059376 0.4873811  3.500213 4.648868e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene22 gene22  65.004576     -1.5658595 0.4520391 -3.463992 5.322230e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene68 gene68  39.847172     -1.5991166 0.4738890 -3.374454 7.396235e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene17 gene17  10.767377     -2.7640067 0.8589323 -3.217956 1.291074e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene54 gene54  34.836688     -1.7177991 0.5427615 -3.164924 1.551233e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene35 gene35  17.597058     -2.0160959 0.6501023 -3.101198 1.927394e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene37 gene37 965.206104      1.0004767 0.3243701  3.084368 2.039852e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene97 gene97  93.222142      1.1971031 0.3974222  3.012170 2.593876e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene91 gene91   5.390343     -4.1752900 1.5146143 -2.756669 5.839345e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene82 gene82  50.659143     -1.2197948 0.4510429 -2.704387 6.843046e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene69 gene69 124.315379     -0.9054537 0.3471010 -2.608617 9.090886e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene38 gene38  19.249041      1.8780708 0.7477486  2.511634 1.201736e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene36 gene36   7.383298     -2.2559963 0.9223209 -2.445999 1.444512e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene48 gene48   3.195330     -4.7137990 1.9415523 -2.427851 1.518860e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene25 gene25  21.876995     -1.6155091 0.6717073 -2.405079 1.616897e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                padj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene72 1.700352e-19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene73 1.700352e-19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene84 4.599911e-13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene12 7.685341e-13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene28 1.592785e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene50 7.059033e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene90 7.059033e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene26 7.095171e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene65 7.095171e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene29 2.616504e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene80 2.616504e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene11 3.286167e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6  1.013672e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene18 1.550618e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene58 1.550618e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene86 4.066818e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene24 1.288496e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene33 3.704763e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene92 5.648544e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene55 1.128597e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene51 2.885622e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene30 3.279644e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene83 3.742961e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene99 3.975556e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  4.072860e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene63 4.072860e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene40 6.267891e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene42 9.813344e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene47 1.603058e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene22 1.774077e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene68 2.385882e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene17 4.034607e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene54 4.700707e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene35 5.668805e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene37 5.828148e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene97 7.205212e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene91 1.578201e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene82 1.800802e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene69 2.330996e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene38 3.004340e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene36 3.523201e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene48 3.616333e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene25 3.760226e-02</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://federicomarini.github.io" class="external-link">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

