<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate a tidy table with the DE genes from the results of DESeq — deseqresult2DEgenes • ideal</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a tidy table with the DE genes from the results of DESeq — deseqresult2DEgenes"><meta property="og:description" content="Generate a tidy table with the DE genes from the results of DESeq"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.21.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/federicomarini/ideal" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate a tidy table with the DE genes from the results of DESeq</h1>
    <small class="dont-index">Source: <a href="https://github.com/federicomarini/ideal/blob/HEAD/R/res2tbl.R" class="external-link"><code>R/res2tbl.R</code></a></small>
    <div class="hidden name"><code>deseqresult2DEgenes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a tidy table with the DE genes from the results of DESeq</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">deseqresult2DEgenes</span><span class="op">(</span><span class="va">deseqresult</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>deseqresult</dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html" class="external-link">DESeqResults</a></code> object</p></dd>


<dt>FDR</dt>
<dd><p>Numeric value, the significance level for thresholding adjusted p-values</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A "tidy" data.frame with only genes marked as differentially expressed</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># with simulated data...</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/makeExampleDESeqDataSet.html" class="external-link">makeExampleDESeqDataSet</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, m <span class="op">=</span> <span class="fl">8</span>, betaSD <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">deseqresult2DEgenes</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            id   baseMean log2FoldChange     lfcSE      stat       pvalue</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene73 gene73 159.046069       4.244326 0.4623663  9.179576 4.327926e-20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene40 gene40 610.470607       3.529247 0.4586178  7.695401 1.410517e-14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene42 gene42  71.612573       2.817219 0.4525573  6.225109 4.812202e-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5   gene5  65.523447       3.154648 0.5381724  5.861779 4.579335e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene36 gene36  48.772285      -3.343422 0.5828485 -5.736348 9.674001e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene74 gene74 229.841405       2.367616 0.4143246  5.714398 1.100930e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene72 gene72 164.536955       2.848661 0.5206205  5.471666 4.458249e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene27 gene27  40.385721       4.362489 0.8611647  5.065801 4.066869e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene26 gene26 130.739989      -1.836943 0.3648586 -5.034671 4.786691e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9   gene9  21.910881      -3.198543 0.6402995 -4.995386 5.871826e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7   gene7  13.606113      -5.257772 1.0678894 -4.923517 8.500228e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene86 gene86  30.789347      -2.985548 0.6136684 -4.865084 1.144082e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene61 gene61  61.036687      -2.194397 0.4546204 -4.826879 1.386897e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene58 gene58  17.222718      -3.262287 0.7277067 -4.482969 7.361162e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene30 gene30  99.103357      -2.141405 0.4809889 -4.452088 8.503932e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene32 gene32  24.015400       3.896497 0.8798725  4.428479 9.489996e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene56 gene56  57.988010       2.411936 0.5662139  4.259761 2.046452e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene82 gene82 132.671768       2.490235 0.5878539  4.236146 2.273894e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene71 gene71   8.772735      -5.233068 1.2527687 -4.177202 2.951171e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene25 gene25  34.309814       2.600892 0.6262576  4.153071 3.280436e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene23 gene23  62.453330      -1.792103 0.4507716 -3.975634 7.019208e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene31 gene31  11.307524      -3.312569 0.9009529 -3.676739 2.362345e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene38 gene38  56.861614       2.217419 0.6815784  3.253359 1.140492e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene12 gene12  10.781049       2.677839 0.8406589  3.185404 1.445518e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene60 gene60  11.023779      -2.450048 0.7919950 -3.093515 1.978008e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene29 gene29   6.981685       3.543255 1.1873143  2.984261 2.842645e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene95 gene95  19.721867      -1.903890 0.6400424 -2.974631 2.933407e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene47 gene47  43.621141       1.470279 0.5033448  2.921017 3.488902e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene50 gene50 120.031775       1.381421 0.4737079  2.916186 3.543391e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene24 gene24   8.420716      -2.535457 0.9512418 -2.665418 7.689261e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene84 gene84 140.691703       1.409270 0.5543674  2.542123 1.101814e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene67 gene67   2.213289      -4.708244 1.9686496 -2.391611 1.677462e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                padj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene73 4.111530e-18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene40 6.699954e-13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene42 1.523864e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5  1.087592e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene36 1.743139e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene74 1.743139e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene72 6.050480e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene27 4.829407e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene26 5.052618e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9  5.578234e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  7.341106e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene86 9.057317e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene61 1.013501e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene58 4.995074e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene30 5.385824e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene32 5.634685e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene56 1.143606e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene82 1.200111e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene71 1.475585e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene25 1.558207e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene23 3.175356e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene31 1.020104e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene38 4.710727e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene12 5.721843e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene60 7.516429e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene29 1.032125e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene95 1.032125e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene47 1.160766e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene50 1.160766e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene24 2.434933e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene84 3.376526e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene67 4.979966e-02</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://federicomarini.github.io" class="external-link">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

