<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interactive Differential Expression AnaLysis • ideal</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Interactive Differential Expression AnaLysis">
<meta property="og:description" content="This package provides functions for an Interactive
    Differential Expression AnaLysis of RNA-sequencing datasets, to
    extract quickly and effectively information downstream the step
    of differential expression. A Shiny application encapsulates
    the whole package.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/federicomarini/ideal">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">


<div id="ideal---interactive-differential-expression-analysis-in-rna-seq-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#ideal---interactive-differential-expression-analysis-in-rna-seq-data" class="anchor"></a><code>ideal</code> - Interactive Differential Expression AnaLysis in RNA-seq data</h1></div>
<p><code>ideal</code> is a Bioconductor package containing a Shiny application for analyzing RNA-Seq data in the context of differential expression. This enables an interactive and at the same time reproducible analysis, keeping the functionality accessible, and yet providing a comprehensive selection of graphs and tables to mine the datasetat hand.</p>
<p><code>ideal</code> is an R package which fully leverages the infrastructure of the Bioconductor project in order to deliver an interactive yet reproducible analysis for the detection of differentially expressed genes in RNA-Seq datasets. Graphs, tables, and interactive HTML reports can be readily exported and shared across collaborators. The dynamic user interface displays a broad level of content and information, subdivided by thematic tasks. All in all, it aims to enforce a proper analysis, by reaching out both life scientists and experienced bioinformaticians, and also fosters the communication between the two sides, offering robust statistical methods and high standard of accessible documentation.</p>
<p>It is structured in a similar way to the <code>pcaExplorer</code>, also designed as an interactive companion tool for RNA-seq analysis focused rather on the exploratory data analysis e.g. using principal components analysis as a main tool.</p>
<p>The interactive/reactive design of the app, with a dynamically generated user interface makes it easy and immediate to apply the gold standard methods in a way that is information-rich and accessible also to the bench biologist, while also providing additional insight also for the experienced data analyst. Reproducibility is supported via state saving and automated report generation.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><code>ideal</code> can be easily installed using <code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install()</a></code>:</p>
<pre><code>if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("ideal")</code></pre>
<p>or, optionally,</p>
<pre><code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install("federicomarini/ideal")
# or alternatively...
devtools::install_github("federicomarini/ideal")</a></code></pre>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h2>
<p>This command loads the <code>ideal</code> package</p>
<pre><code><a href="https://www.rdocumentation.org/packages/base/topics/library">library("ideal")</a></code></pre>
<p>The main parameters for <code>ideal</code> are</p>
<ul>
<li>
<code>dds_obj</code> - a <code>DESeqDataSet</code> object. If not provided, then a <code>countmatrix</code> and a <code>expdesign</code> need to be provided. If none of the above is provided, it is possible to upload the data during the execution of the Shiny App</li>
<li>
<code>res_obj</code> - a <code>DESeqResults</code> object. If not provided, it can be computed during the execution of the application</li>
<li>
<code>annotation_obj</code> - a <code>data.frame</code> object, with row.names as gene identifiers (e.g. ENSEMBL ids) and a column, <code>gene_name</code>, containing e.g. HGNC-based gene symbols. If not provided, it can be constructed during the execution via the <code>org.eg.XX.db</code> packages</li>
<li>
<code>countmatrix</code> - a count matrix, with genes as rows and samples as columns. If not provided, it is possible to upload the data during the execution of the Shiny App</li>
<li>
<code>expdesign</code> -a <code>data.frame</code> containing the info on the experimental covariates of each sample. If not provided, it is possible to upload the data during the execution of the Shiny App</li>
</ul>
<p>The <code>ideal</code> app can be launched in different modes:</p>
<ul>
<li>
<code><a href="reference/ideal.html">ideal(dds_obj = dds, res_obj = res, annotation_obj = anno)</a></code>, where the objects are precomputed in the current session and provided as parameters</li>
<li>
<code><a href="reference/ideal.html">ideal(dds_obj = dds)</a></code>, as in the command above, but where the result object is assembled at runtime</li>
<li>
<code><a href="reference/ideal.html">ideal(countmatrix = countmatrix, expdesign = expdesign)</a></code>, where instead of passing the defined <code>DESeqDataSet</code> object, its components are given, namely the count matrix (e.g. generated after a run of featureCounts or HTSeq-count) and a data frame with the experimental covariates. The design formula can be constructed interactively at runtime</li>
<li>
<code><a href="reference/ideal.html">ideal()</a></code>, where the count matrix and experimental design can simply be uploaded at runtime, where all the derived objects can be extracted and computed live. These files have to be formatted as tabular text files, and a function in the package tries to guess the separator, based on heuristics of occurrencies per line of commonly used characters</li>
</ul>
</div>
<div id="accessing-the-public-instance-of-ideal" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-the-public-instance-of-ideal" class="anchor"></a>Accessing the public instance of <code>ideal</code>
</h2>
<p>To use <code>ideal</code> without installing any additional software, you can access the public instance of the Shiny Server made available at the Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI) in Mainz.</p>
<p>This resource is accessible at this address:</p>
<p><a href="http://shiny.imbei.uni-mainz.de:3838/ideal" class="uri">http://shiny.imbei.uni-mainz.de:3838/ideal</a></p>
</div>
<div id="deploying-to-a-shiny-server" class="section level2">
<h2 class="hasAnchor">
<a href="#deploying-to-a-shiny-server" class="anchor"></a>Deploying to a Shiny Server</h2>
<p>A deployment-oriented version of the package is available at <a href="https://github.com/federicomarini/ideal_serveredition" class="uri">https://github.com/federicomarini/ideal_serveredition</a>. This repository contains also detailed instruction to setup the running instance of a Shiny Server, where <code>ideal</code> can be run without further installation for the end-users.</p>
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>For additional details regarding the functions of <strong>ideal</strong>, please consult the documentation or write an email to <a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>.</p>
<div id="bug-reportsissuesnew-features" class="section level3">
<h3 class="hasAnchor">
<a href="#bug-reportsissuesnew-features" class="anchor"></a>Bug reports/Issues/New features</h3>
<p>Please use <a href="https://github.com/federicomarini/ideal/issues" class="uri">https://github.com/federicomarini/ideal/issues</a> for reporting bugs, issues or for suggesting new features to be implemented.</p>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/ideal">https://​www.bioconductor.org/​packages/​ideal</a>
</li>
<li>Browse source code at <br><a href="https://github.com/federicomarini/ideal">https://​github.com/​federicomarini/​ideal</a>
</li>
<li>Report a bug at <br><a href="https://github.com/federicomarini/ideal/issues">https://​github.com/​federicomarini/​ideal/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing ideal</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://federicomarini.github.io">Federico Marini</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3252-7758" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/federicomarini/ideal"><img src="https://travis-ci.org/federicomarini/ideal.svg?branch=master" alt="Linux Build Status"></a></li>
<li><a href="https://codecov.io/github/federicomarini/ideal?branch=master"><img src="https://codecov.io/github/federicomarini/ideal/coverage.svg?branch=master" alt="codecov.io"></a></li>
<li><a href="https://ci.appveyor.com/project/federicomarini/ideal"><img src="https://ci.appveyor.com/api/projects/status/github/federicomarini/ideal?svg=true" alt="Windows Build status"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by <a href="https://federicomarini.github.io">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
