<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ideal User's Guide • ideal</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ideal User's Guide">
<meta property="og:description" content="ideal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/federicomarini/ideal">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ideal-usersguide_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="ideal-usersguide_files/jquery-1.12.4/jquery.min.js"></script><link href="ideal-usersguide_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="ideal-usersguide_files/datatables-binding-0.13/datatables.js"></script><link href="ideal-usersguide_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="ideal-usersguide_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="ideal-usersguide_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="ideal-usersguide_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="ideal-usersguide_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ideal User’s Guide</h1>
                        <h4 class="author">Federico Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                              <h4 class="author">Harald Binder</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz<br><h4 class="date">26 May 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/federicomarini/ideal/blob/master/vignettes/ideal-usersguide.Rmd"><code>vignettes/ideal-usersguide.Rmd</code></a></small>
      <div class="hidden name"><code>ideal-usersguide.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      In the scope of differential expression analysis, we provide the R/Bioconductor package <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>, which serves as a web application to allow simultaneously for interactive and reproducible analysis. <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> guides the user throughout the steps of Differential Expression analysis and produces a wealth of effective visualizations to facilitate data interpretation, in a comprehensive and accessible way for a wide range of scientists.
    </div>
    
<!-- **Package**: *ideal* -->
<!-- **Authors**: Federico Marini [aut, cre] (<https://orcid.org/0000-0003-3252-7758>) -->
<!-- **Version**: 1.12.1 -->
<p><strong>Compiled date</strong>: 2020-05-26</p>
<p><strong>Last edited</strong>: 2019-02-06</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<p align="center">
<img src="ideal_logo_v2.png"></p>
<p><em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> is an R package distributed as part of the <a href="http://bioconductor.org">Bioconductor</a> project. To install the package, start R and enter:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"ideal"</span>)</pre></body></html></div>
<p>The GitHub repository for <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> is <a href="https://github.com/federicomarini/ideal" class="uri">https://github.com/federicomarini/ideal</a>. This is the place to file an issue, report a bug, or provide a pull request.</p>
<p>Once <em>ideal</em> is installed, it can be loaded by the following command.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ideal"</span>)</pre></body></html></div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> is a Bioconductor package containing a Shiny application for analyzing RNA-Seq data in the context of differential expression. This enables an interactive and at the same time analysis, keeping the functionality accessible, and yet providing a comprehensive selection of graphs and tables to mine the dataset at hand.</p>
<p><em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> is an R package which fully leverages the infrastructure of the Bioconductor project in order to deliver an interactive yet reproducible analysis for the detection of differentially expressed genes in RNA-Seq datasets. Graphs, tables, and interactive HTML reports can be readily exported and shared across collaborators. The dynamic user interface displays a broad level of content and information, subdivided by thematic tasks. All in all, it aims to enforce a proper analysis, by reaching out both life scientists and experienced bioinformaticians, and also fosters the communication between the two sides, offering robust statistical methods and high standard of accessible documentation.</p>
<p>It is structured in a similar way to the <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em>, also designed as an interactive companion tool for RNA-seq analysis focused rather on the exploratory data analysis e.g. using principal components analysis as a main tool.</p>
<p>The interactive/reactive design of the app, with a dynamically generated user interface makes it easy and immediate to apply the gold standard methods (in the current implementation, based on <em><a href="https://bioconductor.org/packages/3.12/DESeq2">DESeq2</a></em>) in a way that is information-rich and accessible also to the bench biologist, while also providing additional insight also for the experienced data analyst. Reproducibility is supported via state saving and automated report generation.</p>
<div id="citation-info" class="section level2">
<h2 class="hasAnchor">
<a href="#citation-info" class="anchor"></a>Citation info</h2>
<p>If you use <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> for your analysis, please cite it as here below:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span>(<span class="st">"ideal"</span>)</pre></body></html></div>
<pre><code>## 
## To cite the package 'ideal' in publications use:
## 
##   Federico Marini (2020). ideal: an R/Bioconductor package for
##   Interactive Differential Expression Analysis, bioRxiv URL
##   http://bioconductor.org/packages/ideal/ doi:
##   10.1101/2020.01.10.901652
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {ideal: an R/Bioconductor package for Interactive Differential Expression Analysis},
##     journal = {bioRxiv},
##     author = {Federico Marini and Jan Linke and Harald Binder},
##     year = {2020},
##     doi = {10.1101/2020.01.10.901652},
##     url = {http://bioconductor.org/packages/ideal/},
##   }</code></pre>
</div>
</div>
<div id="using-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#using-the-application" class="anchor"></a>Using the application</h1>
<p>There are different ways to use <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> for interactive differential expression analysis.</p>
<div id="launching-ideal-locally" class="section level2">
<h2 class="hasAnchor">
<a href="#launching-ideal-locally" class="anchor"></a>Launching <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> locally</h2>
<p>First load the library:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ideal"</span>)</pre></body></html></div>
<p>and then launch the app with the <code>ideal</code> function. This takes the following essential parameters as input:</p>
<ul>
<li>
<code>dds_obj</code> - a <code>DESeqDataSet</code> object. If not provided, then a <code>countmatrix</code> and a <code>expdesign</code> need to be provided. If none of the above is provided, it is possible to upload the data during the execution of the Shiny App</li>
<li>
<code>res_obj</code> - a <code>DESeqResults</code> object. If not provided, it can be computed during the execution of the application</li>
<li>
<code>annotation_obj</code> - a <code>data.frame</code> object, with row.names as gene identifiers (e.g. ENSEMBL ids) and a column, <code>gene_name</code>, containing e.g. HGNC-based gene symbols. If not provided, it can be constructed during the execution via the <code>org.eg.XX.db</code> packages</li>
<li>
<code>countmatrix</code> - a count matrix, with genes as rows and samples as columns. If not provided, it is possible to upload the data during the execution of the Shiny App</li>
<li>
<code>expdesign</code> -a <code>data.frame</code> containing the info on the experimental covariates of each sample. If not provided, it is possible to upload the data during the execution of the Shiny App</li>
</ul>
<p>Different modalities are supported to launch the application:</p>
<ul>
<li>
<code><a href="../reference/ideal.html">ideal(dds_obj = dds, res_obj = res, annotation_obj = anno)</a></code>, where the objects are precomputed in the current session and provided as parameters</li>
<li>
<code><a href="../reference/ideal.html">ideal(dds_obj = dds)</a></code>, as in the command above, but where the result object is assembled at runtime</li>
<li>
<code><a href="../reference/ideal.html">ideal(countmatrix = countmatrix, expdesign = expdesign)</a></code>, where instead of passing the defined <code>DESeqDataSet</code> object, its components are given, namely the count matrix (e.g. generated after a run of featureCounts or HTSeq-count) and a data frame with the experimental covariates. The design formula can be constructed interactively at runtime</li>
<li>
<code><a href="../reference/ideal.html">ideal()</a></code>, where the count matrix and experimental design can simply be uploaded at runtime, where all the derived objects can be extracted and computed live. These files have to be formatted as tabular text files, and a function in the package tries to guess the separator, based on heuristics of occurrencies per line of commonly used characters</li>
</ul>
</div>
<div id="accessing-the-public-instance-of-ideal" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-the-public-instance-of-ideal" class="anchor"></a>Accessing the public instance of <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>
</h2>
<p>To use <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> without installing any additional software, you can access the public instance of the Shiny Server made available at the Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI) in Mainz.</p>
<p>This resource is accessible at this address:</p>
<p><a href="http://shiny.imbei.uni-mainz.de:3838/ideal" class="uri">http://shiny.imbei.uni-mainz.de:3838/ideal</a></p>
</div>
<div id="deploying-to-a-shiny-server" class="section level2">
<h2 class="hasAnchor">
<a href="#deploying-to-a-shiny-server" class="anchor"></a>Deploying to a Shiny Server</h2>
<p>A deployment-oriented version of the package is available at <a href="https://github.com/federicomarini/ideal_serveredition" class="uri">https://github.com/federicomarini/ideal_serveredition</a>. This repository contains also detailed instruction to setup the running instance of a Shiny Server, where <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> can be run without further installation for the end-users.</p>
<p>Please note that you still need <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> to be installed there once during the setup phase - for this operation, you might require root administrator permissions.</p>
<!-- Following the structure of \texttt{pcaExplorer}, the deployment as a standalone web application is easily achieved also for \texttt{ideal}, with the only requirement of a running Shiny Server. Handling of sensitive human patient data can thus occur in a secured way, e.g. with restricted access to an internal instance of the application. This will also avoid the need of external software installation for the users, which could moreover seamlessly access higher performing computing infrastructures. -->
<!-- This implementation of \texttt{ideal} is made available at github.com/federicomarini/ideal\_serveredition  -->
</div>
</div>
<div id="getting-to-know-the-user-interface-and-the-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-to-know-the-user-interface-and-the-functionality" class="anchor"></a>Getting to know the user interface and the functionality</h1>
<p>The user interface is dynamically displayed according to the provided and computed objects, with tabs that are actively usable only once the required input is effectively available.</p>
<p>Moreover, for some relevant UI widgets, the user can receive additional information by hovering over with the mouse, with the functionality powered by the <em><a href="https://CRAN.R-project.org/package=shinyBS">shinyBS</a></em> package.</p>
<p>For the user which is either new with the app UI/functionality, or not extensively familiar with the topic of differential expression, it is possible to obtain a small <em>guided tour</em> of the App by clicking on the respective help buttons, marked in the app like this - please note that this button is clickable but does not start any tour.</p>
<button id="btn" type="button" class="btn btn-default action-button" style="color: #ffffff; background-color: #0092AC; border-color: #2e6da4">
<i class="fa fa-info"></i> Click me for a quick tour
</button>
<p>These trigger the start of a step-by-step guide and feature introduction, powered by the <em><a href="https://CRAN.R-project.org/package=rintrojs">rintrojs</a></em> package.</p>
<div id="the-controls-sidebar" class="section level2">
<h2 class="hasAnchor">
<a href="#the-controls-sidebar" class="anchor"></a>The controls sidebar</h2>
<p>Some of the input controls which affect different tabs are located in the sidebar, while others are as well in the individual tabs of the app. By changing one or more of the input parameters, the user can get a fine control on what is computed and displayed.</p>
<div id="app-settings" class="section level3">
<h3 class="hasAnchor">
<a href="#app-settings" class="anchor"></a>App settings</h3>
<ul>
<li>
<strong>Group/color by</strong> - Select the group of samples to stratify the analysis for plotting. Can also assume multiple values.</li>
<li>
<strong>Select the gene(s) of interest - ids</strong> - Select a subset of genes for deeper analysis. If an annotation object is provided, the user can handily select the genes e.g. based on their HGNC symbol</li>
<li>
<strong>False Discovery Rate</strong> - Set as default to 0.05, it is the FDR value for the Benjamini-Hochberg procedure for adjusting p-values in the multiple testing comparison scenario</li>
</ul>
</div>
<div id="plot-export-settings" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-export-settings" class="anchor"></a>Plot export settings</h3>
<p><strong>Width</strong> and <strong>Height</strong> for the figures to export are input here in cm.</p>
</div>
<div id="quick-viewer" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-viewer" class="anchor"></a>Quick viewer</h3>
<p>This displays a list of the underlying objects with which basically all of the analysis can be performed. A green tick icon appears close to each when the respective component is either provided or calculated. For obtaining the best analysis experience in <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>, it is recommended to provide all of them.</p>
</div>
<div id="first-steps-help" class="section level3">
<h3 class="hasAnchor">
<a href="#first-steps-help" class="anchor"></a>First steps help</h3>
<p>Clicking on this button activated the <code>intro.js</code> based tour for getting to know the components and the structure of the app. Dedicated step-by-step procedures are also available in each individual tab.</p>
</div>
</div>
<div id="the-task-menu" class="section level2">
<h2 class="hasAnchor">
<a href="#the-task-menu" class="anchor"></a>The task menu</h2>
<p>The task menu, accessible by clicking on the cog icon in the upper right part of the application, provides two functionalities:</p>
<ul>
<li>
<code>Exit ideal &amp; save</code> will close the application and store the content of the <code>input</code> and <code>values</code> reactive objects in a list of two elements in the <code>ideal_env</code> environment, respectively called <code>ideal_inputs_YYYYMMDD_HHMMSS</code> and <code>ideal_values_YYYYMMDD_HHMMSS</code>
</li>
<li>
<code>Save State as .RData</code> will similarly store <code>LiveInputs</code> and <code>r_data</code> in a binary file named <code>idealState_YYYYMMDD_HHMMSS.Rdata</code>, without closing the application</li>
</ul>
</div>
</div>
<div id="the-main-app-panels" class="section level1">
<h1 class="hasAnchor">
<a href="#the-main-app-panels" class="anchor"></a>The main app panels</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> app is a one-paged dashboard, structured in different panels, where each of them is focused on a different aspect of the data exploration.</p>
<p>On top of the panels, three <code>valueBox</code> objects serve as guiding elements for having an overview of the data at hand: how many genes and samples are in the data, how many entries are in the annotation object, and how many genes were found to be differentially expressed in the results. Whenever each of the underlying objects is available, the background color turns from red to green.</p>
<p>For the main analysis, the available panels are described in the following subsections.</p>
<div id="welcome" class="section level2">
<h2 class="hasAnchor">
<a href="#welcome" class="anchor"></a>Welcome!</h2>
<p>The landing page for the app is also where you might likely be reading this text (otherwise in the package vignette).</p>
</div>
<div id="data-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#data-setup" class="anchor"></a>Data Setup</h2>
<p>The Data Setup panel is where you can upload or inspect the required inputs for running the app. This builds on the primary idea used by <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em> and extends it with the following aspects:</p>
<ul>
<li>the panel structure appears dynamically in three consecutive mandatory steps, marked with color from red to yellow to green, with optional steps in light blue.</li>
<li>the optional step of retrieving the annotation on the fly relieves the user from the task of composing the <code>data.frame</code> in advance, and is based on the widely adopted <code>org.XX.eg.db</code> Bioconductor packages.</li>
<li>when the objects are already passed as parameters, or computed, a brief overview/summary for them is displayed</li>
<li>to tighten the concert operations between similar tools with different scope (as <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em> and <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> are), the information flow can move from the data exploration to decisions taken at the moment of testing</li>
</ul>
<p>A diagnostic mean-dispersion plot is also provided in a collapsible element at the bottom of the panel, shown when the <code>DESeqDataSet</code> is generated and the <code>DESeq</code> command from the <code>DESeq2</code> package has been applied.</p>
</div>
<div id="counts-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#counts-overview" class="anchor"></a>Counts Overview</h2>
<p align="center">
<img src="ss_05_overview.png" style="width:800px;"></p>
<p>As in <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em>, interactive tables for the raw, normalized or (r)log-transformed counts are shown in this tab. The user can also generate a sample-to-sample correlation scatter plot with the selected data.</p>
<p>Additionally, <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> has an option to include a filter step at the gene level by removing genes with low absolute or averages low values. After this, it might be possible to have to re-run the analysis in step 3 from the Data Setup panel.</p>
</div>
<div id="extract-results" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-results" class="anchor"></a>Extract Results</h2>
<p>This tab is an interface for generating the summary tables after testing for DE. It is usually based on the Wald test, as implemented in DESeq2, but when the factor of interest is assuming more than two levels, the user can also perform an ANOVA-like test across the groups with the likelihood ratio test. Options for enabling/disabling automated independent filtering, adding the additional column of unshrunken log2 fold change values (instead of the moderated estimates used by default), as well as using the Independent Hypothesis Weighting (<em><a href="https://bioconductor.org/packages/3.12/IHW">IHW</a></em>) framework, are provided.</p>
<p>The False Discovery Rate (FDR) can be set from the sidebar panel, and a couple of diagnostic plots, such as the histogram of raw p-values and the distribution of log2fc, are shown below the interactive enhanced version of the table - with clickable elements to link to ENSEMBL database and NCBI website.</p>
</div>
<div id="summary-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#summary-plots" class="anchor"></a>Summary Plots</h2>
<p align="center">
<img src="ss_08_summaryplots.png" style="width:800px;"></p>
<p>In this tab an interactive MA plot for the contrast selected in the Extract Results tab is displayed. Clicking on a single gene in the zoomed plot (enabled by brushing in the main plot), it is possible to obtain a boxplot for its expression values, flanked by an overview of information accessed live from the Entrez database. Alternatively, a volcano plot of -log10(p-value) versus log fold change can provide a slightly different perspective. The subset of selected genes are also here presented in static and interactive heatmaps, with the underlying data accessible from the collapsible box element.</p>
</div>
<div id="gene-finder" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-finder" class="anchor"></a>Gene Finder</h2>
<p align="center">
<img src="ss_09_genefinder.png" style="width:800px;"></p>
<p>The functionality in the Gene Finder builds upon the one provided by <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em>, and allows to query up to four genes in the same view, which can here be selected from a dropdown input list which supports autocompletion.</p>
<p>A combined summary table (with both normalized counts and results statistics) is located below an MA plot where the selected genes are marked and annotated on the plot. To avoid repeating this manually, the user can also quickly upload a list of genes as text file (one gene identifier per line), such as members of gene families (e.g. all cytokines, all immunoglobulines, …) or defined by common function (e.g. all housekeeping genes, or others based on any annotation).</p>
</div>
<div id="functional-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#functional-analysis" class="anchor"></a>Functional Analysis</h2>
<p align="center">
<img src="ss_10_funcanalysis.png" style="width:800px;"></p>
<p>The Functional Analysis tab takes the user from the simple lists of DE genes to insight on the affected biological pathways, with three approaches based on the Gene Ontology (GO) databases. This panel of ideal has a slim interface to the following methods for performing functional analysis:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/limma/man/goana.html">limma::goana</a></code> for the quick yet standard implementation</li>
<li>
<code>topGO</code>, particularly valuable for pruning terms which are topologically less meaningful than their specific nodes</li>
<li>
<code>goseq</code>, which accounts for the specific length bias intrinsic in RNA-Seq assays (longer genes have higher chances of being called DE).</li>
</ul>
<p><em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> allows the user to work simultaneously with more gene lists, two of which can be uploaded in a custom way (e.g. list of gene families, or extracted from other existing publications).</p>
<p>The interaction among these lists can be visually represented in Venn diagrams, as well as with the appealing alternative from the UpSetR package, where all combination of sets are explicitly shown.</p>
<p>Each of the methods for GO enrichment delivers its own interactive <code>DT</code>-based table, which can then be explored interactively with the display of a heatmap for all the (DE) genes annotated to a particular term, picking the normalized transformed values for comparing robustly the expression values. This is simply triggered by clicking any of the rows for the results tables. Another useful feature is provided by the clickable link to the AmiGO database on each of the GO term identifiers.</p>
</div>
<div id="signatures-explorer" class="section level2">
<h2 class="hasAnchor">
<a href="#signatures-explorer" class="anchor"></a>Signatures Explorer</h2>
<p>The Signatures Explorer tab allows the user to check the behavior of a number of provided gene signatures in the data at hand, displaying this as a heatmap.</p>
<p>This panel is composed by different well panels:</p>
<ul>
<li><p>in the Setup Options, you can select and upload a gene signature file, in <code>gmt</code> format (e.g. like the ones provided in the MSigDB database, or from WikiPathways), and quickly compute the variance stabilized transformed version of your data, which is more amenable for visualization than raw or normalized counts</p></li>
<li><p>in the Conversion options tab, you can create an annotation vector, used to bring the ids from your data and the ids the <code>gmt</code> used for encoding the signature elements. This works based on the <code>org.XX.eg.db</code> packages.</p></li>
<li><p>the lower well panels control the appearance of the heatmap, also with an option to display all genes annotated in that pathway, or only the ones detected as differentially expressed (for this you need to provide or compute the result object)</p></li>
</ul>
</div>
<div id="report-editor" class="section level2">
<h2 class="hasAnchor">
<a href="#report-editor" class="anchor"></a>Report Editor</h2>
<p>The Report Editor tab works in the same way of <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em>, with the scope of providing an interface to full computational reproducibility of the analyses.</p>
<p>General <code>Markdown options</code> and <code>Editor options</code> are available, and the text editor, based on the <code>shinyAce</code> package, contains a comprehensive template report, that can be edited to the best convenience of the user.</p>
<p>The code contained in the template report fetches the latest state of the reactive values in the ongoing session, and its output is a comprehensive HTML file that can be expanded, edited, previewed in the tab itself, downloaded, and shared with a few mouse clicks.</p>
<p>As for <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em>, the functionality to display the report preview is based on <code>knit2html</code>, and some elements such as <code>DataTable</code> objects might not render correctly. To render them correctly, please install the PhantomJS executable before launching the app. This can be done by using the <em><a href="https://CRAN.R-project.org/package=webshot">webshot</a></em> package and calling <code><a href="https://rdrr.io/pkg/webshot/man/install_phantomjs.html">webshot::install_phantomjs()</a></code> - HTML widgets will be rendered automatically as screenshots. Keep in mind that the fully rendered report (the one you can obtain with the “Generate &amp; Save” button) is not affected by this, since it uses <code><a href="https://rdrr.io/pkg/rmarkdown/man/render.html">rmarkdown::render()</a></code>.</p>
</div>
<div id="about" class="section level2">
<h2 class="hasAnchor">
<a href="#about" class="anchor"></a>About</h2>
<p>The About tab contains the output of <code>sessionInfo</code>, plus general information on <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>, including the link to the Github development version. If requested, the modular structure of the app can be easily expanded, and many new operations on the same set of input data and derived results can be embedded in the same framework.</p>
</div>
</div>
<div id="running-ideal-on-an-exemplary-data-set" class="section level1">
<h1 class="hasAnchor">
<a href="#running-ideal-on-an-exemplary-data-set" class="anchor"></a>Running <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> on an exemplary data set</h1>
<p>We can run <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> for demonstration purpose on published datasets that are available as SummarizedExperiment in an experiment Bioconductor packages.</p>
<p>We will use the <em><a href="https://bioconductor.org/packages/3.12/airway">airway</a></em> dataset, which can be installed with this command:</p>
<!-- [][]or use pasilla[][] -->
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"airway"</span>)</pre></body></html></div>
<p>This package provides a <code>RangedSummarizedExperiment</code> object of read counts in genes for an RNA-Seq experiment on four human airway smooth muscle cell lines treated with dexamethasone. More details such as gene models and count quantifications can be found in the <em><a href="https://bioconductor.org/packages/3.12/airway">airway</a></em> package vignette.</p>
<p>To run <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> on this dataset, the following commands are required. First, prepare the objects to be passed as parameters of <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">airway</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DESeq2</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">airway</span>)

<span class="no">dds_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSet</a></span>(<span class="no">airway</span>,<span class="kw">design</span><span class="kw">=</span> ~ <span class="no">cell</span> + <span class="no">dex</span>)
<span class="no">dds_airway</span></pre></body></html></div>
<pre><code>## class: DESeqDataSet 
## dim: 64102 8 
## metadata(2): '' version
## assays(1): counts
## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99
## rowData names(0):
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># run deseq on it</span>
<span class="no">dds_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(<span class="no">dds_airway</span>)
<span class="co"># extract the results</span>
<span class="no">res_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(<span class="no">dds_airway</span>, <span class="kw">contrast</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dex"</span>,<span class="st">"trt"</span>,<span class="st">"untrt"</span>),<span class="kw">alpha</span> <span class="kw">=</span> <span class="fl">0.05</span>)</pre></body></html></div>
<p>Then launch the app itself.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="../reference/ideal.html">ideal</a></span>(<span class="kw">dds_obj</span> <span class="kw">=</span> <span class="no">dds_airway</span>)
<span class="co"># or also providing the results object</span>
<span class="fu"><a href="../reference/ideal.html">ideal</a></span>(<span class="kw">dds_obj</span> <span class="kw">=</span> <span class="no">dds_airway</span>,<span class="kw">res_obj</span> <span class="kw">=</span> <span class="no">res_airway</span>)</pre></body></html></div>
<p>The <code>annotation</code> for this dataset can be built manually by exploiting the <em><a href="https://bioconductor.org/packages/3.12/org.Hs.eg.db">org.Hs.eg.db</a></em> package.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)
<span class="no">genenames_airway</span> <span class="kw">&lt;-</span> <span class="fu">mapIds</span>(<span class="no">org.Hs.eg.db</span>,<span class="kw">keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dds_airway</span>),<span class="kw">column</span> <span class="kw">=</span> <span class="st">"SYMBOL"</span>,<span class="kw">keytype</span><span class="kw">=</span><span class="st">"ENSEMBL"</span>)
<span class="no">annotation_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">gene_id</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dds_airway</span>),
                                <span class="kw">gene_name</span> <span class="kw">=</span> <span class="no">genenames_airway</span>,
                                <span class="kw">row.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dds_airway</span>),
                                <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annotation_airway</span>)</pre></body></html></div>
<pre><code>##                         gene_id gene_name
## ENSG00000000003 ENSG00000000003    TSPAN6
## ENSG00000000005 ENSG00000000005      TNMD
## ENSG00000000419 ENSG00000000419      DPM1
## ENSG00000000457 ENSG00000000457     SCYL3
## ENSG00000000460 ENSG00000000460  C1orf112
## ENSG00000000938 ENSG00000000938       FGR</code></pre>
<p>or alternatively, can be handily created at runtime in the optional step.</p>
<p>Then again, the app can be launched with:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="../reference/ideal.html">ideal</a></span>(<span class="kw">dds_obj</span> <span class="kw">=</span> <span class="no">dds_airway</span>,
      <span class="kw">annotation_obj</span> <span class="kw">=</span> <span class="no">annotation_airway</span>)</pre></body></html></div>
<p>If desired, alternatives can be used. See the well written annotation workflow available at the Bioconductor site (<a href="https://bioconductor.org/help/workflows/annotation/annotation/" class="uri">https://bioconductor.org/help/workflows/annotation/annotation/</a>).</p>
<div id="coming-from-edgerlimma-voom" class="section level2">
<h2 class="hasAnchor">
<a href="#coming-from-edgerlimma-voom" class="anchor"></a>Coming from <code>edgeR</code>/<code>limma-voom</code>
</h2>
<p>Let’s suppose you performed part of your analysis with <em><a href="https://bioconductor.org/packages/3.12/edgeR">edgeR</a></em> or <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em>/<code>voom</code> and you want to continue inspecting your data interactively using <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em>. You can use the functionality provided by the <em><a href="https://bioconductor.org/packages/3.12/DEFormats">DEFormats</a></em> package to convert the object with:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DEFormats</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">edgeR</span>)</pre></body></html></div>
<pre><code>## Loading required package: limma</code></pre>
<pre><code>## 
## Attaching package: 'limma'</code></pre>
<pre><code>## The following object is masked from 'package:DESeq2':
## 
##     plotMA</code></pre>
<pre><code>## The following object is masked from 'package:BiocGenerics':
## 
##     plotMA</code></pre>
<pre><code>## 
## Attaching package: 'edgeR'</code></pre>
<pre><code>## The following object is masked from 'package:DEFormats':
## 
##     DGEList</code></pre>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">limma</span>)
<span class="no">dge_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEFormats/man/as.DGEList.html">as.DGEList</a></span>(<span class="no">dds_airway</span>) <span class="co"># this is your initial object</span>
<span class="co"># your factors for the design:</span>
<span class="no">dex</span> <span class="kw">&lt;-</span> <span class="fu">colData</span>(<span class="no">dds_airway</span>)$<span class="no">dex</span>
<span class="no">cell</span> <span class="kw">&lt;-</span> <span class="fu">colData</span>(<span class="no">dds_airway</span>)$<span class="no">cell</span>

<span class="no">redo_dds_airway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEFormats/man/as.DESeqDataSet.html">as.DESeqDataSet</a></span>(<span class="no">dge_airway</span>)
<span class="co"># force the design to ~cell + dex</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/design.html">design</a></span>(<span class="no">redo_dds_airway</span>) <span class="kw">&lt;-</span> ~<span class="no">cell</span>+<span class="no">group</span>  <span class="co">#TODO: this is due to the not 100% re-conversion via DEFormats </span>

<span class="co">### with edgeR</span>
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="no">dge_airway</span>)
<span class="no">design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(~ <span class="no">cell</span> + <span class="no">dex</span>)
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span>(<span class="no">y</span>,<span class="no">design</span>)
<span class="co"># If you performed quasi-likelihood F-tests</span>
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span>(<span class="no">y</span>,<span class="no">design</span>)
<span class="no">qlf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span>(<span class="no">fit</span>) <span class="co"># contrast for dexamethasone treatment</span>
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span>(<span class="no">qlf</span>)</pre></body></html></div>
<pre><code>## Coefficient:  dexuntrt 
##                                                                                                  seqnames
## ENSG00000109906 11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11
## ENSG00000152583             4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4
## ENSG00000096060                                                   6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6
## ENSG00000101347          20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20
## ENSG00000211445                                             5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5
## ENSG00000163884                                                                                 3,3,3,3,3
## ENSG00000189221                                 X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X
## ENSG00000157514                 X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X
## ENSG00000162692                                                           1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
## ENSG00000112936                         5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5
##                                                                                                                                                                                                                                                                                                                                                                                             start
## ENSG00000109906                                                                       113930315,113930447,113930459,113930979,113931229,113932093,113933933,113933933,113933933,113933933,113935134,113992540,114027059,114027059,114051488,114057674,114057674,114057674,114107929,114112889,114112889,114117512,114117920,114117920,114117920,114117920,114121048,114121048,114121048,114121048
## ENSG00000152583                    88394487,88394500,88394570,88400582,88401524,88403576,88411425,88411912,88412770,88414734,88414734,88415521,88415521,88415528,88415569,88415592,88415685,88416133,88416133,88416133,88416165,88418283,88420673,88420673,88425974,88449278,88449707,88449707,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88451697,88452130,88452130
## ENSG00000096060                                                                                                                                                                                               35541362,35541367,35543176,35544771,35547813,35548334,35554811,35558907,35565025,35565025,35586873,35587909,35604791,35610497,35656509,35656581,35656581,35656581,35688105,35696285
## ENSG00000101347                                                                                                                                35518632,35525971,35526225,35526843,35532560,35532560,35533767,35533767,35539621,35540751,35540864,35545125,35545352,35547767,35555585,35555585,35559163,35559163,35563432,35563432,35563744,35569442,35569442,35575141,35575141,35579839,35579839
## ENSG00000211445                                                                                                                                             150400124,150400153,150400165,150400167,150400389,150400389,150404901,150404901,150404905,150406006,150406272,150406461,150406461,150406461,150406527,150406993,150406993,150406993,150406993,150407383,150407470,150407470,150407470
## ENSG00000163884                                                                                                                                                                                                                                                                                                                                 126061478,126070684,126071518,126072855,126076080
## ENSG00000189221                                                                                                              43515467,43515548,43515550,43517110,43542761,43542761,43552538,43552538,43571119,43571119,43571119,43571952,43571952,43587420,43587420,43590488,43590488,43590941,43591946,43595474,43599928,43601197,43602951,43603041,43603356,43603356,43603614,43603614,43603614
## ENSG00000157514 106956451,106956454,106957230,106957600,106957717,106957907,106959003,106959128,106959129,106959129,106959129,106959133,106959142,106959153,106959545,106959920,106959920,106965089,107018330,107018330,107018330,107018330,107018348,107018397,107018476,107019010,107019010,107019233,107019376,107019829,107020200,107020200,107020250,107020250,107020250,107020502,107020502
## ENSG00000162692                                                                                                                                                                                                                   101185298,101185311,101185316,101185320,101186032,101186032,101188576,101190180,101194663,101196754,101197974,101200058,101200731,101203679,101203679,101203679
## ENSG00000112936                                                                          40909354,40928682,40931166,40934425,40934427,40936440,40936440,40937340,40937654,40937654,40945300,40945300,40947704,40950006,40955489,40958135,40959551,40962187,40964581,40964843,40972505,40972505,40972505,40972623,40976765,40976852,40976852,40976852,40979827,40979827,40980089,40981494,40981494
##                                                                                                                                                                                                                                                                                                                                                                                               end
## ENSG00000109906                                                                       113930604,113930604,113930604,113931016,113931306,113932283,113934368,113934466,113935290,113935290,113935290,113992636,114027156,114027156,114051664,114057760,114057760,114060486,114108062,114113059,114113059,114118087,114118018,114118066,114118087,114118087,114121198,114121279,114121374,114121398
## ENSG00000152583                    88394955,88394955,88394955,88400730,88401672,88403712,88411545,88412030,88412842,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88416279,88416279,88416295,88416279,88418391,88420737,88420737,88426075,88449385,88449766,88449771,88450244,88450253,88450393,88450524,88450556,88450609,88450655,88450656,88451783,88452201,88452213
## ENSG00000096060                                                                                                                                                                                               35543719,35543719,35543719,35545010,35547998,35554894,35554894,35558997,35565181,35565181,35586987,35588051,35604935,35610620,35656692,35656692,35656714,35656719,35688325,35696360
## ENSG00000101347                                                                                                                                35521469,35526362,35526362,35526947,35532652,35532652,35533797,35533906,35539736,35540955,35540955,35545233,35545452,35547922,35555655,35555655,35559278,35559278,35563592,35563592,35563806,35569514,35569514,35575207,35575207,35580111,35580246
## ENSG00000211445                                                                                                                                             150400302,150400302,150400302,150400302,150400415,150400814,150405054,150405054,150405054,150406578,150406431,150406578,150406578,150406578,150406578,150407022,150407051,150407092,150407092,150407546,150407727,150408515,150408554
## ENSG00000163884                                                                                                                                                                                                                                                                                                                                 126062738,126071790,126071790,126073128,126076285
## ENSG00000189221                                                                                                              43515662,43515662,43515662,43517217,43542855,43542855,43552675,43552675,43571223,43571223,43571223,43572043,43572043,43587561,43587561,43590629,43590637,43591100,43592042,43595527,43599985,43601294,43603152,43603152,43603418,43603418,43603905,43604409,43606068
## ENSG00000157514 106957979,106957979,106957979,106957979,106957979,106957979,106959180,106959180,106959180,106959180,106959631,106959180,106959180,106959180,106959775,106960292,106960365,106965502,107018524,107018741,107018755,107019017,107018755,107018755,107018755,107019215,107019218,107019345,107019488,107019944,107020285,107020297,107020308,107020414,107020572,107020559,107020567
## ENSG00000162692                                                                                                                                                                                                                   101185480,101185480,101185480,101185480,101186121,101186307,101188896,101190446,101194938,101197074,101198240,101200324,101204595,101204169,101204594,101204601
## ENSG00000112936                                                                          40909718,40928737,40931241,40934568,40934568,40936587,40936587,40937589,40937792,40937792,40945462,40945470,40947947,40950116,40955655,40958363,40959722,40962274,40964975,40964975,40972673,40972696,40972696,40972696,40976942,40976942,40976942,40976943,40980011,40980011,40980452,40981868,40983041
##                                                                                                                                                              width
## ENSG00000109906                                 290,158,146,38,78,191,436,534,1358,1358,157,97,98,98,177,87,87,2813,134,171,171,576,99,147,168,168,151,232,327,351
## ENSG00000152583 469,456,386,149,149,137,121,119,73,1017,1017,230,230,223,182,159,66,147,147,163,115,109,65,65,102,108,60,65,48,57,197,328,360,413,459,460,87,72,84
## ENSG00000096060                                                                    2358,2353,544,240,186,6561,84,91,157,157,115,143,145,124,184,112,134,139,221,76
## ENSG00000101347                                                  2838,392,138,105,93,93,31,140,116,205,92,109,101,156,71,71,116,116,161,161,63,73,73,67,67,273,408
## ENSG00000211445                                                          179,150,138,136,27,426,154,154,150,573,160,118,118,118,52,30,59,100,100,164,258,1046,1085
## ENSG00000163884                                                                                                                              1261,1107,273,274,206
## ENSG00000189221                                         196,115,113,108,95,95,138,138,105,105,105,92,92,142,142,142,150,160,97,54,58,98,202,112,63,63,292,796,2455
## ENSG00000157514          1529,1526,750,380,263,73,178,53,52,52,503,48,39,28,231,373,446,414,195,412,426,688,408,359,280,206,209,113,113,116,86,98,59,165,323,58,66
## ENSG00000162692                                                                                    183,170,165,161,90,276,321,267,276,321,267,267,3865,491,916,923
## ENSG00000112936                      365,56,76,144,142,148,148,250,139,139,163,171,244,111,167,229,172,88,395,133,169,192,192,74,178,91,91,92,185,185,364,375,1548
##                                                                                        strand
## ENSG00000109906                   +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000152583 -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000096060                                       -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000101347                         -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000211445                                 +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000163884                                                                     -,-,-,-,-
## ENSG00000189221                     +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000157514     -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000162692                                               +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000112936             +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
##                                                                                                                                                                                                                                                                                          exon_id
## ENSG00000109906                                                                360958,360959,360960,360961,360962,360963,360964,360965,360967,360966,360968,360969,360971,360970,360972,360974,360973,360975,360976,360977,360978,360979,360980,360981,360982,360983,360984,360985,360986,360987
## ENSG00000152583 172689,172690,172691,172692,172693,172694,172695,172696,172697,172699,172698,172700,172701,172702,172703,172704,172705,172707,172706,172708,172709,172710,172711,172712,172713,172714,172715,172716,172717,172718,172719,172720,172721,172722,172723,172724,172725,172726,172727
## ENSG00000096060                                                                                                                                      228563,228564,228565,228566,228567,228568,228569,228570,228571,228572,228573,228574,228575,228576,228578,228579,228580,228581,228583,228584
## ENSG00000101347                                                                                     626560,626561,626562,626563,626565,626564,626566,626567,626568,626569,626570,626571,626572,626573,626575,626574,626576,626577,626578,626579,626580,626581,626582,626584,626583,626585,626586
## ENSG00000211445                                                                                                                 191301,191302,191303,191304,191305,191306,191308,191307,191309,191310,191311,191314,191312,191313,191315,191316,191317,191319,191318,191320,191321,191322,191323
## ENSG00000163884                                                                                                                                                                                                                                               145255,145256,145257,145258,145259
## ENSG00000189221                                                                       653982,653983,653984,653985,653986,653987,653989,653988,653990,653991,653992,653993,653994,653995,653996,653997,653998,653999,654000,654001,654002,654003,654004,654005,654006,654007,654008,654009,654010
## ENSG00000157514               667855,667856,667857,667858,667859,667860,667861,667862,667864,667863,667865,667866,667867,667868,667869,667870,667871,667872,667873,667874,667875,667876,667877,667878,667879,667880,667881,667882,667883,667884,667885,667886,667887,667888,667889,667890,667891
## ENSG00000162692                                                                                                                                                                                  14792,14793,14794,14795,14796,14797,14798,14799,14800,14801,14802,14803,14804,14805,14806,14807
## ENSG00000112936                                           181403,181404,181405,181406,181407,181409,181408,181410,181412,181411,181413,181414,181415,181416,181417,181418,181419,181420,181421,181422,181423,181425,181424,181426,181427,181428,181429,181430,181431,181432,181433,181434,181435
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       exon_name
## ENSG00000109906                                                                                                                                                 ENSE00001344383,ENSE00002217296,ENSE00002239785,ENSE00002278969,ENSE00001777728,ENSE00002277816,ENSE00002264739,ENSE00002267450,ENSE00003532715,ENSE00003522546,ENSE00002223922,ENSE00002239280,ENSE00003606532,ENSE00003503866,ENSE00002304588,ENSE00003667676,ENSE00003652618,ENSE00002237247,ENSE00002308595,ENSE00003574983,ENSE00003648066,ENSE00002216844,ENSE00002254046,ENSE00002202712,ENSE00003533854,ENSE00003675344,ENSE00002251269,ENSE00002289800,ENSE00002238115,ENSE00001513888
## ENSG00000152583 ENSE00001914146,ENSE00001135423,ENSE00002056886,ENSE00001175672,ENSE00001175678,ENSE00001175684,ENSE00001175691,ENSE00001006057,ENSE00001006053,ENSE00003651115,ENSE00003561541,ENSE00003487906,ENSE00003598972,ENSE00001786211,ENSE00002316459,ENSE00002028083,ENSE00001789990,ENSE00003602905,ENSE00003459209,ENSE00002234840,ENSE00001678790,ENSE00002045268,ENSE00003592507,ENSE00003613792,ENSE00001605148,ENSE00002310772,ENSE00002225747,ENSE00002211631,ENSE00002063937,ENSE00002061881,ENSE00002270604,ENSE00002252787,ENSE00002269371,ENSE00001175712,ENSE00001603831,ENSE00001942085,ENSE00002020252,ENSE00002083738,ENSE00002034086
## ENSG00000096060                                                                                                                                                                                                                                                                                                                 ENSE00001193215,ENSE00001354861,ENSE00002324002,ENSE00000747301,ENSE00000747302,ENSE00001461801,ENSE00000747303,ENSE00000747304,ENSE00003659116,ENSE00003674427,ENSE00000747342,ENSE00001391213,ENSE00000747346,ENSE00001370159,ENSE00002312621,ENSE00002286243,ENSE00001354888,ENSE00001519071,ENSE00001424352,ENSE00002227265
## ENSG00000101347                                                                                                                                                                                                 ENSE00000800455,ENSE00001930417,ENSE00000661812,ENSE00000661813,ENSE00003692177,ENSE00003551249,ENSE00001842453,ENSE00000661815,ENSE00001622813,ENSE00001604475,ENSE00001711142,ENSE00001793017,ENSE00001667406,ENSE00001752622,ENSE00003598833,ENSE00003555165,ENSE00003535434,ENSE00003608826,ENSE00003546823,ENSE00003559540,ENSE00002218059,ENSE00003462531,ENSE00003673643,ENSE00003619308,ENSE00003541183,ENSE00001957470,ENSE00001898623
## ENSG00000211445                                                                                                                                                                                                                                                                 ENSE00002107942,ENSE00002109589,ENSE00002103911,ENSE00002122901,ENSE00002118129,ENSE00002124493,ENSE00003675685,ENSE00003665164,ENSE00002136348,ENSE00002125088,ENSE00002105542,ENSE00003651491,ENSE00003520309,ENSE00003538636,ENSE00002138234,ENSE00002126956,ENSE00002097854,ENSE00003670314,ENSE00003461760,ENSE00002108594,ENSE00002108389,ENSE00002112476,ENSE00001504048
## ENSG00000163884                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001241585,ENSE00001079374,ENSE00002029422,ENSE00002086065,ENSE00001241575
## ENSG00000189221                                                                                                                                                                 ENSE00001415055,ENSE00002218265,ENSE00001834008,ENSE00001836223,ENSE00003501661,ENSE00003656032,ENSE00003691177,ENSE00003550793,ENSE00003487993,ENSE00003504764,ENSE00003590154,ENSE00003495252,ENSE00003524833,ENSE00003526715,ENSE00003672501,ENSE00001811266,ENSE00001383783,ENSE00001382812,ENSE00001377890,ENSE00001373498,ENSE00001369566,ENSE00001677475,ENSE00001816354,ENSE00001614970,ENSE00003583432,ENSE00003609031,ENSE00001875207,ENSE00002311052,ENSE00001389556
## ENSG00000157514                                 ENSE00001457703,ENSE00001457656,ENSE00001927655,ENSE00002041041,ENSE00002038888,ENSE00001837666,ENSE00002071248,ENSE00002029666,ENSE00003665736,ENSE00003656931,ENSE00001927337,ENSE00002058631,ENSE00002076150,ENSE00002024886,ENSE00001457675,ENSE00002294110,ENSE00002065188,ENSE00001457644,ENSE00002035416,ENSE00001457652,ENSE00002038214,ENSE00001457647,ENSE00002071593,ENSE00002026941,ENSE00001929174,ENSE00002056274,ENSE00002082126,ENSE00002049455,ENSE00002037364,ENSE00002051687,ENSE00002035177,ENSE00001870013,ENSE00002065447,ENSE00002038769,ENSE00002054083,ENSE00002030770,ENSE00002038361
## ENSG00000162692                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001167786,ENSE00001811760,ENSE00001825682,ENSE00001451872,ENSE00001451878,ENSE00001067757,ENSE00001067755,ENSE00001067758,ENSE00001067756,ENSE00001067753,ENSE00001067759,ENSE00001067751,ENSE00003565397,ENSE00001892498,ENSE00001451871,ENSE00001167778
## ENSG00000112936                                                                                                 ENSE00001880051,ENSE00001270622,ENSE00000821201,ENSE00002078394,ENSE00000742287,ENSE00003553702,ENSE00003475879,ENSE00001958007,ENSE00003642523,ENSE00003502339,ENSE00002062072,ENSE00000742293,ENSE00000742294,ENSE00000742295,ENSE00001008711,ENSE00000971156,ENSE00000971157,ENSE00000971158,ENSE00001869809,ENSE00000742300,ENSE00001874313,ENSE00003653847,ENSE00003602374,ENSE00001935044,ENSE00001850992,ENSE00003554311,ENSE00003588324,ENSE00002045506,ENSE00003589391,ENSE00003615437,ENSE00001896230,ENSE00001832121,ENSE00001143083
##                     logFC   logCPM        F        PValue           FDR
## ENSG00000109906 -7.141494 4.143740 1719.009  0.000000e+00  0.000000e+00
## ENSG00000152583 -4.549790 5.528678 1693.943  0.000000e+00  0.000000e+00
## ENSG00000096060 -3.992959 6.895112 1560.869  0.000000e+00  0.000000e+00
## ENSG00000101347 -3.752212 9.206504 1575.112  0.000000e+00  0.000000e+00
## ENSG00000211445 -3.713684 9.158160 1542.044  0.000000e+00  0.000000e+00
## ENSG00000163884 -4.472519 4.685058 1348.292 3.819016e-294 4.080109e-290
## ENSG00000189221 -3.332051 6.766776 1184.595 8.385137e-259 7.678629e-255
## ENSG00000157514 -3.180372 6.675013 1062.798 1.743036e-232 1.396651e-228
## ENSG00000162692  3.665549 4.599336 1034.572 2.202493e-226 1.568713e-222
## ENSG00000112936 -3.182516 6.515306 1033.852 3.152321e-226 2.020701e-222</code></pre>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co"># If you performed likelihood ratio tests</span>
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html">glmFit</a></span>(<span class="no">y</span>,<span class="no">design</span>)
<span class="no">lrt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html">glmLRT</a></span>(<span class="no">fit</span>)
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span>(<span class="no">lrt</span>)</pre></body></html></div>
<pre><code>## Coefficient:  dexuntrt 
##                                                                                                              seqnames
## ENSG00000109906             11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11
## ENSG00000152583                         4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4
## ENSG00000165995 10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10
## ENSG00000163884                                                                                             3,3,3,3,3
## ENSG00000189221                                             X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X
## ENSG00000127954                                                                                   7,7,7,7,7,7,7,7,7,7
## ENSG00000162692                                                                       1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
## ENSG00000101347                      20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20
## ENSG00000171819                                                                                       1,1,1,1,1,1,1,1
## ENSG00000120129                                                                                               5,5,5,5
##                                                                                                                                                                                                                                                                                                                                                                          start
## ENSG00000109906                                                    113930315,113930447,113930459,113930979,113931229,113932093,113933933,113933933,113933933,113933933,113935134,113992540,114027059,114027059,114051488,114057674,114057674,114057674,114107929,114112889,114112889,114117512,114117920,114117920,114117920,114117920,114121048,114121048,114121048,114121048
## ENSG00000152583 88394487,88394500,88394570,88400582,88401524,88403576,88411425,88411912,88412770,88414734,88414734,88415521,88415521,88415528,88415569,88415592,88415685,88416133,88416133,88416133,88416165,88418283,88420673,88420673,88425974,88449278,88449707,88449707,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88451697,88452130,88452130
## ENSG00000165995                                              18429606,18429666,18429676,18429721,18429887,18430109,18439812,18439812,18439812,18470736,18549698,18550190,18629646,18629666,18689742,18690853,18690853,18787284,18787284,18789741,18795400,18803165,18803440,18803909,18807265,18807839,18816517,18823005,18825030,18827109,18828159,18828159,18828159,18828159
## ENSG00000163884                                                                                                                                                                                                                                                                                                              126061478,126070684,126071518,126072855,126076080
## ENSG00000189221                                                                                           43515467,43515548,43515550,43517110,43542761,43542761,43552538,43552538,43571119,43571119,43571119,43571952,43571952,43587420,43587420,43590488,43590488,43590941,43591946,43595474,43599928,43601197,43602951,43603041,43603356,43603356,43603614,43603614,43603614
## ENSG00000127954                                                                                                                                                                                                                                                                      87905744,87907343,87910230,87910798,87911956,87913129,87920231,87936107,87936107,87936107
## ENSG00000162692                                                                                                                                                                                                101185298,101185311,101185316,101185320,101186032,101186032,101188576,101190180,101194663,101196754,101197974,101200058,101200731,101203679,101203679,101203679
## ENSG00000101347                                                                                                             35518632,35525971,35526225,35526843,35532560,35532560,35533767,35533767,35539621,35540751,35540864,35545125,35545352,35547767,35555585,35555585,35559163,35559163,35563432,35563432,35563744,35569442,35569442,35575141,35575141,35579839,35579839
## ENSG00000171819                                                                                                                                                                                                                                                                                        11249398,11252327,11253637,11253684,11254518,11254583,11254911,11254911
## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172195093,172196578,172197164,172197589
##                                                                                                                                                                                                                                                                                                                                                                            end
## ENSG00000109906                                                    113930604,113930604,113930604,113931016,113931306,113932283,113934368,113934466,113935290,113935290,113935290,113992636,114027156,114027156,114051664,114057760,114057760,114060486,114108062,114113059,114113059,114118087,114118018,114118066,114118087,114118087,114121198,114121279,114121374,114121398
## ENSG00000152583 88394955,88394955,88394955,88400730,88401672,88403712,88411545,88412030,88412842,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88416279,88416279,88416295,88416279,88418391,88420737,88420737,88426075,88449385,88449766,88449771,88450244,88450253,88450393,88450524,88450556,88450609,88450655,88450656,88451783,88452201,88452213
## ENSG00000165995                                              18429785,18429785,18429785,18429785,18430144,18430144,18439904,18439904,18440198,18471036,18550246,18550246,18629906,18629906,18690029,18690972,18690972,18787406,18787722,18789877,18795476,18803298,18803459,18803970,18807345,18807897,18816626,18823156,18825125,18827294,18828653,18830038,18830056,18830798
## ENSG00000163884                                                                                                                                                                                                                                                                                                              126062738,126071790,126071790,126073128,126076285
## ENSG00000189221                                                                                           43515662,43515662,43515662,43517217,43542855,43542855,43552675,43552675,43571223,43571223,43571223,43572043,43572043,43587561,43587561,43590629,43590637,43591100,43592042,43595527,43599985,43601294,43603152,43603152,43603418,43603418,43603905,43604409,43606068
## ENSG00000127954                                                                                                                                                                                                                                                                      87908943,87908943,87910394,87912483,87912483,87913586,87920329,87936195,87936203,87936206
## ENSG00000162692                                                                                                                                                                                                101185480,101185480,101185480,101185480,101186121,101186307,101188896,101190446,101194938,101197074,101198240,101200324,101204595,101204169,101204594,101204601
## ENSG00000101347                                                                                                             35521469,35526362,35526362,35526947,35532652,35532652,35533797,35533906,35539736,35540955,35540955,35545233,35545452,35547922,35555655,35555655,35559278,35559278,35563592,35563592,35563806,35569514,35569514,35575207,35575207,35580111,35580246
## ENSG00000171819                                                                                                                                                                                                                                                                                        11250012,11252427,11253831,11253831,11254716,11254716,11255235,11256038
## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172196135,172196797,172197309,172198198
##                                                                                                                                                              width
## ENSG00000109906                                 290,158,146,38,78,191,436,534,1358,1358,157,97,98,98,177,87,87,2813,134,171,171,576,99,147,168,168,151,232,327,351
## ENSG00000152583 469,456,386,149,149,137,121,119,73,1017,1017,230,230,223,182,159,66,147,147,163,115,109,65,65,102,108,60,65,48,57,197,328,360,413,459,460,87,72,84
## ENSG00000165995                    180,120,110,65,258,36,93,93,387,301,549,57,261,241,288,120,120,123,439,137,77,134,20,62,81,59,110,152,96,186,495,1880,1898,2640
## ENSG00000163884                                                                                                                              1261,1107,273,274,206
## ENSG00000189221                                         196,115,113,108,95,95,138,138,105,105,105,92,92,142,142,142,150,160,97,54,58,98,202,112,63,63,292,796,2455
## ENSG00000127954                                                                                                            3200,1601,165,1686,528,458,99,89,97,100
## ENSG00000162692                                                                                    183,170,165,161,90,276,321,267,276,321,267,267,3865,491,916,923
## ENSG00000101347                                                  2838,392,138,105,93,93,31,140,116,205,92,109,101,156,71,71,116,116,161,161,63,73,73,67,67,273,408
## ENSG00000171819                                                                                                                   615,101,195,148,199,134,325,1128
## ENSG00000120129                                                                                                                                   1043,220,146,610
##                                                                                        strand
## ENSG00000109906                   +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000152583 -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000165995           +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000163884                                                                     -,-,-,-,-
## ENSG00000189221                     +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000127954                                                           -,-,-,-,-,-,-,-,-,-
## ENSG00000162692                                               +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+
## ENSG00000101347                         -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-
## ENSG00000171819                                                               +,+,+,+,+,+,+,+
## ENSG00000120129                                                                       -,-,-,-
##                                                                                                                                                                                                                                                                                          exon_id
## ENSG00000109906                                                                360958,360959,360960,360961,360962,360963,360964,360965,360967,360966,360968,360969,360971,360970,360972,360974,360973,360975,360976,360977,360978,360979,360980,360981,360982,360983,360984,360985,360986,360987
## ENSG00000152583 172689,172690,172691,172692,172693,172694,172695,172696,172697,172699,172698,172700,172701,172702,172703,172704,172705,172707,172706,172708,172709,172710,172711,172712,172713,172714,172715,172716,172717,172718,172719,172720,172721,172722,172723,172724,172725,172726,172727
## ENSG00000165995                                    321791,321792,321793,321794,321795,321796,321797,321798,321799,321800,321801,321802,321803,321804,321805,321806,321807,321808,321809,321810,321811,321812,321813,321814,321815,321816,321817,321818,321819,321820,321821,321822,321823,321824
## ENSG00000163884                                                                                                                                                                                                                                               145255,145256,145257,145258,145259
## ENSG00000189221                                                                       653982,653983,653984,653985,653986,653987,653989,653988,653990,653991,653992,653993,653994,653995,653996,653997,653998,653999,654000,654001,654002,654003,654004,654005,654006,654007,654008,654009,654010
## ENSG00000127954                                                                                                                                                                                                            263537,263538,263539,263540,263541,263542,263543,263544,263545,263546
## ENSG00000162692                                                                                                                                                                                  14792,14793,14794,14795,14796,14797,14798,14799,14800,14801,14802,14803,14804,14805,14806,14807
## ENSG00000101347                                                                                     626560,626561,626562,626563,626565,626564,626566,626567,626568,626569,626570,626571,626572,626573,626575,626574,626576,626577,626578,626579,626580,626581,626582,626584,626583,626585,626586
## ENSG00000171819                                                                                                                                                                                                                                          2221,2222,2223,2224,2225,2226,2227,2228
## ENSG00000120129                                                                                                                                                                                                                                                      207637,207638,207639,207640
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       exon_name
## ENSG00000109906                                                                                                                                                 ENSE00001344383,ENSE00002217296,ENSE00002239785,ENSE00002278969,ENSE00001777728,ENSE00002277816,ENSE00002264739,ENSE00002267450,ENSE00003532715,ENSE00003522546,ENSE00002223922,ENSE00002239280,ENSE00003606532,ENSE00003503866,ENSE00002304588,ENSE00003667676,ENSE00003652618,ENSE00002237247,ENSE00002308595,ENSE00003574983,ENSE00003648066,ENSE00002216844,ENSE00002254046,ENSE00002202712,ENSE00003533854,ENSE00003675344,ENSE00002251269,ENSE00002289800,ENSE00002238115,ENSE00001513888
## ENSG00000152583 ENSE00001914146,ENSE00001135423,ENSE00002056886,ENSE00001175672,ENSE00001175678,ENSE00001175684,ENSE00001175691,ENSE00001006057,ENSE00001006053,ENSE00003651115,ENSE00003561541,ENSE00003487906,ENSE00003598972,ENSE00001786211,ENSE00002316459,ENSE00002028083,ENSE00001789990,ENSE00003602905,ENSE00003459209,ENSE00002234840,ENSE00001678790,ENSE00002045268,ENSE00003592507,ENSE00003613792,ENSE00001605148,ENSE00002310772,ENSE00002225747,ENSE00002211631,ENSE00002063937,ENSE00002061881,ENSE00002270604,ENSE00002252787,ENSE00002269371,ENSE00001175712,ENSE00001603831,ENSE00001942085,ENSE00002020252,ENSE00002083738,ENSE00002034086
## ENSG00000165995                                                                                 ENSE00001552836,ENSE00001323316,ENSE00001912501,ENSE00001877551,ENSE00001563168,ENSE00001473561,ENSE00003566244,ENSE00003685211,ENSE00001824310,ENSE00001935997,ENSE00001473551,ENSE00001696833,ENSE00001402115,ENSE00001811759,ENSE00001473523,ENSE00003471380,ENSE00003510985,ENSE00001005446,ENSE00001886774,ENSE00001005448,ENSE00001005464,ENSE00001005450,ENSE00001367687,ENSE00001385143,ENSE00001098801,ENSE00001098797,ENSE00001098795,ENSE00001098805,ENSE00001737387,ENSE00001098799,ENSE00001266433,ENSE00001846919,ENSE00001548772,ENSE00001552865
## ENSG00000163884                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001241585,ENSE00001079374,ENSE00002029422,ENSE00002086065,ENSE00001241575
## ENSG00000189221                                                                                                                                                                 ENSE00001415055,ENSE00002218265,ENSE00001834008,ENSE00001836223,ENSE00003501661,ENSE00003656032,ENSE00003691177,ENSE00003550793,ENSE00003487993,ENSE00003504764,ENSE00003590154,ENSE00003495252,ENSE00003524833,ENSE00003526715,ENSE00003672501,ENSE00001811266,ENSE00001383783,ENSE00001382812,ENSE00001377890,ENSE00001373498,ENSE00001369566,ENSE00001677475,ENSE00001816354,ENSE00001614970,ENSE00003583432,ENSE00003609031,ENSE00001875207,ENSE00002311052,ENSE00001389556
## ENSG00000127954                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001917465,ENSE00001277946,ENSE00001163797,ENSE00001753614,ENSE00001483677,ENSE00001120706,ENSE00001621319,ENSE00001673636,ENSE00001905312,ENSE00001871078
## ENSG00000162692                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001167786,ENSE00001811760,ENSE00001825682,ENSE00001451872,ENSE00001451878,ENSE00001067757,ENSE00001067755,ENSE00001067758,ENSE00001067756,ENSE00001067753,ENSE00001067759,ENSE00001067751,ENSE00003565397,ENSE00001892498,ENSE00001451871,ENSE00001167778
## ENSG00000101347                                                                                                                                                                                                 ENSE00000800455,ENSE00001930417,ENSE00000661812,ENSE00000661813,ENSE00003692177,ENSE00003551249,ENSE00001842453,ENSE00000661815,ENSE00001622813,ENSE00001604475,ENSE00001711142,ENSE00001793017,ENSE00001667406,ENSE00001752622,ENSE00003598833,ENSE00003555165,ENSE00003535434,ENSE00003608826,ENSE00003546823,ENSE00003559540,ENSE00002218059,ENSE00003462531,ENSE00003673643,ENSE00003619308,ENSE00003541183,ENSE00001957470,ENSE00001898623
## ENSG00000171819                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001471825,ENSE00001168153,ENSE00001168146,ENSE00001948794,ENSE00001168140,ENSE00001823259,ENSE00001841505,ENSE00001471787
## ENSG00000120129                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001876542,ENSE00000769464,ENSE00000769465,ENSE00000812856
##                     logFC   logCPM        LR        PValue           FDR
## ENSG00000109906 -7.141823 4.143740 1293.2297 3.345918e-283 2.144800e-278
## ENSG00000152583 -4.552568 5.528678  901.7878 4.010221e-198 1.285316e-193
## ENSG00000165995 -3.274517 4.504348  749.6857 4.696169e-165 1.003446e-160
## ENSG00000163884 -4.458891 4.685058  703.8821 4.280694e-155 6.860026e-151
## ENSG00000189221 -3.335100 6.766776  647.7867 6.767998e-143 8.676844e-139
## ENSG00000127954 -5.197020 3.652849  638.6801 6.471522e-141 6.913959e-137
## ENSG00000162692  3.677380 4.599336  634.3726 5.595308e-140 5.123863e-136
## ENSG00000101347 -3.751932 9.206504  631.1278 2.841479e-139 2.276806e-135
## ENSG00000171819 -5.668260 3.500283  615.0520 8.911131e-136 6.346904e-132
## ENSG00000120129 -2.932273 7.308257  603.5237 2.866583e-133 1.837537e-129</code></pre>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co"># lrt to DESeqResults</span>

<span class="no">tbledger</span> <span class="kw">&lt;-</span> <span class="no">lrt</span>$<span class="no">table</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>) <span class="kw">==</span> <span class="st">'PValue'</span>] <span class="kw">&lt;-</span> <span class="st">'pvalue'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>) <span class="kw">==</span> <span class="st">'logFC'</span>] <span class="kw">&lt;-</span> <span class="st">'log2FoldChange'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbledger</span>) <span class="kw">==</span> <span class="st">'logCPM'</span>] <span class="kw">&lt;-</span> <span class="st">'baseMean'</span>
<span class="co"># get from the logcpm to something more the baseMean for better </span>
<span class="no">tbledger</span>$<span class="no">baseMean</span> <span class="kw">&lt;-</span> (<span class="fl">2</span>^<span class="no">tbledger</span>$<span class="no">baseMean</span>) * <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">dge_airway</span>$<span class="no">samples</span>$<span class="no">lib.size</span>) / <span class="fl">1e6</span>
<span class="co"># use the constructor for DESeqResults </span>
<span class="no">edger_resu</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html">DESeqResults</a></span>(<span class="fu">DataFrame</span>(<span class="no">tbledger</span>))
<span class="no">edger_resu</span> <span class="kw">&lt;-</span> <span class="kw pkg">DESeq2</span><span class="kw ns">:::</span><span class="fu">pvalueAdjustment</span>(<span class="no">edger_resu</span>, <span class="kw">independentFiltering</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                                        <span class="kw">alpha</span> <span class="kw">=</span> <span class="fl">0.05</span>, <span class="kw">pAdjustMethod</span> <span class="kw">=</span> <span class="st">"BH"</span>)

<span class="co"># cfr with res_airway</span>
<span class="co"># plot(res_airway$pvalue,edger_resu$pvalue)</span>



<span class="co">### with limma-voom</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="no">dge_airway</span>)
<span class="no">design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(~<span class="fl">0</span>+<span class="no">dex</span>+<span class="no">cell</span>)
<span class="no">contr.matrix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html">makeContrasts</a></span>(<span class="no">dextrt</span> - <span class="no">dexuntrt</span>,<span class="kw">levels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">design</span>))
<span class="no">v</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html">voom</a></span>(<span class="no">x</span>, <span class="no">design</span>)
<span class="no">vfit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span>(<span class="no">v</span>, <span class="no">design</span>)
<span class="no">vfit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span>(<span class="no">vfit</span>, <span class="kw">contrasts</span><span class="kw">=</span><span class="no">contr.matrix</span>)
<span class="no">efit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span>(<span class="no">vfit</span>)

<span class="no">tbllimma</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span>(<span class="no">efit</span>,<span class="kw">number</span><span class="kw">=</span> <span class="fl">Inf</span>, <span class="kw">sort.by</span> <span class="kw">=</span> <span class="st">"none"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>) <span class="kw">==</span> <span class="st">'P.Value'</span>] <span class="kw">&lt;-</span> <span class="st">'pvalue'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>) <span class="kw">==</span> <span class="st">'logFC'</span>] <span class="kw">&lt;-</span> <span class="st">'log2FoldChange'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>) <span class="kw">==</span> <span class="st">'AveExpr'</span>] <span class="kw">&lt;-</span> <span class="st">'baseMean'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tbllimma</span>) <span class="kw">==</span> <span class="st">'adj.P.Val'</span>] <span class="kw">&lt;-</span> <span class="st">'padj'</span>
<span class="co"># get from the logcpm to something more the baseMean for better </span>
<span class="no">tbllimma</span>$<span class="no">baseMean</span> <span class="kw">&lt;-</span> (<span class="fl">2</span>^<span class="no">tbllimma</span>$<span class="no">baseMean</span>) * <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">dge_airway</span>$<span class="no">samples</span>$<span class="no">lib.size</span>) / <span class="fl">1e6</span>
<span class="co"># use the constructor for DESeqResults </span>
<span class="no">limma_resu</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html">DESeqResults</a></span>(<span class="fu">DataFrame</span>(<span class="no">tbllimma</span>))

<span class="co"># cfr with res_airway</span>
<span class="co"># plot(res_airway$pvalue,limma_resu$pvalue)</span></pre></body></html></div>
<p>After preparing the objects, you can launch the app with the following command:</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="../reference/ideal.html">ideal</a></span>(<span class="no">redo_dds_airway</span>,<span class="no">edger_resu</span>)
<span class="co"># or ...</span>
<span class="fu"><a href="../reference/ideal.html">ideal</a></span>(<span class="no">redo_dds_airway</span>,<span class="no">limma_resu</span>)</pre></body></html></div>
<p>Basically, you need a <code>data.frame</code> (e.g. <code>myresults_df</code>) with 3 columns to create the <code>DESeqResults</code> object to be passed to the app, with this names:</p>
<ul>
<li>
<code>baseMean</code> as an average expression value</li>
<li>
<code>log2FoldChange</code> as an indicator for the effect size</li>
<li>
<code>pvalue</code> as a measure of significance</li>
</ul>
<p>Then, you just need to call <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html">DESeqResults(DataFrame(myresults_df))</a></code>. <code>padj</code> can be computed either by <code>p.adjust</code> or using <code>DESeq2:::pvalueAdjustment</code> as in the example above.</p>
</div>
</div>
<div id="functions-exported-by-the-package-for-standalone-usage" class="section level1">
<h1 class="hasAnchor">
<a href="#functions-exported-by-the-package-for-standalone-usage" class="anchor"></a>Functions exported by the package for standalone usage</h1>
<p>The functions exported by the <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> package can be also used in a standalone scenario, provided the required objects are in the working environment. They are listed here for an overview, but please refer to the documentation for additional details. Where possible, for each function a code snippet will be provided for its typical usage.</p>
<div id="deseqresult2degenes-and-deseqresult2tbl" class="section level2">
<h2 class="hasAnchor">
<a href="#deseqresult2degenes-and-deseqresult2tbl" class="anchor"></a><code>deseqresult2DEgenes</code> and <code>deseqresult2tbl</code>
</h2>
<p><code>deseqresult2DEgenes</code> and <code>deseqresult2tbl</code> generate a tidy table with the results of DESeq2, sorted by the values in the <code>padj</code> column.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/summary.html">summary</a></span>(<span class="no">res_airway</span>)</pre></body></html></div>
<pre><code>## 
## out of 33469 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 2211, 6.6%
## LFC &lt; 0 (down)     : 1817, 5.4%
## outliers [1]       : 0, 0%
## low counts [2]     : 16676, 50%
## (mean count &lt; 7)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results</code></pre>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">res_airway</span></pre></body></html></div>
<pre><code>## log2 fold change (MLE): dex trt vs untrt 
## Wald test p-value: dex trt vs untrt 
## DataFrame with 64102 rows and 6 columns
##                  baseMean log2FoldChange     lfcSE      stat      pvalue
##                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
## ENSG00000000003  708.6022     -0.3812540  0.100654 -3.787752 0.000152016
## ENSG00000000005    0.0000             NA        NA        NA          NA
## ENSG00000000419  520.2979      0.2068126  0.112219  1.842943 0.065337292
## ENSG00000000457  237.1630      0.0379204  0.143445  0.264356 0.791505742
## ENSG00000000460   57.9326     -0.0881682  0.287142 -0.307054 0.758801924
## ...                   ...            ...       ...       ...         ...
## LRG_94                  0             NA        NA        NA          NA
## LRG_96                  0             NA        NA        NA          NA
## LRG_97                  0             NA        NA        NA          NA
## LRG_98                  0             NA        NA        NA          NA
## LRG_99                  0             NA        NA        NA          NA
##                       padj
##                  &lt;numeric&gt;
## ENSG00000000003 0.00119738
## ENSG00000000005         NA
## ENSG00000000419 0.18631502
## ENSG00000000457 0.90374373
## ENSG00000000460 0.88619241
## ...                    ...
## LRG_94                  NA
## LRG_96                  NA
## LRG_97                  NA
## LRG_98                  NA
## LRG_99                  NA</code></pre>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/deseqresult2tbl.html">deseqresult2tbl</a></span>(<span class="no">res_airway</span>))</pre></body></html></div>
<pre><code>##                id   baseMean log2FoldChange     lfcSE     stat        pvalue
## 1 ENSG00000152583   997.4398       4.574919 0.1840561 24.85611 2.220933e-136
## 2 ENSG00000165995   495.0929       3.291062 0.1331737 24.71255 7.839410e-135
## 3 ENSG00000120129  3409.0294       2.947810 0.1214377 24.27426 3.666925e-130
## 4 ENSG00000101347 12703.3871       3.766995 0.1554374 24.23481 9.562330e-130
## 5 ENSG00000189221  2341.7673       3.353580 0.1417823 23.65301 1.098955e-123
## 6 ENSG00000211445 12285.6151       3.730403 0.1658306 22.49526 4.618318e-112
##            padj
## 1 3.729613e-132
## 2 6.582360e-131
## 3 2.052622e-126
## 4 4.014505e-126
## 5 3.690950e-120
## 6 1.292590e-108</code></pre>
<p>In particular, <code>deseqresult2DEgenes</code> only includes genes detected as DE.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span>(<span class="no">res_airway</span>,<span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.05</span>))</pre></body></html></div>
<pre><code>##                id   baseMean log2FoldChange     lfcSE     stat        pvalue
## 1 ENSG00000152583   997.4398       4.574919 0.1840561 24.85611 2.220933e-136
## 2 ENSG00000165995   495.0929       3.291062 0.1331737 24.71255 7.839410e-135
## 3 ENSG00000120129  3409.0294       2.947810 0.1214377 24.27426 3.666925e-130
## 4 ENSG00000101347 12703.3871       3.766995 0.1554374 24.23481 9.562330e-130
## 5 ENSG00000189221  2341.7673       3.353580 0.1417823 23.65301 1.098955e-123
## 6 ENSG00000211445 12285.6151       3.730403 0.1658306 22.49526 4.618318e-112
##            padj
## 1 3.729613e-132
## 2 6.582360e-131
## 3 2.052622e-126
## 4 4.014505e-126
## 5 3.690950e-120
## 6 1.292590e-108</code></pre>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="co"># the output in the first lines is the same, but</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">res_airway</span>)</pre></body></html></div>
<pre><code>## [1] 64102     6</code></pre>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span>(<span class="no">res_airway</span>))</pre></body></html></div>
<pre><code>## [1] 4028    7</code></pre>
<p>This tables can be enhanced with clickable links to the ENSEMBL and NCBI gene databases by the following commands:</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">myde</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span>(<span class="no">res_airway</span>,<span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.05</span>)
<span class="no">myde</span>$<span class="no">symbol</span> <span class="kw">&lt;-</span> <span class="fu">mapIds</span>(<span class="no">org.Hs.eg.db</span>,<span class="kw">keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">myde</span>$<span class="no">id</span>),<span class="kw">column</span> <span class="kw">=</span> <span class="st">"SYMBOL"</span>,<span class="kw">keytype</span><span class="kw">=</span><span class="st">"ENSEMBL"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="no">myde_enhanced</span> <span class="kw">&lt;-</span> <span class="no">myde</span>
<span class="no">myde_enhanced</span>$<span class="no">id</span> <span class="kw">&lt;-</span> <span class="kw pkg">ideal</span><span class="kw ns">:::</span><span class="fu">createLinkENS</span>(<span class="no">myde_enhanced</span>$<span class="no">id</span>,<span class="kw">species</span> <span class="kw">=</span> <span class="st">"Homo_sapiens"</span>)
<span class="no">myde_enhanced</span>$<span class="no">symbol</span> <span class="kw">&lt;-</span> <span class="kw pkg">ideal</span><span class="kw ns">:::</span><span class="fu">createLinkGeneSymbol</span>(<span class="no">myde_enhanced</span>$<span class="no">symbol</span>)
<span class="kw pkg">DT</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="no">myde_enhanced</span>[<span class="fl">1</span>:<span class="fl">100</span>,], <span class="kw">escape</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<div id="htmlwidget-30e4409849e39179307f" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-30e4409849e39179307f">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100"],["<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000152583\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000152583<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000165995\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000165995<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000120129\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000120129<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000101347\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000101347<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000189221\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000189221<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000211445\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000211445<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157214\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000157214<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162614\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162614<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000125148\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000125148<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000154734\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000154734<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000139132\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000139132<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162493\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162493<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134243\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134243<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000179094\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000179094<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162692\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162692<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163884\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163884<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000178695\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000178695<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000198624\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000198624<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000107562\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000107562<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000148848\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000148848<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000146250\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000146250<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000106976\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000106976<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000154930\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000154930<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163394\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163394<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000253368\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000253368<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163661\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163661<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162616\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162616<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000148175\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000148175<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124766\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124766<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000197381\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000197381<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000172403\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000172403<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000116285\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000116285<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163083\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163083<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000166741\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000166741<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000070404\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000070404<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000105989\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000105989<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000122035\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000122035<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000119139\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000119139<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000136478\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000136478<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196569\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196569<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000108821\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000108821<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000221866\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000221866<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000100242\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000100242<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000145685\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000145685<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134686\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134686<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000172986\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000172986<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000102804\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000102804<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000060718\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000060718<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000099337\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000099337<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000126803\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000126803<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157150\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000157150<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000133142\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000133142<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000138316\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000138316<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000103196\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000103196<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000131389\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000131389<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000152268\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000152268<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000168309\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000168309<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163171\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163171<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124440\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124440<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000141401\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000141401<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000103742\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000103742<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000112837\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000112837<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000164647\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000164647<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000068383\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000068383<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000156675\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000156675<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000136999\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000136999<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000206538\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000206538<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000227051\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000227051<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134363\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134363<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196517\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196517<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124762\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124762<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000149218\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000149218<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000109906\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000109906<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000185432\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000185432<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163378\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163378<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000167641\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000167641<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196975\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196975<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000106484\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000106484<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000126016\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000126016<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000151726\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000151726<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000150907\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000150907<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000130513\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000130513<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000221869\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000221869<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000169715\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000169715<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000147027\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000147027<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000145390\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000145390<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000185950\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000185950<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000013293\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000013293<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000131831\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000131831<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000146122\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000146122<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000137672\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000137672<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000020577\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000020577<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000119138\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000119138<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000138735\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000138735<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163110\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163110<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000071282\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000071282<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000101825\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000101825<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162407\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162407<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163513\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163513<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000246430\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000246430<\/a>"],[997.439773207048,495.092906698546,3409.02937530956,12703.3870621821,2341.76725275591,12285.6151498691,3009.26322295794,5393.10168349223,3656.252782247,30315.1354711078,1223.46614081102,1099.62586594636,5510.95825597916,776.596665468835,508.170232439755,561.107166129071,2649.85014895882,2057.19173059348,25136.3075719089,1365.17398590702,313.689977800113,884.745882993046,670.046703530433,1216.6020649876,1040.17649034906,8565.70116426611,1467.63582703992,11193.718773887,1281.46865801595,1969.04835631005,8014.7816508389,3232.63543712134,418.921476555884,7432.9363516342,2128.11402480981,333.214688163837,943.340574231144,5029.85142077092,3701.52859767166,17178.4260430908,100990.694351515,847.493946021912,2236.2383212374,2018.16368824202,2737.98198594824,756.366829107728,5397.33684521952,363.762294004358,530.060485021926,449.17684817333,167.615505683076,3055.97384805426,799.568584106461,3082.06913512982,740.410803126942,1597.83892249109,159.526919764023,1357.57154103081,197.043622803841,627.527233035192,370.82982255279,2754.96812806571,1920.71012986285,1247.24185187194,1631.87785808313,435.712716722131,12754.9476812811,4615.74006753287,2906.3410530882,229.640742740425,8020.52881680403,2027.18573134681,385.071028862516,3885.6492404017,1805.28931002148,231.671616289766,4908.8181516426,677.213238740759,386.415162103212,1460.17800298032,678.868174958036,451.004765299293,1963.57429862035,554.194641254197,12593.3295910107,8442.05371962528,1974.88307917017,244.497326184804,1017.17597249954,1942.41443351691,655.449436784439,3105.00555637675,1683.2682077277,6850.26951693182,1314.18432532042,3659.86884296353,4767.14944686345,25853.6620173955,13216.6476019075,2162.53881470905],[4.57491873843491,3.29106176403315,2.9478098639195,3.76699543436408,3.35357994378782,3.73040266633607,1.97677251857873,2.03566491652844,2.21097855583723,2.3456036828563,2.22890298187202,1.89121701527402,2.19571163110442,3.19174990854463,-3.69266058082503,4.45912820866763,-2.52817456060374,2.91843572447791,-1.91166983641132,-1.81454314533375,-2.75624278406194,-1.7562564406907,1.89852992350102,-2.18850417776046,1.98905883196779,2.26026682055985,1.51911806699615,1.44787382405786,-2.4119901286651,2.32109081388685,1.74435752623225,2.42451299177805,3.44998558648886,2.18737318780403,1.8155243546487,-3.04184070272801,2.288870596122,1.92793686438767,1.46583077408219,2.03064228241387,-1.3049464420229,2.75315191919184,1.85904607960478,-1.23994363586989,1.38754886372375,-1.57290367426743,1.20138799368316,2.42734208791257,2.71827771300548,2.45871257645932,2.94532021155259,1.53909067526407,-2.50719276017938,2.63126675642951,-2.12491614560924,1.91295694919961,4.83584740187653,1.50562831455222,2.52988705621146,2.32686807604604,-2.19734247824709,-1.83356244471417,1.56739405864982,1.38112413393709,1.54725188157684,-2.55248079258346,1.36622032630791,-1.151117902147,-1.80754278766037,-2.22924544131829,-1.00797847291904,1.92643097844724,7.35262580898062,2.24936617078873,1.26185128385828,2.89652209582583,1.18507402021209,-1.8637650860243,-2.37428208838473,1.60249023356877,2.22784502832455,-1.86853337535912,1.49982909698427,2.22518024290245,1.16474952177727,2.14750844417035,2.12423823311707,-2.88290688674632,1.42847388936671,1.68143695109109,1.95430726838003,1.20330797579411,2.08291934981393,-1.17882788438342,1.60545647508075,1.60527035076755,-1.32376820049595,-1.0014599405466,1.45838180541525,2.0512069721771],[0.184056092094016,0.133173695928173,0.121437689778823,0.155437396880673,0.141782348968743,0.165830584142872,0.0899823243685651,0.0941823072413265,0.10564078059373,0.115808058684839,0.112839437391307,0.0957795930235963,0.112007681687192,0.163962920457615,0.190185898203412,0.234863687585881,0.13443012375092,0.161296341724212,0.105748095217685,0.100483648865865,0.154888457370009,0.0997092669086191,0.108680978814825,0.125426994390838,0.114476718104887,0.130784242562768,0.0879565365437062,0.0843415499900068,0.140805797626506,0.135754467110297,0.103292012983118,0.144277322345034,0.208810170664016,0.134467866594971,0.113376796611315,0.191370959052672,0.144873254515938,0.122239410953865,0.0931376320932493,0.132700384998169,0.0854759991473077,0.181860564601197,0.122956027076028,0.0820420233819913,0.091834240766912,0.104277445142647,0.0798459534985851,0.161719375784471,0.181277396346021,0.164437635025724,0.197284112665953,0.103653790190652,0.169059459494136,0.179650210652512,0.145233369288729,0.131071764516462,0.331701711174415,0.103308740078303,0.174271007852363,0.160804044218343,0.152138396320904,0.127342051010235,0.109388229962853,0.0970434464891289,0.109445841286016,0.182762156358851,0.0979619407698337,0.0827635615809139,0.130787597667298,0.161327083682904,0.0731177286642894,0.140256157842824,0.536388708511622,0.164369569871279,0.0924377332110804,0.212578531690494,0.0871650987197041,0.137305943233996,0.175174674129628,0.118381204117079,0.164891665536523,0.138568499803701,0.111243466177348,0.166765250747559,0.0873411821868367,0.161424020150622,0.159990380628694,0.218489977266558,0.1087276278804,0.128926345050477,0.149999816008704,0.0924720105542745,0.160522464318899,0.0909257082542852,0.123921626523685,0.12465344741089,0.103195285203347,0.0782627728552586,0.114687209749098,0.161456466228165],[24.8561114516005,24.7125510867265,24.274258422475,24.2348077744505,23.6530144138544,22.4952633774848,21.9684536096437,21.6140905458219,20.9292144890536,20.2542354089507,19.7528721642115,19.7455110798821,19.6032236185056,19.4662909128268,-19.416058791465,18.9860265522617,-18.806607403621,18.0936262613316,-18.0775817519559,-18.0580936880185,-17.7950173361055,-17.6137734750398,17.4688334997039,-17.4484303669189,17.3752258528704,17.2824093810466,17.2712356203485,17.1667917441571,-17.1299063626841,17.0977122395613,16.8876322171914,16.8045327732096,16.5221146820479,16.2668840013101,16.0131914899024,-15.8949963870473,15.7991245780303,15.7718108206142,15.7383298365864,15.3024596156363,-15.2668170602368,15.1388066193968,15.1196010786463,-15.113518472072,15.1092757138978,-15.0838340171814,15.0463228384447,15.0095934772069,14.9951277312967,14.9522496846582,14.9293329896244,14.8483781676791,-14.830242375561,14.6466110274651,-14.631046267231,14.5947295075848,14.5789039940579,14.5740652089168,14.5169703635082,14.4702086776287,-14.4430500871868,-14.3987192774742,14.328726766875,14.2320185844987,14.1371464040684,-13.9661341463476,13.9464399701707,-13.9085109456244,-13.8204449037932,-13.8181723144514,-13.7856918059782,13.7350901955126,13.7076446470746,13.6848090102703,13.6508246148449,13.6256567057442,13.5957400108376,-13.5738121899653,-13.55379766043,13.5366948285465,13.5109620069372,-13.484546473449,13.4824016953337,13.3431889013306,13.3356280807567,13.3035247304988,13.2772872017037,-13.1946871101972,13.1380948634144,13.0418414516659,13.0287311036877,13.0126723597931,12.9758744899153,-12.9647368936262,12.9554180341065,12.8778656676551,-12.8277972960437,-12.7961213743184,12.7161678151013,12.7043965478496],[2.2209328519317e-136,7.8394097956377e-135,3.66692460961555e-130,9.56233000573123e-130,1.09895484197084e-123,4.61831818403322e-112,5.76997456475883e-107,1.32384913484128e-103,2.9023972834987e-97,3.25971323970826e-91,7.57828359063729e-87,8.76733638137227e-87,1.45132059890544e-85,2.1208340519519e-84,5.64602181588782e-84,2.22544194072977e-80,6.66702728287544e-79,3.5775330782885e-73,4.78616718475531e-73,6.81352218173249e-73,7.72434768571471e-71,1.93129966917865e-69,2.47503405302977e-68,3.5382280496688e-68,1.2710488026287e-67,6.38252132216978e-67,7.74659023654771e-67,4.70729581433583e-66,8.87975152400637e-66,1.54345145378225e-65,5.54854607580999e-64,2.26086984828426e-63,2.54322372376398e-61,1.69570082981641e-59,1.03368774602184e-57,6.86324182619446e-57,3.15466953718e-56,4.86351232160208e-56,8.25950320243983e-56,7.36209124965006e-53,1.27233637702658e-52,8.98289683269555e-52,1.20268967382562e-51,1.31904430444092e-51,1.4067749635356e-51,2.0689734900155e-51,3.64964500980089e-51,6.35359057985157e-51,7.90105490455152e-51,1.50575176028388e-50,2.12379849989549e-50,7.12721238012237e-50,9.33951312492951e-50,1.41604595820289e-48,1.78028214753118e-48,3.0341372137649e-48,3.82609967986668e-48,4.1070594319248e-48,9.45998119695722e-48,1.86904264326675e-47,2.77294264901206e-47,5.2711248004355e-47,1.44750076939657e-46,5.79835717746779e-46,2.24192605766704e-45,2.50879642253542e-44,3.30705380291383e-44,5.62380907333168e-44,1.91880253469702e-43,1.98034227626564e-43,3.10773049723513e-43,6.25780915677731e-43,9.1376213400172e-43,1.25135510349183e-42,1.99606962289157e-42,2.81863330346939e-42,4.24445806880519e-42,5.72645001340065e-42,7.5234810118038e-42,9.49666080286536e-42,1.34749044114089e-41,1.92847671571333e-41,1.98537308478103e-41,1.29786596669394e-40,1.43639834361794e-40,2.20808507994037e-40,3.13549010487034e-40,9.41449518313524e-40,1.99178480164259e-39,7.07279009916007e-39,8.39928375241299e-39,1.0365279749941e-38,1.67671662138361e-38,1.93894555958544e-38,2.18940138537665e-38,5.99711477942039e-38,1.14575459559673e-37,1.72346569667874e-37,4.80871289178758e-37,5.58989544173854e-37],[3.7296125382489e-132,6.5823604349072e-131,2.05262216564246e-126,4.01450519465611e-126,3.69094973224327e-120,1.29259028774116e-108,1.38421689808564e-103,2.77892481517371e-100,5.41555084242152e-94,5.47403644344208e-88,1.15692833034156e-83,1.22691566543654e-83,1.87477129364762e-82,2.5439404453163e-81,6.32090962361361e-81,2.33574040691719e-77,6.58584642125455e-76,3.33763961020549e-70,4.23021608071558e-70,5.72097389989169e-70,6.17690336600986e-68,1.4741961520235e-66,1.80709768924039e-65,2.47572765158701e-65,8.53788901701751e-65,4.12237232935373e-64,4.81809221638317e-64,2.82320066464791e-63,5.14198852905652e-63,8.63972675445512e-63,3.00570110487346e-61,1.18646210506993e-60,1.29419260585359e-58,8.37526589267853e-57,4.95963380541278e-55,3.20151166631343e-54,1.43179366318551e-53,2.14928848464904e-53,3.55645736611723e-53,3.09078995888434e-50,5.21130360473351e-50,3.59166158360611e-49,4.69692271919851e-49,5.03425250101737e-49,5.24977154725628e-49,7.55310256909353e-49,1.30401039679971e-48,2.22283013765515e-48,2.70780438800273e-48,5.05721786208945e-48,6.99312709975392e-48,2.3016784134499e-47,2.95921592277248e-47,4.40364069927799e-46,5.43568692790748e-46,9.0986189697775e-46,1.12722266533335e-45,1.18913532828126e-45,2.69256719051699e-45,5.23113885139642e-45,7.63377473850155e-45,1.42770965764054e-44,3.85839371753597e-44,1.52143456376901e-43,5.79210219790809e-43,6.38336641267231e-42,8.28885888243761e-42,1.3888327318891e-41,4.66992042973437e-41,4.75084112076126e-41,7.35043918874219e-41,1.45954707180224e-40,2.10202842688916e-40,2.8397305747214e-40,4.46933295696242e-40,6.22806698225809e-40,9.25677718823967e-40,1.23287532147483e-39,1.59926350166103e-39,1.99346781078147e-39,2.79363049112087e-39,3.94937920572853e-39,4.01691207382264e-39,2.59465037841563e-38,2.83781616286777e-38,4.31167124970216e-38,6.05221670472272e-38,1.7965638364817e-37,3.7582069858409e-37,1.31970404594661e-36,1.54999090169529e-36,1.89200155261694e-36,3.02764540031129e-36,3.46390561511896e-36,3.87017025943474e-36,1.04905779677923e-35,1.98357287874802e-35,2.95328157595164e-35,8.15683995876655e-35,9.38711141531154e-35],["<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SPARCL1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SPARCL1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CACNB2[sym]\" target=\"_blank\" class=\"btn btn-primary\">CACNB2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DUSP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">DUSP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SAMHD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SAMHD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MAOA[sym]\" target=\"_blank\" class=\"btn btn-primary\">MAOA<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GPX3[sym]\" target=\"_blank\" class=\"btn btn-primary\">GPX3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STEAP2[sym]\" target=\"_blank\" class=\"btn btn-primary\">STEAP2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NEXN[sym]\" target=\"_blank\" class=\"btn btn-primary\">NEXN<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MT2A[sym]\" target=\"_blank\" class=\"btn btn-primary\">MT2A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAMTS1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAMTS1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FGD4[sym]\" target=\"_blank\" class=\"btn btn-primary\">FGD4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDPN[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDPN<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SORT1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SORT1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PER1[sym]\" target=\"_blank\" class=\"btn btn-primary\">PER1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=VCAM1[sym]\" target=\"_blank\" class=\"btn btn-primary\">VCAM1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KLF15[sym]\" target=\"_blank\" class=\"btn btn-primary\">KLF15<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KCTD12[sym]\" target=\"_blank\" class=\"btn btn-primary\">KCTD12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCDC69[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCDC69<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CXCL12[sym]\" target=\"_blank\" class=\"btn btn-primary\">CXCL12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAM12[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAM12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PRSS35[sym]\" target=\"_blank\" class=\"btn btn-primary\">PRSS35<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DNM1[sym]\" target=\"_blank\" class=\"btn btn-primary\">DNM1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ACSS1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ACSS1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCKAR[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCKAR<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TRNP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">TRNP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PTX3[sym]\" target=\"_blank\" class=\"btn btn-primary\">PTX3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DNAJB4[sym]\" target=\"_blank\" class=\"btn btn-primary\">DNAJB4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STOM[sym]\" target=\"_blank\" class=\"btn btn-primary\">STOM<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SOX4[sym]\" target=\"_blank\" class=\"btn btn-primary\">SOX4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADARB1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADARB1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SYNPO2[sym]\" target=\"_blank\" class=\"btn btn-primary\">SYNPO2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ERRFI1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ERRFI1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=INHBB[sym]\" target=\"_blank\" class=\"btn btn-primary\">INHBB<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NNMT[sym]\" target=\"_blank\" class=\"btn btn-primary\">NNMT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FSTL3[sym]\" target=\"_blank\" class=\"btn btn-primary\">FSTL3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=WNT2[sym]\" target=\"_blank\" class=\"btn btn-primary\">WNT2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RASL11A[sym]\" target=\"_blank\" class=\"btn btn-primary\">RASL11A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TJP2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TJP2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TEX2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TEX2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LAMA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">LAMA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=COL1A1[sym]\" target=\"_blank\" class=\"btn btn-primary\">COL1A1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PLXNA4[sym]\" target=\"_blank\" class=\"btn btn-primary\">PLXNA4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SUN2[sym]\" target=\"_blank\" class=\"btn btn-primary\">SUN2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LHFPL2[sym]\" target=\"_blank\" class=\"btn btn-primary\">LHFPL2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PHC2[sym]\" target=\"_blank\" class=\"btn btn-primary\">PHC2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GXYLT2[sym]\" target=\"_blank\" class=\"btn btn-primary\">GXYLT2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TSC22D1[sym]\" target=\"_blank\" class=\"btn btn-primary\">TSC22D1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=COL11A1[sym]\" target=\"_blank\" class=\"btn btn-primary\">COL11A1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KCNK6[sym]\" target=\"_blank\" class=\"btn btn-primary\">KCNK6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=HSPA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">HSPA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TIMP4[sym]\" target=\"_blank\" class=\"btn btn-primary\">TIMP4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TCEAL4[sym]\" target=\"_blank\" class=\"btn btn-primary\">TCEAL4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAMTS14[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAMTS14<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CRISPLD2[sym]\" target=\"_blank\" class=\"btn btn-primary\">CRISPLD2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC6A6[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC6A6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NA[sym]\" target=\"_blank\" class=\"btn btn-primary\">NA<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FAM107A[sym]\" target=\"_blank\" class=\"btn btn-primary\">FAM107A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CDC42EP3[sym]\" target=\"_blank\" class=\"btn btn-primary\">CDC42EP3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=HIF3A[sym]\" target=\"_blank\" class=\"btn btn-primary\">HIF3A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IMPA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">IMPA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IGDCC4[sym]\" target=\"_blank\" class=\"btn btn-primary\">IGDCC4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TBX18[sym]\" target=\"_blank\" class=\"btn btn-primary\">TBX18<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STEAP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">STEAP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=INPP5A[sym]\" target=\"_blank\" class=\"btn btn-primary\">INPP5A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RAB11FIP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">RAB11FIP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCN3[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCN3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=VGLL3[sym]\" target=\"_blank\" class=\"btn btn-primary\">VGLL3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=C14orf132[sym]\" target=\"_blank\" class=\"btn btn-primary\">C14orf132<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FST[sym]\" target=\"_blank\" class=\"btn btn-primary\">FST<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC6A9[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC6A9<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CDKN1A[sym]\" target=\"_blank\" class=\"btn btn-primary\">CDKN1A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ENDOD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ENDOD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ZBTB16[sym]\" target=\"_blank\" class=\"btn btn-primary\">ZBTB16<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=METTL7A[sym]\" target=\"_blank\" class=\"btn btn-primary\">METTL7A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=EOGT[sym]\" target=\"_blank\" class=\"btn btn-primary\">EOGT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PPP1R14A[sym]\" target=\"_blank\" class=\"btn btn-primary\">PPP1R14A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ANXA4[sym]\" target=\"_blank\" class=\"btn btn-primary\">ANXA4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MEST[sym]\" target=\"_blank\" class=\"btn btn-primary\">MEST<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=AMOT[sym]\" target=\"_blank\" class=\"btn btn-primary\">AMOT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ACSL1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ACSL1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FOXO1[sym]\" target=\"_blank\" class=\"btn btn-primary\">FOXO1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GDF15[sym]\" target=\"_blank\" class=\"btn btn-primary\">GDF15<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CEBPD[sym]\" target=\"_blank\" class=\"btn btn-primary\">CEBPD<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MT1E[sym]\" target=\"_blank\" class=\"btn btn-primary\">MT1E<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TMEM47[sym]\" target=\"_blank\" class=\"btn btn-primary\">TMEM47<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=USP53[sym]\" target=\"_blank\" class=\"btn btn-primary\">USP53<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IRS2[sym]\" target=\"_blank\" class=\"btn btn-primary\">IRS2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC7A14[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC7A14<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RAI2[sym]\" target=\"_blank\" class=\"btn btn-primary\">RAI2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DAAM2[sym]\" target=\"_blank\" class=\"btn btn-primary\">DAAM2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TRPC6[sym]\" target=\"_blank\" class=\"btn btn-primary\">TRPC6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SAMD4A[sym]\" target=\"_blank\" class=\"btn btn-primary\">SAMD4A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KLF9[sym]\" target=\"_blank\" class=\"btn btn-primary\">KLF9<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDE5A[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDE5A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDLIM5[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDLIM5<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LMCD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">LMCD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MXRA5[sym]\" target=\"_blank\" class=\"btn btn-primary\">MXRA5<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PLPP3[sym]\" target=\"_blank\" class=\"btn btn-primary\">PLPP3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TGFBR2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TGFBR2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LINC00968[sym]\" target=\"_blank\" class=\"btn btn-primary\">LINC00968<\/a>"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>id<\/th>\n      <th>baseMean<\/th>\n      <th>log2FoldChange<\/th>\n      <th>lfcSE<\/th>\n      <th>stat<\/th>\n      <th>pvalue<\/th>\n      <th>padj<\/th>\n      <th>symbol<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="ggplotcounts" class="section level2">
<h2 class="hasAnchor">
<a href="#ggplotcounts" class="anchor"></a><code>ggplotCounts</code>
</h2>
<p><code>ggplotCounts</code> extends the functionality of the <code>plotCounts</code> function of <em><a href="https://bioconductor.org/packages/3.12/DESeq2">DESeq2</a></em>, and plots the normalized counts of a single gene as a boxplot, with jittered points superimposed.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="../reference/ggplotCounts.html">ggplotCounts</a></span>(<span class="kw">dds</span> <span class="kw">=</span> <span class="no">dds_airway</span>,
             <span class="kw">gene</span> <span class="kw">=</span> <span class="st">"ENSG00000103196"</span>, <span class="co"># CRISPLD2 in the original publication</span>
             <span class="kw">intgroup</span> <span class="kw">=</span> <span class="st">"dex"</span>)</pre></body></html></div>
<p><img src="ideal-usersguide_files/figure-html/ggpc1-1.png" width="700"></p>
<p>If an <code>annotation_obj</code> is provided, their gene name can also be included in the title.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="../reference/ggplotCounts.html">ggplotCounts</a></span>(<span class="kw">dds</span> <span class="kw">=</span> <span class="no">dds_airway</span>,
             <span class="kw">gene</span> <span class="kw">=</span> <span class="st">"ENSG00000103196"</span>, <span class="co"># CRISPLD2 in the original publication</span>
             <span class="kw">intgroup</span> <span class="kw">=</span> <span class="st">"dex"</span>,
             <span class="kw">annotation_obj</span> <span class="kw">=</span> <span class="no">annotation_airway</span>)</pre></body></html></div>
<p><img src="ideal-usersguide_files/figure-html/ggpc2-1.png" width="700"></p>
<p>When used in the context of the app, it is possible to seamless search for genes also by their gene name, making exploration even more immediate.</p>
</div>
<div id="goseqtable" class="section level2">
<h2 class="hasAnchor">
<a href="#goseqtable" class="anchor"></a><code>goseqTable</code>
</h2>
<p><code>goseqTable</code> is a wrapper to extract the functional GO terms enriched in in a list of (DE) genes, based on the algorithm and the implementation in the <em><a href="https://bioconductor.org/packages/3.12/goseq">goseq</a></em> package.</p>
<p>Its counterpart, based on the <em><a href="https://bioconductor.org/packages/3.12/topGO">topGO</a></em> package, can be found in the <em><a href="https://bioconductor.org/packages/3.12/pcaExplorer">pcaExplorer</a></em> package.</p>
<p><em>Please note that the following code chunks are not run only for reducing compilation time. The reader is invited to try out the functionality by simply executing these commands.</em></p>
<!-- maybe eval=FALSE here? -->
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">res_subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span>(<span class="no">res_airway</span>)[<span class="fl">1</span>:<span class="fl">100</span>,] <span class="co"># taking only a subset of the DE genes</span>
<span class="no">myde</span> <span class="kw">&lt;-</span> <span class="no">res_subset</span>$<span class="no">id</span>
<span class="no">myassayed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">res_airway</span>)
<span class="no">mygo</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/goseqTable.html">goseqTable</a></span>(<span class="kw">de.genes</span> <span class="kw">=</span> <span class="no">myde</span>,
                   <span class="kw">assayed.genes</span> <span class="kw">=</span> <span class="no">myassayed</span>,
                   <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
                   <span class="kw">id</span> <span class="kw">=</span> <span class="st">"ensGene"</span>,
                   <span class="kw">testCats</span> <span class="kw">=</span> <span class="st">"GO:BP"</span>,
                   <span class="kw">addGeneToTerms</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">mygo</span>)</pre></body></html></div>
<p>As for the results, this table can be enhanced by adding the links for each category to the AmiGO database.</p>
<!-- maybe eval=FALSE here? -->
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">mygo_enhanced</span> <span class="kw">&lt;-</span> <span class="no">mygo</span>
<span class="co"># using the unexported function to link the GO term to the entry in the AmiGO db</span>
<span class="no">mygo_enhanced</span>$<span class="no">category</span> <span class="kw">&lt;-</span> <span class="kw pkg">ideal</span><span class="kw ns">:::</span><span class="fu">createLinkGO</span>(<span class="no">mygo_enhanced</span>$<span class="no">category</span>)
<span class="kw pkg">DT</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="no">mygo_enhanced</span>,<span class="kw">escape</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
</div>
<div id="plot_ma" class="section level2">
<h2 class="hasAnchor">
<a href="#plot_ma" class="anchor"></a><code>plot_ma</code>
</h2>
<p>The MA plot provided by <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> displays the gene-wise log2-fold changes (logFCs) versus the average expression value. As a main input parameter, a <code>DESeqResults</code> object is required. Control on the appearance of the plot can be applied by selecting the False Discovery Rate (<code>FDR</code>), the point transparency (<code>point_alpha</code>), adding horizontal lines at particular logFC values (<code>hlines</code>). The user can also decide to add rug plots in the margins (setting <code>add_rug</code> to <code>TRUE</code>).</p>
<p>To facilitate the inspection of a particular gene or gene set, <code>intgenes</code> can assume the value of a vector of genes (either the IDs or the gene symbols if <code>symbol</code> column is provided in <code>res_obj</code>. Labels can be added via <code>labels_intgenes</code>, while classical labels/title can be also edited as preferred (see <code>plot_ma</code> for all details).</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span>(<span class="no">res_airway</span>, <span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.05</span>, <span class="kw">hlines</span> <span class="kw">=</span> <span class="fl">1</span>,
        <span class="kw">title</span> <span class="kw">=</span><span class="st">"Adding horizontal lines"</span>, <span class="kw">add_rug</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span>(<span class="no">res_airway</span>, <span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.1</span>,
        <span class="kw">intgenes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ENSG00000103196"</span>,  <span class="co"># CRISPLD2</span>
                     <span class="st">"ENSG00000120129"</span>,  <span class="co"># DUSP1</span>
                     <span class="st">"ENSG00000163884"</span>,  <span class="co"># KLF15</span>
                     <span class="st">"ENSG00000179094"</span>), <span class="co"># PER1</span>
        <span class="kw">title</span> <span class="kw">=</span> <span class="st">"Providing a shortlist of genes"</span>,
        <span class="kw">add_rug</span> <span class="kw">=</span> <span class="fl">FALSE</span>
       )</pre></body></html></div>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">res_airway2</span> <span class="kw">&lt;-</span> <span class="no">res_airway</span>
<span class="no">res_airway2</span>$<span class="no">symbol</span> <span class="kw">&lt;-</span> <span class="fu">mapIds</span>(<span class="no">org.Hs.eg.db</span>,<span class="kw">keys</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">res_airway2</span>),<span class="kw">column</span> <span class="kw">=</span> <span class="st">"SYMBOL"</span>,<span class="kw">keytype</span><span class="kw">=</span><span class="st">"ENSEMBL"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span>(<span class="no">res_airway2</span>, <span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.05</span>,
        <span class="kw">intgenes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CRISPLD2"</span>,  <span class="co"># CRISPLD2</span>
                     <span class="st">"DUSP1"</span>,  <span class="co"># DUSP1</span>
                     <span class="st">"KLF15"</span>,  <span class="co"># KLF15</span>
                     <span class="st">"PER1"</span>), <span class="co"># PER1</span>
        <span class="kw">annotation_obj</span> <span class="kw">=</span> <span class="no">annotation_airway</span>,
        <span class="kw">hlines</span> <span class="kw">=</span> <span class="fl">2</span>,
        <span class="kw">add_rug</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
        <span class="kw">title</span> <span class="kw">=</span> <span class="st">"Putting gene names..."</span>
       )</pre></body></html></div>
<p><img src="ideal-usersguide_files/figure-html/plotma2-1.png" width="700"></p>
</div>
<div id="plot_volcano" class="section level2">
<h2 class="hasAnchor">
<a href="#plot_volcano" class="anchor"></a><code>plot_volcano</code>
</h2>
<p>The volcano plot gives a different flavor for the gene overview, displaying log2-fold changes and log p-values.</p>
<p>In a way similar to <code>plot_ma</code>, genes can be annotated with <code>intgenes</code>, and vertical lines can be added via <code>vlines</code>. <code>ylim_up</code> controls the y axis upper limit to visualize better the bulk of genes - keep in mind that very small p-values due to robust differences/large effect sizes can be “cut out”.</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span>(<span class="no">res_airway</span>)</pre></body></html></div>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span>(<span class="no">res_airway2</span>, <span class="kw">FDR</span> <span class="kw">=</span> <span class="fl">0.05</span>,
        <span class="kw">intgenes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CRISPLD2"</span>,  <span class="co"># CRISPLD2</span>
                     <span class="st">"DUSP1"</span>,  <span class="co"># DUSP1</span>
                     <span class="st">"KLF15"</span>,  <span class="co"># KLF15</span>
                     <span class="st">"PER1"</span>), <span class="co"># PER1</span>
        <span class="kw">title</span> <span class="kw">=</span> <span class="st">"Selecting the handful of genes - and displaying the full range for the -log10(pvalue)"</span>,
        <span class="kw">ylim_up</span> <span class="kw">=</span> -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">res_airway2</span>$<span class="no">pvalue</span>, <span class="kw">na.rm</span> <span class="kw">=</span><span class="fl">TRUE</span>)))</pre></body></html></div>
<p><img src="ideal-usersguide_files/figure-html/plotvolcano2-1.png" width="700"></p>
</div>
<div id="sepguesser" class="section level2">
<h2 class="hasAnchor">
<a href="#sepguesser" class="anchor"></a><code>sepguesser</code>
</h2>
<p><code>sepguesser</code> makes an educated guess on the separator character for the input text file (<code>file</code>). The separator list can be provided as a vector in <code>sep_list</code> (defaults to comma, tab, semicolon, and whitespace - which ideally could cover most of the cases). The heuristics is based on the number of occurrencies of each separator in each line.</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/design_commas.txt"</span>,<span class="kw">package</span> <span class="kw">=</span> <span class="st">"ideal"</span>))</pre></body></html></div>
<pre><code>## [1] ","</code></pre>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/design_semicolons.txt"</span>,<span class="kw">package</span> <span class="kw">=</span> <span class="st">"ideal"</span>))</pre></body></html></div>
<pre><code>## [1] ";"</code></pre>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/design_spaces.txt"</span>,<span class="kw">package</span> <span class="kw">=</span> <span class="st">"ideal"</span>))</pre></body></html></div>
<pre><code>## [1] " "</code></pre>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="no">anyfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/design_tabs.txt"</span>,<span class="kw">package</span> <span class="kw">=</span> <span class="st">"ideal"</span>) <span class="co"># we know it is going to be TAB</span>
<span class="no">guessed_sep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span>(<span class="no">anyfile</span>)
<span class="no">guessed_sep</span></pre></body></html></div>
<pre><code>## [1] "\t"</code></pre>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="co"># to be used for reading in the same file, without having to specify the sep</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="no">anyfile</span>, <span class="kw">header</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">as.is</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
           <span class="kw">sep</span> <span class="kw">=</span> <span class="no">guessed_sep</span>,
           <span class="kw">quote</span> <span class="kw">=</span> <span class="st">""</span>, <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">check.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>##            SampleName    cell   dex albut        Run avgLength Experiment
## SRR1039508 GSM1275862  N61311 untrt untrt SRR1039508       126  SRX384345
## SRR1039509 GSM1275863  N61311   trt untrt SRR1039509       126  SRX384346
## SRR1039512 GSM1275866 N052611 untrt untrt SRR1039512       126  SRX384349
## SRR1039513 GSM1275867 N052611   trt untrt SRR1039513        87  SRX384350
## SRR1039516 GSM1275870 N080611 untrt untrt SRR1039516       120  SRX384353
## SRR1039517 GSM1275871 N080611   trt untrt SRR1039517       126  SRX384354
## SRR1039520 GSM1275874 N061011 untrt untrt SRR1039520       101  SRX384357
## SRR1039521 GSM1275875 N061011   trt untrt SRR1039521        98  SRX384358
##               Sample    BioSample sizeFactor
## SRR1039508 SRS508568 SAMN02422669  1.0236476
## SRR1039509 SRS508567 SAMN02422675  0.8961667
## SRR1039512 SRS508571 SAMN02422678  1.1794861
## SRR1039513 SRS508572 SAMN02422670  0.6700538
## SRR1039516 SRS508575 SAMN02422682  1.1776714
## SRR1039517 SRS508576 SAMN02422673  1.3990365
## SRR1039520 SRS508579 SAMN02422683  0.9207787
## SRR1039521 SRS508580 SAMN02422677  0.9445141</code></pre>
</div>
</div>
<div id="creating-and-sharing-output-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-and-sharing-output-objects" class="anchor"></a>Creating and sharing output objects</h1>
<p>While running the app, the user can</p>
<ul>
<li>generate and save graphics</li>
<li>create and export tables</li>
<li>generate, preview, download/export an HTML report</li>
<li>save the values of the <code>reactiveValues</code> in an environment, or in binary format</li>
</ul>
<p>This functionality to retrieve and share the output is provided by action buttons that are placed close to each element of interest.</p>
</div>
<div id="enhancing-ideal" class="section level1">
<h1 class="hasAnchor">
<a href="#enhancing-ideal" class="anchor"></a>Enhancing <code>ideal</code>
</h1>
<p>The <code>annotation_obj</code> is a quick helper to make your data and results easier to read. For creating it, you can exploit the corresponding <code>org.XX.eg.db</code> packages, available in Bioconductor.</p>
<p>Currently available are the following (in alphabetical order):</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Ag.eg.db">org.Ag.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.At.tair.db">org.At.tair.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Bt.eg.db">org.Bt.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Ce.eg.db">org.Ce.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Cf.eg.db">org.Cf.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Dm.eg.db">org.Dm.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Dr.eg.db">org.Dr.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.EcK12.eg.db">org.EcK12.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.EcSakai.eg.db">org.EcSakai.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Gg.eg.db">org.Gg.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Hs.eg.db">org.Hs.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Mm.eg.db">org.Mm.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Mmu.eg.db">org.Mmu.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Pf.plasmo.db">org.Pf.plasmo.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Pt.eg.db">org.Pt.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Rn.eg.db">org.Rn.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Sc.sgd.db">org.Sc.sgd.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Sco.eg.db">org.Sco.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Ss.eg.db">org.Ss.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Tgondii.eg.db">org.Tgondii.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/org.Xl.eg.db">org.Xl.eg.db</a></em></li>
</ul>
<p>They can all be easily installed with <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install("orgdb_packagename")</a></code>.</p>
<p>In a similar way, for using at best the <code>goseq</code> package, the gene length information can also be computed if the corresponding <code>TxDb</code> packages are installed. Currently, following <code>TxDb</code> packages are available:</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Athaliana.BioMart.plantsmart22">TxDb.Athaliana.BioMart.plantsmart22</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Athaliana.BioMart.plantsmart25">TxDb.Athaliana.BioMart.plantsmart25</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Athaliana.BioMart.plantsmart28">TxDb.Athaliana.BioMart.plantsmart28</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Btaurus.UCSC.bosTau8.refGene">TxDb.Btaurus.UCSC.bosTau8.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Celegans.UCSC.ce11.refGene">TxDb.Celegans.UCSC.ce11.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Celegans.UCSC.ce6.ensGene">TxDb.Celegans.UCSC.ce6.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Cfamiliaris.UCSC.canFam3.refGene">TxDb.Cfamiliaris.UCSC.canFam3.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Dmelanogaster.UCSC.dm3.ensGene">TxDb.Dmelanogaster.UCSC.dm3.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Dmelanogaster.UCSC.dm6.ensGene">TxDb.Dmelanogaster.UCSC.dm6.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Drerio.UCSC.danRer10.refGene">TxDb.Drerio.UCSC.danRer10.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Ggallus.UCSC.galGal4.refGene">TxDb.Ggallus.UCSC.galGal4.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Hsapiens.UCSC.hg18.knownGene">TxDb.Hsapiens.UCSC.hg18.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Hsapiens.UCSC.hg19.knownGene">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Hsapiens.UCSC.hg19.lincRNAsTranscripts">TxDb.Hsapiens.UCSC.hg19.lincRNAsTranscripts</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Hsapiens.UCSC.hg38.knownGene">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Mmulatta.UCSC.rheMac3.refGene">TxDb.Mmulatta.UCSC.rheMac3.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Mmulatta.UCSC.rheMac8.refGene">TxDb.Mmulatta.UCSC.rheMac8.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Mmusculus.UCSC.mm10.ensGene">TxDb.Mmusculus.UCSC.mm10.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Mmusculus.UCSC.mm10.knownGene">TxDb.Mmusculus.UCSC.mm10.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Mmusculus.UCSC.mm9.knownGene">TxDb.Mmusculus.UCSC.mm9.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Ptroglodytes.UCSC.panTro4.refGene">TxDb.Ptroglodytes.UCSC.panTro4.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Rnorvegicus.UCSC.rn4.ensGene">TxDb.Rnorvegicus.UCSC.rn4.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Rnorvegicus.UCSC.rn5.refGene">TxDb.Rnorvegicus.UCSC.rn5.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Rnorvegicus.UCSC.rn6.refGene">TxDb.Rnorvegicus.UCSC.rn6.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Scerevisiae.UCSC.sacCer2.sgdGene">TxDb.Scerevisiae.UCSC.sacCer2.sgdGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Scerevisiae.UCSC.sacCer3.sgdGene">TxDb.Scerevisiae.UCSC.sacCer3.sgdGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.12/TxDb.Sscrofa.UCSC.susScr3.refGene">TxDb.Sscrofa.UCSC.susScr3.refGene</a></em></li>
</ul>
<p>As for the <code>org.XX.eg.db</code> above, these can all be quickly installed with <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install("txdb_packagename")</a></code>.</p>
</div>
<div id="further-development" class="section level1">
<h1 class="hasAnchor">
<a href="#further-development" class="anchor"></a>Further development</h1>
<p>Additional functionality for the <em><a href="https://bioconductor.org/packages/3.12/ideal">ideal</a></em> will be added in the future, as it is tightly related to a topic under current development research.</p>
<p>Improvements, suggestions, bugs, issues and feedback of any type can be sent to <a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>.</p>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Mojave 10.14.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] edgeR_3.31.0                limma_3.45.0               
##  [3] DEFormats_1.17.0            org.Hs.eg.db_3.11.1        
##  [5] DESeq2_1.29.0               airway_1.9.0               
##  [7] SummarizedExperiment_1.19.2 DelayedArray_0.15.1        
##  [9] matrixStats_0.56.0          GenomicRanges_1.41.1       
## [11] GenomeInfoDb_1.25.0         ideal_1.12.1               
## [13] bigmemory_4.5.36            topGO_2.41.0               
## [15] SparseM_1.78                GO.db_3.11.1               
## [17] AnnotationDbi_1.51.0        IRanges_2.23.4             
## [19] S4Vectors_0.27.5            Biobase_2.49.0             
## [21] graph_1.67.0                BiocGenerics_0.35.2        
## [23] BiocStyle_2.17.0           
## 
## loaded via a namespace (and not attached):
##   [1] shinydashboard_0.7.1     tidyselect_1.0.0         heatmaply_1.1.0         
##   [4] RSQLite_2.2.0            htmlwidgets_1.5.1        grid_4.0.0              
##   [7] TSP_1.1-10               BiocParallel_1.23.0      IHW_1.17.0              
##  [10] munsell_0.5.0            codetools_0.2-16         rentrez_1.2.2           
##  [13] DT_0.13                  withr_2.2.0              colorspace_1.4-1        
##  [16] Category_2.55.0          pcaExplorer_2.15.0       knitr_1.28              
##  [19] rstudioapi_0.11          NMF_0.22.0               labeling_0.3            
##  [22] slam_0.1-47              GenomeInfoDbData_1.2.3   lpsymphony_1.17.0       
##  [25] farver_2.0.3             bit64_0.9-7              pheatmap_1.0.12         
##  [28] rprojroot_1.3-2          vctrs_0.3.0              xfun_0.14               
##  [31] BiocFileCache_1.13.0     gclus_1.3.2              R6_2.4.1                
##  [34] doParallel_1.0.15        seriation_1.2-8          locfit_1.5-9.4          
##  [37] bitops_1.0-6             shinyAce_0.4.1           assertthat_0.2.1        
##  [40] d3heatmap_0.6.1.2        promises_1.1.0           scales_1.1.1            
##  [43] gtable_0.3.0             rlang_0.4.6              genefilter_1.71.0       
##  [46] splines_4.0.0            rtracklayer_1.49.1       lazyeval_0.2.2          
##  [49] checkmate_2.0.0          shinyBS_0.61             BiocManager_1.30.10     
##  [52] yaml_2.2.1               reshape2_1.4.4           GenomicFeatures_1.41.0  
##  [55] threejs_0.3.3            crosstalk_1.1.0.1        backports_1.1.6         
##  [58] httpuv_1.5.2             RBGL_1.65.0              tools_4.0.0             
##  [61] bookdown_0.18            gridBase_0.4-7           ggplot2_3.3.0           
##  [64] ellipsis_0.3.1           gplots_3.0.3             RColorBrewer_1.1-2      
##  [67] Rcpp_1.0.4.6             plyr_1.8.6               base64enc_0.1-3         
##  [70] progress_1.2.2           zlibbioc_1.35.0          purrr_0.3.4             
##  [73] RCurl_1.98-1.2           BiasedUrn_1.07           prettyunits_1.1.1       
##  [76] openssl_1.4.1            viridis_0.5.1            ggrepel_0.8.2           
##  [79] cluster_2.1.0            fs_1.4.1                 magrittr_1.5            
##  [82] data.table_1.12.8        goseq_1.41.0             hms_0.5.3               
##  [85] mime_0.9                 evaluate_0.14            xtable_1.8-4            
##  [88] XML_3.99-0.3             gridExtra_2.3            compiler_4.0.0          
##  [91] biomaRt_2.45.0           tibble_3.0.1             KernSmooth_2.23-17      
##  [94] crayon_1.3.4             htmltools_0.4.0          GOstats_2.55.0          
##  [97] mgcv_1.8-31              later_1.0.0              tidyr_1.0.3             
## [100] geneplotter_1.67.0       DBI_1.1.0                geneLenDataBase_1.25.0  
## [103] dbplyr_1.4.3             MASS_7.3-51.6            rappdirs_0.3.1          
## [106] Matrix_1.2-18            gdata_2.18.0             igraph_1.2.5            
## [109] pkgconfig_2.0.3          bigmemory.sri_0.1.3      pkgdown_1.5.1           
## [112] GenomicAlignments_1.25.0 registry_0.5-1           plotly_4.9.2.1          
## [115] foreach_1.5.0            annotate_1.67.0          rngtools_1.5            
## [118] pkgmaker_0.31.1          webshot_0.5.2            XVector_0.29.0          
## [121] AnnotationForge_1.31.1   bibtex_0.4.2.2           stringr_1.4.0           
## [124] digest_0.6.25            Biostrings_2.57.0        rmarkdown_2.1           
## [127] rintrojs_0.2.2           dendextend_1.13.4        GSEABase_1.51.0         
## [130] curl_4.3                 shiny_1.4.0.2            Rsamtools_2.5.0         
## [133] gtools_3.8.2             lifecycle_0.2.0          nlme_3.1-147            
## [136] jsonlite_1.6.1           desc_1.2.0               viridisLite_0.3.0       
## [139] askpass_1.1              pillar_1.4.4             lattice_0.20-41         
## [142] fastmap_1.0.1            httr_1.4.1               survival_3.1-12         
## [145] glue_1.4.1               fdrtool_1.2.15           UpSetR_1.4.0            
## [148] png_0.1-7                iterators_1.0.12         bit_1.1-15.2            
## [151] Rgraphviz_2.33.0         stringi_1.4.6            blob_1.2.1              
## [154] caTools_1.18.0           memoise_1.1.0            dplyr_0.8.5</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://federicomarini.github.io">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
