<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ideal User's Guide • ideal</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ideal User's Guide">
<meta property="og:description" content="ideal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ideal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.27.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ideal-usersguide.html">ideal User's Guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/federicomarini/ideal" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="ideal-usersguide_files/htmltools-fill-0.5.8/fill.css" rel="stylesheet">
<script src="ideal-usersguide_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="ideal-usersguide_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="ideal-usersguide_files/datatables-binding-0.32/datatables.js"></script><link href="ideal-usersguide_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="ideal-usersguide_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="ideal-usersguide_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="ideal-usersguide_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="ideal-usersguide_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ideal User’s Guide</h1>
                        <h4 data-toc-skip class="author">Federico
Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), MainzCenter for Thrombosis and Hemostasis (CTH),
Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                              <h4 data-toc-skip class="author">Harald
Binder</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), Mainz<br><h4 data-toc-skip class="date">27 March 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/federicomarini/ideal/blob/HEAD/vignettes/ideal-usersguide.Rmd" class="external-link"><code>vignettes/ideal-usersguide.Rmd</code></a></small>
      <div class="hidden name"><code>ideal-usersguide.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>In the scope of differential expression analysis, we provide
      the R/Bioconductor package <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>,
      which serves as a web application to allow simultaneously for
      interactive and reproducible analysis. <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
      guides the user throughout the steps of Differential Expression
      analysis and produces a wealth of effective visualizations to
      facilitate data interpretation, in a comprehensive and accessible
      way for a wide range of scientists.</p>
    </div>
    
<!-- **Package**: *ideal* -->
<!-- **Authors**: Federico Marini [aut, cre] (<https://orcid.org/0000-0003-3252-7758>) -->
<!-- **Version**: 1.27.0 -->
<p><strong>Compiled date</strong>: 2024-03-27</p>
<p><strong>Last edited</strong>: 2019-02-06</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p align="center">
<img src="ideal_logo_v2.png"></p>
<p><em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> is an
R package distributed as part of the <a href="http://bioconductor.org" class="external-link">Bioconductor</a> project. To install the
package, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ideal"</span><span class="op">)</span></span></code></pre></div>
<p>The GitHub repository for <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> is <a href="https://github.com/federicomarini/ideal" class="external-link uri">https://github.com/federicomarini/ideal</a>. This is the
place to file an issue, report a bug, or provide a pull request.</p>
<p>Once <em>ideal</em> is installed, it can be loaded by the following
command.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/federicomarini/ideal" class="external-link">"ideal"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> is a
Bioconductor package containing a Shiny application for analyzing
RNA-Seq data in the context of differential expression. This enables an
interactive and at the same time analysis, keeping the functionality
accessible, and yet providing a comprehensive selection of graphs and
tables to mine the dataset at hand.</p>
<p><em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> is an
R package which fully leverages the infrastructure of the Bioconductor
project in order to deliver an interactive yet reproducible analysis for
the detection of differentially expressed genes in RNA-Seq datasets.
Graphs, tables, and interactive HTML reports can be readily exported and
shared across collaborators. The dynamic user interface displays a broad
level of content and information, subdivided by thematic tasks. All in
all, it aims to enforce a proper analysis, by reaching out both life
scientists and experienced bioinformaticians, and also fosters the
communication between the two sides, offering robust statistical methods
and high standard of accessible documentation.</p>
<p>It is structured in a similar way to the <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>,
also designed as an interactive companion tool for RNA-seq analysis
focused rather on the exploratory data analysis e.g. using principal
components analysis as a main tool.</p>
<p>The interactive/reactive design of the app, with a dynamically
generated user interface makes it easy and immediate to apply the gold
standard methods (in the current implementation, based on <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>) in
a way that is information-rich and accessible also to the bench
biologist, while also providing additional insight also for the
experienced data analyst. Reproducibility is supported via state saving
and automated report generation.</p>
<div class="section level3">
<h3 id="citation-info">Citation info<a class="anchor" aria-label="anchor" href="#citation-info"></a>
</h3>
<p>If you use <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> for
your analysis, please cite it as here below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"ideal"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Please cite the articles below for the 'ideal' software itself, or its</span></span>
<span><span class="co">## usage in combined workflows with the 'pcaExplorer' or 'GeneTonic'</span></span>
<span><span class="co">## software packages:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Federico Marini, Jan Linke, Harald Binder (2020). ideal: an</span></span>
<span><span class="co">##   R/Bioconductor package for Interactive Differential Expression</span></span>
<span><span class="co">##   Analysis. BMC Bioinformatics, 21, 565,</span></span>
<span><span class="co">##   &lt;doi:10.1186/s12859-020-03819-5&gt; &lt;doi:10.18129/B9.bioc.ideal&gt;.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Annekathrin Ludt, Arsenij Ustjanzew, Harald Binder, Konstantin</span></span>
<span><span class="co">##   Strauch, Federico Marini (2022). Interactive and Reproducible</span></span>
<span><span class="co">##   Workflows for Exploring and Modeling RNA-seq Data with pcaExplorer,</span></span>
<span><span class="co">##   ideal, and GeneTonic. Current Protocols, 2 (4), e411,</span></span>
<span><span class="co">##   &lt;doi:10.1002/cpz1.411&gt;.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## To see these entries in BibTeX format, use 'print(&lt;citation&gt;,</span></span>
<span><span class="co">## bibtex=TRUE)', 'toBibtex(.)', or set</span></span>
<span><span class="co">## 'options(citation.bibtex.max=999)'.</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="using-the-application">Using the application<a class="anchor" aria-label="anchor" href="#using-the-application"></a>
</h2>
<p>There are different ways to use <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> for
interactive differential expression analysis.</p>
<div class="section level3">
<h3 id="launching-ideal-locally">Launching <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
locally<a class="anchor" aria-label="anchor" href="#launching-ideal-locally"></a>
</h3>
<p>First load the library:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/federicomarini/ideal" class="external-link">"ideal"</a></span><span class="op">)</span></span></code></pre></div>
<p>and then launch the app with the <code>ideal</code> function. This
takes the following essential parameters as input:</p>
<ul>
<li>
<code>dds_obj</code> - a <code>DESeqDataSet</code> object. If not
provided, then a <code>countmatrix</code> and a <code>expdesign</code>
need to be provided. If none of the above is provided, it is possible to
upload the data during the execution of the Shiny App</li>
<li>
<code>res_obj</code> - a <code>DESeqResults</code> object. If not
provided, it can be computed during the execution of the
application</li>
<li>
<code>annotation_obj</code> - a <code>data.frame</code> object, with
row.names as gene identifiers (e.g. ENSEMBL ids) and a column,
<code>gene_name</code>, containing e.g. HGNC-based gene symbols. If not
provided, it can be constructed during the execution via the
<code>org.eg.XX.db</code> packages</li>
<li>
<code>countmatrix</code> - a count matrix, with genes as rows and
samples as columns. If not provided, it is possible to upload the data
during the execution of the Shiny App</li>
<li>
<code>expdesign</code> -a <code>data.frame</code> containing the
info on the experimental covariates of each sample. If not provided, it
is possible to upload the data during the execution of the Shiny
App</li>
</ul>
<p>Different modalities are supported to launch the application:</p>
<ul>
<li>
<code>ideal(dds_obj = dds, res_obj = res, annotation_obj = anno)</code>,
where the objects are precomputed in the current session and provided as
parameters</li>
<li>
<code>ideal(dds_obj = dds)</code>, as in the command above, but
where the result object is assembled at runtime</li>
<li>
<code>ideal(countmatrix = countmatrix, expdesign = expdesign)</code>,
where instead of passing the defined <code>DESeqDataSet</code> object,
its components are given, namely the count matrix (e.g. generated after
a run of featureCounts or HTSeq-count) and a data frame with the
experimental covariates. The design formula can be constructed
interactively at runtime</li>
<li>
<code><a href="../reference/ideal.html">ideal()</a></code>, where the count matrix and experimental design
can simply be uploaded at runtime, where all the derived objects can be
extracted and computed live. These files have to be formatted as tabular
text files, and a function in the package tries to guess the separator,
based on heuristics of occurrencies per line of commonly used
characters</li>
</ul>
<p>You can obtain more information on the formats that are expected and
required to work with <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> by
pressing the small round button in Step 1 of the Data Setup panel -
these are common to the formats used by <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>.
For the sake of completeness, you can see an overview reported in the
image below.</p>
<p><img src="../../../_temp/Library/pcaExplorer/www/help_dataformats.png" width="100%"></p>
<p>The value to specify for the experimental design is best selected
live in Step 2 of the Data Setup panel (see more in Section
@ref(datasetup)) - or if passing directly the <code>dds</code> object in
the call to <code>ideal</code>, is already contained in the
<code>dds</code> itself.</p>
</div>
<div class="section level3">
<h3 id="accessing-the-public-instance-of-ideal">Accessing the public instance of <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em><a class="anchor" aria-label="anchor" href="#accessing-the-public-instance-of-ideal"></a>
</h3>
<p>To use <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
without installing any additional software, you can access the public
instance of the Shiny Server made available at the Institute of Medical
Biostatistics, Epidemiology and Informatics (IMBEI) in Mainz.</p>
<p>This resource is accessible at this address:</p>
<p><a href="http://shiny.imbei.uni-mainz.de:3838/ideal" class="external-link uri">http://shiny.imbei.uni-mainz.de:3838/ideal</a></p>
</div>
<div class="section level3">
<h3 id="deploying-to-a-shiny-server">Deploying to a Shiny Server<a class="anchor" aria-label="anchor" href="#deploying-to-a-shiny-server"></a>
</h3>
<p>A deployment-oriented version of the package is available at <a href="https://github.com/federicomarini/ideal_serveredition" class="external-link uri">https://github.com/federicomarini/ideal_serveredition</a>.
This repository contains also detailed instruction to setup the running
instance of a Shiny Server, where <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> can
be run without further installation for the end-users.</p>
<p>Please note that you still need <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> to be
installed there once during the setup phase - for this operation, you
might require root administrator permissions.</p>
<!-- Following the structure of \texttt{pcaExplorer}, the deployment as a standalone web application is easily achieved also for \texttt{ideal}, with the only requirement of a running Shiny Server. Handling of sensitive human patient data can thus occur in a secured way, e.g. with restricted access to an internal instance of the application. This will also avoid the need of external software installation for the users, which could moreover seamlessly access higher performing computing infrastructures. -->
<!-- This implementation of \texttt{ideal} is made available at github.com/federicomarini/ideal\_serveredition  -->
</div>
</div>
<div class="section level2">
<h2 id="getting-to-know-the-user-interface-and-the-functionality">Getting to know the user interface and the functionality<a class="anchor" aria-label="anchor" href="#getting-to-know-the-user-interface-and-the-functionality"></a>
</h2>
<p>The user interface is dynamically displayed according to the provided
and computed objects, with tabs that are actively usable only once the
required input is effectively available.</p>
<p>Moreover, for some relevant UI widgets, the user can receive
additional information by hovering over with the mouse, with the
functionality powered by the <em><a href="https://CRAN.R-project.org/package=shinyBS" class="external-link">shinyBS</a></em>
package.</p>
<p>For the user which is either new with the app UI/functionality, or
not extensively familiar with the topic of differential expression, it
is possible to obtain a small <em>guided tour</em> of the App by
clicking on the respective help buttons, marked in the app like this -
please note that this button is clickable but does not start any
tour.</p>
<button id="btn" type="button" class="btn btn-default action-button" style="color: #ffffff; background-color: #0092AC; border-color: #2e6da4">
<i class="fa fa-info"></i> Click me for a quick tour
</button>
<p>These trigger the start of a step-by-step guide and feature
introduction, powered by the <em><a href="https://CRAN.R-project.org/package=rintrojs" class="external-link">rintrojs</a></em>
package.</p>
<div class="section level3">
<h3 id="the-controls-sidebar">The controls sidebar<a class="anchor" aria-label="anchor" href="#the-controls-sidebar"></a>
</h3>
<p>Some of the input controls which affect different tabs are located in
the sidebar, while others are as well in the individual tabs of the app.
By changing one or more of the input parameters, the user can get a fine
control on what is computed and displayed.</p>
<div class="section level4">
<h4 id="app-settings">App settings<a class="anchor" aria-label="anchor" href="#app-settings"></a>
</h4>
<ul>
<li>
<strong>Group/color by</strong> - Select the group of samples to
stratify the analysis for plotting. Can also assume multiple
values.</li>
<li>
<strong>Select the gene(s) of interest - ids</strong> - Select a
subset of genes for deeper analysis. If an annotation object is
provided, the user can handily select the genes e.g. based on their HGNC
symbol</li>
<li>
<strong>False Discovery Rate</strong> - Set as default to 0.05, it
is the FDR value for the Benjamini-Hochberg procedure for adjusting
p-values in the multiple testing comparison scenario</li>
</ul>
</div>
<div class="section level4">
<h4 id="plot-export-settings">Plot export settings<a class="anchor" aria-label="anchor" href="#plot-export-settings"></a>
</h4>
<p><strong>Width</strong> and <strong>Height</strong> for the figures to
export are input here in cm.</p>
</div>
<div class="section level4">
<h4 id="quick-viewer">Quick viewer<a class="anchor" aria-label="anchor" href="#quick-viewer"></a>
</h4>
<p>This displays a list of the underlying objects with which basically
all of the analysis can be performed. A green tick icon appears close to
each when the respective component is either provided or calculated. For
obtaining the best analysis experience in <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>, it
is recommended to provide all of them.</p>
</div>
<div class="section level4">
<h4 id="first-steps-help">First steps help<a class="anchor" aria-label="anchor" href="#first-steps-help"></a>
</h4>
<p>Clicking on this button activated the <code>intro.js</code> based
tour for getting to know the components and the structure of the app.
Dedicated step-by-step procedures are also available in each individual
tab.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-task-menu">The task menu<a class="anchor" aria-label="anchor" href="#the-task-menu"></a>
</h3>
<p>The task menu, accessible by clicking on the cog icon in the upper
right part of the application, provides two functionalities:</p>
<ul>
<li>
<code>Exit ideal &amp; save</code> will close the application and
store the content of the <code>input</code> and <code>values</code>
reactive objects in a list of two elements in the <code>ideal_env</code>
environment, respectively called
<code>ideal_inputs_YYYYMMDD_HHMMSS</code> and
<code>ideal_values_YYYYMMDD_HHMMSS</code>
</li>
<li>
<code>Save State as .RData</code> will similarly store
<code>LiveInputs</code> and <code>r_data</code> in a binary file named
<code>idealState_YYYYMMDD_HHMMSS.Rdata</code>, without closing the
application</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="the-main-app-panels">The main app panels<a class="anchor" aria-label="anchor" href="#the-main-app-panels"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> app
is a one-paged dashboard, structured in different panels, where each of
them is focused on a different aspect of the data exploration.</p>
<p>On top of the panels, three <code>valueBox</code> objects serve as
guiding elements for having an overview of the data at hand: how many
genes and samples are in the data, how many entries are in the
annotation object, and how many genes were found to be differentially
expressed in the results. Whenever each of the underlying objects is
available, the background color turns from red to green.</p>
<p>For the main analysis, the available panels are described in the
following subsections.</p>
<div class="section level3">
<h3 id="welcome">Welcome!<a class="anchor" aria-label="anchor" href="#welcome"></a>
</h3>
<p>The landing page for the app is also where you might likely be
reading this text (otherwise in the package vignette).</p>
</div>
<div class="section level3">
<h3 id="datasetup">Data Setup<a class="anchor" aria-label="anchor" href="#datasetup"></a>
</h3>
<p>The Data Setup panel is where you can upload or inspect the required
inputs for running the app. This builds on the primary idea used by
<em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>
and extends it with the following aspects:</p>
<ul>
<li>the panel structure appears dynamically in three consecutive
mandatory steps, marked with color from red to yellow to green, with
optional steps in light blue.</li>
<li>the optional step of retrieving the annotation on the fly relieves
the user from the task of composing the <code>data.frame</code> in
advance, and is based on the widely adopted <code>org.XX.eg.db</code>
Bioconductor packages.</li>
<li>when the objects are already passed as parameters, or computed, a
brief overview/summary for them is displayed</li>
<li>to tighten the concert operations between similar tools with
different scope (as <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>
and <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> are),
the information flow can move from the data exploration to decisions
taken at the moment of testing</li>
</ul>
<p>A diagnostic mean-dispersion plot is also provided in a collapsible
element at the bottom of the panel, shown when the
<code>DESeqDataSet</code> is generated and the <code>DESeq</code>
command from the <code>DESeq2</code> package has been applied.</p>
</div>
<div class="section level3">
<h3 id="counts-overview">Counts Overview<a class="anchor" aria-label="anchor" href="#counts-overview"></a>
</h3>
<p align="center">
<img src="ss_05_overview.png" style="width:800px;"></p>
<p>As in <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>,
interactive tables for the raw, normalized, or variance stabilized
transformed (preferred over regularized logarithm transformation,
because of its speed during the dispersion estimation - especially with
larger number of samples) counts are shown in this tab. The user can
also generate a sample-to-sample correlation scatter plot with the
selected data.</p>
<p>Additionally, <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> has
an option to include a filter step at the gene level by removing genes
with low absolute or averages low values. After this, it might be
possible to have to re-run the analysis in step 3 from the Data Setup
panel.</p>
</div>
<div class="section level3">
<h3 id="extract-results">Extract Results<a class="anchor" aria-label="anchor" href="#extract-results"></a>
</h3>
<p>This tab is an interface for generating the summary tables after
testing for DE. It is usually based on the Wald test, as implemented in
DESeq2, but when the factor of interest is assuming more than two
levels, the user can also perform an ANOVA-like test across the groups
with the likelihood ratio test. Options for enabling/disabling automated
independent filtering, adding the additional column of unshrunken log2
fold change values (instead of the moderated estimates used by default),
as well as using the Independent Hypothesis Weighting (<em><a href="https://bioconductor.org/packages/3.19/IHW" class="external-link">IHW</a></em>)
framework, are provided.</p>
<p>The False Discovery Rate (FDR) can be set from the sidebar panel, and
a couple of diagnostic plots, such as the histogram of raw p-values and
the distribution of log2fc, are shown below the interactive enhanced
version of the table - with clickable elements to link to ENSEMBL
database and NCBI website.</p>
</div>
<div class="section level3">
<h3 id="summary-plots">Summary Plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h3>
<p align="center">
<img src="ss_08_summaryplots.png" style="width:800px;"></p>
<p>In this tab an interactive MA plot for the contrast selected in the
Extract Results tab is displayed. Clicking on a single gene in the
zoomed plot (enabled by brushing in the main plot), it is possible to
obtain a boxplot for its expression values, flanked by an overview of
information accessed live from the Entrez database. Alternatively, a
volcano plot of -log10(p-value) versus log fold change can provide a
slightly different perspective. The subset of selected genes are also
here presented in static and interactive heatmaps, with the underlying
data accessible from the collapsible box element.</p>
</div>
<div class="section level3">
<h3 id="gene-finder">Gene Finder<a class="anchor" aria-label="anchor" href="#gene-finder"></a>
</h3>
<p align="center">
<img src="ss_09_genefinder.png" style="width:800px;"></p>
<p>The functionality in the Gene Finder builds upon the one provided by
<em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>,
and allows to query up to four genes in the same view, which can here be
selected from a dropdown input list which supports autocompletion.</p>
<p>A combined summary table (with both normalized counts and results
statistics) is located below an MA plot where the selected genes are
marked and annotated on the plot. To avoid repeating this manually, the
user can also quickly upload a list of genes as text file (one gene
identifier per line), such as members of gene families (e.g. all
cytokines, all immunoglobulines, …) or defined by common function
(e.g. all housekeeping genes, or others based on any annotation).</p>
</div>
<div class="section level3">
<h3 id="functional-analysis">Functional Analysis<a class="anchor" aria-label="anchor" href="#functional-analysis"></a>
</h3>
<p align="center">
<img src="ss_10_funcanalysis.png" style="width:800px;"></p>
<p>The Functional Analysis tab takes the user from the simple lists of
DE genes to insight on the affected biological pathways, with three
approaches based on the Gene Ontology (GO) databases. This panel of
ideal has a slim interface to the following methods for performing
functional analysis:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::goana</a></code> for the quick yet standard
implementation</li>
<li>
<code>topGO</code>, particularly valuable for pruning terms which
are topologically less meaningful than their specific nodes</li>
<li>
<code>goseq</code>, which accounts for the specific length bias
intrinsic in RNA-Seq assays (longer genes have higher chances of being
called DE).</li>
</ul>
<p><em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
allows the user to work simultaneously with more gene lists, two of
which can be uploaded in a custom way (e.g. list of gene families, or
extracted from other existing publications).</p>
<p>The interaction among these lists can be visually represented in Venn
diagrams, as well as with the appealing alternative from the UpSetR
package, where all combination of sets are explicitly shown.</p>
<p>Each of the methods for GO enrichment delivers its own interactive
<code>DT</code>-based table, which can then be explored interactively
with the display of a heatmap for all the (DE) genes annotated to a
particular term, picking the normalized transformed values for comparing
robustly the expression values. This is simply triggered by clicking any
of the rows for the results tables. Another useful feature is provided
by the clickable link to the AmiGO database on each of the GO term
identifiers.</p>
</div>
<div class="section level3">
<h3 id="signatures-explorer">Signatures Explorer<a class="anchor" aria-label="anchor" href="#signatures-explorer"></a>
</h3>
<p>The Signatures Explorer tab allows the user to check the behavior of
a number of provided gene signatures in the data at hand, displaying
this as a heatmap.</p>
<p>This panel is composed by different well panels:</p>
<ul>
<li><p>in the Setup Options, you can select and upload a gene signature
file, in <code>gmt</code> format (e.g. like the ones provided in the
MSigDB database, or from WikiPathways), and quickly compute the variance
stabilized transformed version of your data, which is more amenable for
visualization than raw or normalized counts</p></li>
<li><p>in the Conversion options tab, you can create an annotation
vector, used to bring the ids from your data and the ids the
<code>gmt</code> used for encoding the signature elements. This works
based on the <code>org.XX.eg.db</code> packages.</p></li>
<li><p>the lower well panels control the appearance of the heatmap, also
with an option to display all genes annotated in that pathway, or only
the ones detected as differentially expressed (for this you need to
provide or compute the result object)</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="report-editor">Report Editor<a class="anchor" aria-label="anchor" href="#report-editor"></a>
</h3>
<p>The Report Editor tab works in the same way of <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>,
with the scope of providing an interface to full computational
reproducibility of the analyses.</p>
<p>General <code>Markdown options</code> and <code>Editor options</code>
are available, and the text editor, based on the <code>shinyAce</code>
package, contains a comprehensive template report, that can be edited to
the best convenience of the user.</p>
<p>The code contained in the template report fetches the latest state of
the reactive values in the ongoing session, and its output is a
comprehensive HTML file that can be expanded, edited, previewed in the
tab itself, downloaded, and shared with a few mouse clicks.</p>
<p>As for <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>,
the functionality to display the report preview is based on
<code>knit2html</code>, and some elements such as <code>DataTable</code>
objects might not render correctly. To render them correctly, please
install the PhantomJS executable before launching the app. This can be
done by using the <em><a href="https://CRAN.R-project.org/package=webshot" class="external-link">webshot</a></em>
package and calling <code><a href="http://wch.github.io/webshot/reference/install_phantomjs.html" class="external-link">webshot::install_phantomjs()</a></code> - HTML
widgets will be rendered automatically as screenshots. Alternatively,
the more recent <em><a href="https://CRAN.R-project.org/package=webshot2" class="external-link">webshot2</a></em>
package uses the headless Chrome browser (via the <em><a href="https://CRAN.R-project.org/package=chromote" class="external-link">chromote</a></em>
package, requiring Google Chrome or other Chromium-based browser). Keep
in mind that the fully rendered report (the one you can obtain with the
“Generate &amp; Save” button) is not affected by this, since it uses
<code><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">rmarkdown::render()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="about">About<a class="anchor" aria-label="anchor" href="#about"></a>
</h3>
<p>The About tab contains the output of <code>sessionInfo</code>, plus
general information on <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>,
including the link to the Github development version. If requested, the
modular structure of the app can be easily expanded, and many new
operations on the same set of input data and derived results can be
embedded in the same framework.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-ideal-on-an-exemplary-data-set">Running <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> on an
exemplary data set<a class="anchor" aria-label="anchor" href="#running-ideal-on-an-exemplary-data-set"></a>
</h2>
<p>We can run <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> for
demonstration purpose on published datasets that are available as
SummarizedExperiment in an experiment Bioconductor packages.</p>
<p>We will use the <em><a href="https://bioconductor.org/packages/3.19/airway" class="external-link">airway</a></em>
dataset, which can be installed with this command:</p>
<!-- [][]or use pasilla[][] -->
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"airway"</span><span class="op">)</span></span></code></pre></div>
<p>This package provides a <code>RangedSummarizedExperiment</code>
object of read counts in genes for an RNA-Seq experiment on four human
airway smooth muscle cell lines treated with dexamethasone. More details
such as gene models and count quantifications can be found in the <em><a href="https://bioconductor.org/packages/3.19/airway" class="external-link">airway</a></em>
package vignette.</p>
<p>To run <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> on
this dataset, the following commands are required. First, prepare the
objects to be passed as parameters of <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dds_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSet</a></span><span class="op">(</span><span class="va">airway</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">cell</span> <span class="op">+</span> <span class="va">dex</span><span class="op">)</span></span>
<span><span class="va">dds_airway</span></span></code></pre></div>
<pre><code><span><span class="co">## class: DESeqDataSet </span></span>
<span><span class="co">## dim: 63677 8 </span></span>
<span><span class="co">## metadata(2): '' version</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span></span>
<span><span class="co">##   ENSG00000273493</span></span>
<span><span class="co">## rowData names(10): gene_id gene_name ... seq_coord_system symbol</span></span>
<span><span class="co">## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span><span class="co">## colData names(9): SampleName cell ... Sample BioSample</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run deseq on it</span></span>
<span><span class="va">dds_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span></span>
<span><span class="co"># extract the results</span></span>
<span><span class="va">res_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds_airway</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dex"</span>, <span class="st">"trt"</span>, <span class="st">"untrt"</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>Then launch the app itself.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ideal.html">ideal</a></span><span class="op">(</span>dds_obj <span class="op">=</span> <span class="va">dds_airway</span><span class="op">)</span></span>
<span><span class="co"># or also providing the results object</span></span>
<span><span class="fu"><a href="../reference/ideal.html">ideal</a></span><span class="op">(</span>dds_obj <span class="op">=</span> <span class="va">dds_airway</span>, res_obj <span class="op">=</span> <span class="va">res_airway</span><span class="op">)</span></span></code></pre></div>
<p>The <code>annotation</code> for this dataset can be built manually by
exploiting the <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">genenames_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span>, keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span><span class="va">annotation_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span>,</span>
<span>  gene_name <span class="op">=</span> <span class="va">genenames_airway</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotation_airway</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         gene_id gene_name</span></span>
<span><span class="co">## ENSG00000000003 ENSG00000000003    TSPAN6</span></span>
<span><span class="co">## ENSG00000000005 ENSG00000000005      TNMD</span></span>
<span><span class="co">## ENSG00000000419 ENSG00000000419      DPM1</span></span>
<span><span class="co">## ENSG00000000457 ENSG00000000457     SCYL3</span></span>
<span><span class="co">## ENSG00000000460 ENSG00000000460     FIRRM</span></span>
<span><span class="co">## ENSG00000000938 ENSG00000000938       FGR</span></span></code></pre>
<p>or alternatively, can be handily created at runtime in the optional
step.</p>
<p>Then again, the app can be launched with:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ideal.html">ideal</a></span><span class="op">(</span></span>
<span>  dds_obj <span class="op">=</span> <span class="va">dds_airway</span>,</span>
<span>  annotation_obj <span class="op">=</span> <span class="va">annotation_airway</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If desired, alternatives can be used. See the well written annotation
workflow available at the Bioconductor site (<a href="https://bioconductor.org/help/workflows/annotation/annotation/" class="external-link uri">https://bioconductor.org/help/workflows/annotation/annotation/</a>).</p>
<div class="section level3">
<h3 id="coming-from-edgerlimma-voom">Coming from <code>edgeR</code>/<code>limma-voom</code><a class="anchor" aria-label="anchor" href="#coming-from-edgerlimma-voom"></a>
</h3>
<p>Let’s suppose you performed part of your analysis with <em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em> or
<em><a href="https://bioconductor.org/packages/3.19/limma" class="external-link">limma</a></em>/<code>voom</code>
and you want to continue inspecting your data interactively using <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>. You
can use the functionality provided by the <em><a href="https://bioconductor.org/packages/3.19/DEFormats" class="external-link">DEFormats</a></em>
package to convert the object with:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/DEFormats" class="external-link">DEFormats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: limma</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'limma'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:DESeq2':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'edgeR'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:DEFormats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     DGEList</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="va">dge_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEFormats/man/as.DGEList.html" class="external-link">as.DGEList</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span> <span class="co"># this is your initial object</span></span>
<span><span class="co"># your factors for the design:</span></span>
<span><span class="va">dex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span></span>
<span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dds_airway</span><span class="op">)</span><span class="op">$</span><span class="va">cell</span></span>
<span></span>
<span><span class="va">redo_dds_airway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEFormats/man/as.DESeqDataSet.html" class="external-link">as.DESeqDataSet</a></span><span class="op">(</span><span class="va">dge_airway</span><span class="op">)</span></span>
<span><span class="co"># force the design to ~cell + dex</span></span>
<span><span class="fu">design</span><span class="op">(</span><span class="va">redo_dds_airway</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">cell</span> <span class="op">+</span> <span class="va">group</span> <span class="co"># TODO: this is due to the not 100% re-conversion via DEFormats</span></span>
<span></span>
<span><span class="co">### with edgeR</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge_airway</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">cell</span> <span class="op">+</span> <span class="va">dex</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="co"># If you performed quasi-likelihood F-tests</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span> <span class="co"># contrast for dexamethasone treatment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">qlf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  dexuntrt </span></span>
<span><span class="co">##                                                                                                              seqnames</span></span>
<span><span class="co">## ENSG00000109906             11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11</span></span>
<span><span class="co">## ENSG00000165995 10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10</span></span>
<span><span class="co">## ENSG00000250978                                                                                         5,5,5,5,5,5,5</span></span>
<span><span class="co">## ENSG00000168309                                                                   3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3</span></span>
<span><span class="co">## ENSG00000152583                         4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4</span></span>
<span><span class="co">## ENSG00000120129                                                                                               5,5,5,5</span></span>
<span><span class="co">## ENSG00000146250                                                                                             6,6,6,6,6</span></span>
<span><span class="co">## ENSG00000189221                                             X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X</span></span>
<span><span class="co">## ENSG00000171819                                                                                       1,1,1,1,1,1,1,1</span></span>
<span><span class="co">## ENSG00000157214                                                 7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                          start</span></span>
<span><span class="co">## ENSG00000109906                                                    113930315,113930447,113930459,113930979,113931229,113932093,113933933,113933933,113933933,113933933,113935134,113992540,114027059,114027059,114051488,114057674,114057674,114057674,114107929,114112889,114112889,114117512,114117920,114117920,114117920,114117920,114121048,114121048,114121048,114121048</span></span>
<span><span class="co">## ENSG00000165995                                              18429606,18429666,18429676,18429721,18429887,18430109,18439812,18439812,18439812,18470736,18549698,18550190,18629646,18629666,18689742,18690853,18690853,18787284,18787284,18789741,18795400,18803165,18803440,18803909,18807265,18807839,18816517,18823005,18825030,18827109,18828159,18828159,18828159,18828159</span></span>
<span><span class="co">## ENSG00000250978                                                                                                                                                                                                                                                                                                 66759637,66760863,66761082,66762701,66763734,66771243,66771243</span></span>
<span><span class="co">## ENSG00000168309                                                                                                                                                                                              58549844,58551188,58551188,58551579,58552935,58552946,58555418,58555418,58556053,58563036,58563036,58572585,58574933,58574933,58594601,58612698,58612698,58613143</span></span>
<span><span class="co">## ENSG00000152583 88394487,88394500,88394570,88400582,88401524,88403576,88411425,88411912,88412770,88414734,88414734,88415521,88415521,88415528,88415569,88415592,88415685,88416133,88416133,88416133,88416165,88418283,88420673,88420673,88425974,88449278,88449707,88449707,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88451697,88452130,88452130</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172195093,172196578,172197164,172197589</span></span>
<span><span class="co">## ENSG00000146250                                                                                                                                                                                                                                                                                                                   84222194,84222257,84231254,84233141,84233141</span></span>
<span><span class="co">## ENSG00000189221                                                                                           43515467,43515548,43515550,43517110,43542761,43542761,43552538,43552538,43571119,43571119,43571119,43571952,43571952,43587420,43587420,43590488,43590488,43590941,43591946,43595474,43599928,43601197,43602951,43603041,43603356,43603356,43603614,43603614,43603614</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                        11249398,11252327,11253637,11253684,11254518,11254583,11254911,11254911</span></span>
<span><span class="co">## ENSG00000157214                                                                                                             89796904,89841000,89841058,89841174,89841187,89841200,89841245,89841590,89843134,89844588,89845805,89845809,89854364,89854364,89854364,89856285,89856285,89856285,89859186,89861651,89861651,89861651,89861651,89866205,89866205,89866205,89867413</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                            end</span></span>
<span><span class="co">## ENSG00000109906                                                    113930604,113930604,113930604,113931016,113931306,113932283,113934368,113934466,113935290,113935290,113935290,113992636,114027156,114027156,114051664,114057760,114057760,114060486,114108062,114113059,114113059,114118087,114118018,114118066,114118087,114118087,114121198,114121279,114121374,114121398</span></span>
<span><span class="co">## ENSG00000165995                                              18429785,18429785,18429785,18429785,18430144,18430144,18439904,18439904,18440198,18471036,18550246,18550246,18629906,18629906,18690029,18690972,18690972,18787406,18787722,18789877,18795476,18803298,18803459,18803970,18807345,18807897,18816626,18823156,18825125,18827294,18828653,18830038,18830056,18830798</span></span>
<span><span class="co">## ENSG00000250978                                                                                                                                                                                                                                                                                                 66759788,66761155,66761155,66762766,66763881,66771305,66771420</span></span>
<span><span class="co">## ENSG00000168309                                                                                                                                                                                              58552422,58552422,58553091,58552422,58553091,58553091,58555592,58555592,58556150,58563115,58563491,58572840,58574996,58574996,58594984,58613147,58613337,58613337</span></span>
<span><span class="co">## ENSG00000152583 88394955,88394955,88394955,88400730,88401672,88403712,88411545,88412030,88412842,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88416279,88416279,88416295,88416279,88418391,88420737,88420737,88426075,88449385,88449766,88449771,88450244,88450253,88450393,88450524,88450556,88450609,88450655,88450656,88451783,88452201,88452213</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172196135,172196797,172197309,172198198</span></span>
<span><span class="co">## ENSG00000146250                                                                                                                                                                                                                                                                                                                   84222413,84222413,84231358,84235419,84235423</span></span>
<span><span class="co">## ENSG00000189221                                                                                           43515662,43515662,43515662,43517217,43542855,43542855,43552675,43552675,43571223,43571223,43571223,43572043,43572043,43587561,43587561,43590629,43590637,43591100,43592042,43595527,43599985,43601294,43603152,43603152,43603418,43603418,43603905,43604409,43606068</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                        11250012,11252427,11253831,11253831,11254716,11254716,11255235,11256038</span></span>
<span><span class="co">## ENSG00000157214                                                                                                             89797227,89841359,89841359,89841248,89841359,89841359,89841359,89841650,89843174,89845917,89845917,89845917,89854660,89854888,89854888,89856337,89856347,89856812,89859350,89861798,89861938,89866908,89866945,89866922,89866945,89866992,89867451</span></span>
<span><span class="co">##                                                                                                                                                              width</span></span>
<span><span class="co">## ENSG00000109906                                 290,158,146,38,78,191,436,534,1358,1358,157,97,98,98,177,87,87,2813,134,171,171,576,99,147,168,168,151,232,327,351</span></span>
<span><span class="co">## ENSG00000165995                    180,120,110,65,258,36,93,93,387,301,549,57,261,241,288,120,120,123,439,137,77,134,20,62,81,59,110,152,96,186,495,1880,1898,2640</span></span>
<span><span class="co">## ENSG00000250978                                                                                                                           152,293,74,66,148,63,178</span></span>
<span><span class="co">## ENSG00000168309                                                                             2579,1235,1904,844,157,146,175,175,98,80,456,256,64,64,384,450,640,195</span></span>
<span><span class="co">## ENSG00000152583 469,456,386,149,149,137,121,119,73,1017,1017,230,230,223,182,159,66,147,147,163,115,109,65,65,102,108,60,65,48,57,197,328,360,413,459,460,87,72,84</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                   1043,220,146,610</span></span>
<span><span class="co">## ENSG00000146250                                                                                                                              220,157,105,2279,2283</span></span>
<span><span class="co">## ENSG00000189221                                         196,115,113,108,95,95,138,138,105,105,105,92,92,142,142,142,150,160,97,54,58,98,202,112,63,63,292,796,2455</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                   615,101,195,148,199,134,325,1128</span></span>
<span><span class="co">## ENSG00000157214                                           324,360,302,75,173,160,115,61,41,1330,113,109,297,525,525,53,63,528,165,148,288,5258,5295,718,741,788,39</span></span>
<span><span class="co">##                                                                                        strand</span></span>
<span><span class="co">## ENSG00000109906                   +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000165995           +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000250978                                                                 -,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000168309                                           -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000152583 -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000120129                                                                       -,-,-,-</span></span>
<span><span class="co">## ENSG00000146250                                                                     +,+,+,+,+</span></span>
<span><span class="co">## ENSG00000189221                     +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000171819                                                               +,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000157214                         +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                          exon_id</span></span>
<span><span class="co">## ENSG00000109906                                                                360958,360959,360960,360961,360962,360963,360964,360965,360967,360966,360968,360969,360971,360970,360972,360974,360973,360975,360976,360977,360978,360979,360980,360981,360982,360983,360984,360985,360986,360987</span></span>
<span><span class="co">## ENSG00000165995                                    321791,321792,321793,321794,321795,321796,321797,321798,321799,321800,321801,321802,321803,321804,321805,321806,321807,321808,321809,321810,321811,321812,321813,321814,321815,321816,321817,321818,321819,321820,321821,321822,321823,321824</span></span>
<span><span class="co">## ENSG00000250978                                                                                                                                                                                                                                 199128,199130,199131,199132,199133,199134,199135</span></span>
<span><span class="co">## ENSG00000168309                                                                                                                                                    140728,140729,140730,140731,140732,140733,140734,140735,140736,140737,140738,140739,140741,140740,140742,140743,140744,140745</span></span>
<span><span class="co">## ENSG00000152583 172689,172690,172691,172692,172693,172694,172695,172696,172697,172699,172698,172700,172701,172702,172703,172704,172705,172707,172706,172708,172709,172710,172711,172712,172713,172714,172715,172716,172717,172718,172719,172720,172721,172722,172723,172724,172725,172726,172727</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                      207637,207638,207639,207640</span></span>
<span><span class="co">## ENSG00000146250                                                                                                                                                                                                                                               218407,218408,218409,218410,218411</span></span>
<span><span class="co">## ENSG00000189221                                                                       653982,653983,653984,653985,653986,653987,653989,653988,653990,653991,653992,653993,653994,653995,653996,653997,653998,653999,654000,654001,654002,654003,654004,654005,654006,654007,654008,654009,654010</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                          2221,2222,2223,2224,2225,2226,2227,2228</span></span>
<span><span class="co">## ENSG00000157214                                                                                     246728,246730,246731,246732,246733,246734,246735,246736,246737,246738,246739,246740,246741,246743,246742,246744,246745,246746,246747,246748,246749,246750,246751,246752,246753,246754,246755</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       exon_name</span></span>
<span><span class="co">## ENSG00000109906                                                                                                                                                 ENSE00001344383,ENSE00002217296,ENSE00002239785,ENSE00002278969,ENSE00001777728,ENSE00002277816,ENSE00002264739,ENSE00002267450,ENSE00003532715,ENSE00003522546,ENSE00002223922,ENSE00002239280,ENSE00003606532,ENSE00003503866,ENSE00002304588,ENSE00003667676,ENSE00003652618,ENSE00002237247,ENSE00002308595,ENSE00003574983,ENSE00003648066,ENSE00002216844,ENSE00002254046,ENSE00002202712,ENSE00003533854,ENSE00003675344,ENSE00002251269,ENSE00002289800,ENSE00002238115,ENSE00001513888</span></span>
<span><span class="co">## ENSG00000165995                                                                                 ENSE00001552836,ENSE00001323316,ENSE00001912501,ENSE00001877551,ENSE00001563168,ENSE00001473561,ENSE00003566244,ENSE00003685211,ENSE00001824310,ENSE00001935997,ENSE00001473551,ENSE00001696833,ENSE00001402115,ENSE00001811759,ENSE00001473523,ENSE00003471380,ENSE00003510985,ENSE00001005446,ENSE00001886774,ENSE00001005448,ENSE00001005464,ENSE00001005450,ENSE00001367687,ENSE00001385143,ENSE00001098801,ENSE00001098797,ENSE00001098795,ENSE00001098805,ENSE00001737387,ENSE00001098799,ENSE00001266433,ENSE00001846919,ENSE00001548772,ENSE00001552865</span></span>
<span><span class="co">## ENSG00000250978                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00002085457,ENSE00002055881,ENSE00002051029,ENSE00002080371,ENSE00002028106,ENSE00002048405,ENSE00002074015</span></span>
<span><span class="co">## ENSG00000168309                                                                                                                                                                                                                                                                                                                                                 ENSE00001419495,ENSE00001913711,ENSE00001833888,ENSE00001891124,ENSE00001228325,ENSE00001929708,ENSE00003652265,ENSE00003666445,ENSE00001518622,ENSE00001876866,ENSE00001518624,ENSE00001736715,ENSE00003684332,ENSE00003505632,ENSE00001872799,ENSE00001849960,ENSE00001901828,ENSE00001854286</span></span>
<span><span class="co">## ENSG00000152583 ENSE00001914146,ENSE00001135423,ENSE00002056886,ENSE00001175672,ENSE00001175678,ENSE00001175684,ENSE00001175691,ENSE00001006057,ENSE00001006053,ENSE00003651115,ENSE00003561541,ENSE00003487906,ENSE00003598972,ENSE00001786211,ENSE00002316459,ENSE00002028083,ENSE00001789990,ENSE00003602905,ENSE00003459209,ENSE00002234840,ENSE00001678790,ENSE00002045268,ENSE00003592507,ENSE00003613792,ENSE00001605148,ENSE00002310772,ENSE00002225747,ENSE00002211631,ENSE00002063937,ENSE00002061881,ENSE00002270604,ENSE00002252787,ENSE00002269371,ENSE00001175712,ENSE00001603831,ENSE00001942085,ENSE00002020252,ENSE00002083738,ENSE00002034086</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001876542,ENSE00000769464,ENSE00000769465,ENSE00000812856</span></span>
<span><span class="co">## ENSG00000146250                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00002225101,ENSE00001450675,ENSE00002300058,ENSE00002228069,ENSE00001450674</span></span>
<span><span class="co">## ENSG00000189221                                                                                                                                                                 ENSE00001415055,ENSE00002218265,ENSE00001834008,ENSE00001836223,ENSE00003501661,ENSE00003656032,ENSE00003691177,ENSE00003550793,ENSE00003487993,ENSE00003504764,ENSE00003590154,ENSE00003495252,ENSE00003524833,ENSE00003526715,ENSE00003672501,ENSE00001811266,ENSE00001383783,ENSE00001382812,ENSE00001377890,ENSE00001373498,ENSE00001369566,ENSE00001677475,ENSE00001816354,ENSE00001614970,ENSE00003583432,ENSE00003609031,ENSE00001875207,ENSE00002311052,ENSE00001389556</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001471825,ENSE00001168153,ENSE00001168146,ENSE00001948794,ENSE00001168140,ENSE00001823259,ENSE00001841505,ENSE00001471787</span></span>
<span><span class="co">## ENSG00000157214                                                                                                                                                                                                 ENSE00001600835,ENSE00001206871,ENSE00001519027,ENSE00001519061,ENSE00001622820,ENSE00002226827,ENSE00001547971,ENSE00001519060,ENSE00001519048,ENSE00001897780,ENSE00001519046,ENSE00001620946,ENSE00001752071,ENSE00003616543,ENSE00003474395,ENSE00001638606,ENSE00001938466,ENSE00001031278,ENSE00001275182,ENSE00001519045,ENSE00001519014,ENSE00001918233,ENSE00001845894,ENSE00003246458,ENSE00001779097,ENSE00001519026,ENSE00001553338</span></span>
<span><span class="co">##                     logFC   logCPM         F       PValue          FDR</span></span>
<span><span class="co">## ENSG00000109906 -7.134561 4.143740 1496.6892 1.434621e-12 9.135235e-08</span></span>
<span><span class="co">## ENSG00000165995 -3.275166 4.504348 1240.0348 5.070416e-12 1.314096e-07</span></span>
<span><span class="co">## ENSG00000250978 -6.158424 1.348282  446.1388 6.191073e-12 1.314096e-07</span></span>
<span><span class="co">## ENSG00000168309 -4.696469 2.775385  806.9353 1.509225e-11 2.402572e-07</span></span>
<span><span class="co">## ENSG00000152583 -4.547789 5.528678  814.2652 4.280080e-11 5.450853e-07</span></span>
<span><span class="co">## ENSG00000120129 -2.930736 7.308257  728.3966 7.513621e-11 7.082571e-07</span></span>
<span><span class="co">## ENSG00000146250  2.765000 3.909086  723.5118 7.785856e-11 7.082571e-07</span></span>
<span><span class="co">## ENSG00000189221 -3.325143 6.766776  668.9023 1.155590e-10 8.791837e-07</span></span>
<span><span class="co">## ENSG00000171819 -5.697205 3.500283  489.6501 1.242623e-10 8.791837e-07</span></span>
<span><span class="co">## ENSG00000157214 -1.961734 7.129056  635.9510 1.491084e-10 8.838056e-07</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you performed likelihood ratio tests</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">lrt</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  dexuntrt </span></span>
<span><span class="co">##                                                                                                              seqnames</span></span>
<span><span class="co">## ENSG00000109906             11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11,11</span></span>
<span><span class="co">## ENSG00000152583                         4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4</span></span>
<span><span class="co">## ENSG00000165995 10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10</span></span>
<span><span class="co">## ENSG00000163884                                                                                             3,3,3,3,3</span></span>
<span><span class="co">## ENSG00000189221                                             X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X,X</span></span>
<span><span class="co">## ENSG00000127954                                                                                   7,7,7,7,7,7,7,7,7,7</span></span>
<span><span class="co">## ENSG00000162692                                                                       1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1</span></span>
<span><span class="co">## ENSG00000101347                      20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20</span></span>
<span><span class="co">## ENSG00000171819                                                                                       1,1,1,1,1,1,1,1</span></span>
<span><span class="co">## ENSG00000120129                                                                                               5,5,5,5</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                          start</span></span>
<span><span class="co">## ENSG00000109906                                                    113930315,113930447,113930459,113930979,113931229,113932093,113933933,113933933,113933933,113933933,113935134,113992540,114027059,114027059,114051488,114057674,114057674,114057674,114107929,114112889,114112889,114117512,114117920,114117920,114117920,114117920,114121048,114121048,114121048,114121048</span></span>
<span><span class="co">## ENSG00000152583 88394487,88394500,88394570,88400582,88401524,88403576,88411425,88411912,88412770,88414734,88414734,88415521,88415521,88415528,88415569,88415592,88415685,88416133,88416133,88416133,88416165,88418283,88420673,88420673,88425974,88449278,88449707,88449707,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88450197,88451697,88452130,88452130</span></span>
<span><span class="co">## ENSG00000165995                                              18429606,18429666,18429676,18429721,18429887,18430109,18439812,18439812,18439812,18470736,18549698,18550190,18629646,18629666,18689742,18690853,18690853,18787284,18787284,18789741,18795400,18803165,18803440,18803909,18807265,18807839,18816517,18823005,18825030,18827109,18828159,18828159,18828159,18828159</span></span>
<span><span class="co">## ENSG00000163884                                                                                                                                                                                                                                                                                                              126061478,126070684,126071518,126072855,126076080</span></span>
<span><span class="co">## ENSG00000189221                                                                                           43515467,43515548,43515550,43517110,43542761,43542761,43552538,43552538,43571119,43571119,43571119,43571952,43571952,43587420,43587420,43590488,43590488,43590941,43591946,43595474,43599928,43601197,43602951,43603041,43603356,43603356,43603614,43603614,43603614</span></span>
<span><span class="co">## ENSG00000127954                                                                                                                                                                                                                                                                      87905744,87907343,87910230,87910798,87911956,87913129,87920231,87936107,87936107,87936107</span></span>
<span><span class="co">## ENSG00000162692                                                                                                                                                                                                101185298,101185311,101185316,101185320,101186032,101186032,101188576,101190180,101194663,101196754,101197974,101200058,101200731,101203679,101203679,101203679</span></span>
<span><span class="co">## ENSG00000101347                                                                                                             35518632,35525971,35526225,35526843,35532560,35532560,35533767,35533767,35539621,35540751,35540864,35545125,35545352,35547767,35555585,35555585,35559163,35559163,35563432,35563432,35563744,35569442,35569442,35575141,35575141,35579839,35579839</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                        11249398,11252327,11253637,11253684,11254518,11254583,11254911,11254911</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172195093,172196578,172197164,172197589</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                            end</span></span>
<span><span class="co">## ENSG00000109906                                                    113930604,113930604,113930604,113931016,113931306,113932283,113934368,113934466,113935290,113935290,113935290,113992636,114027156,114027156,114051664,114057760,114057760,114060486,114108062,114113059,114113059,114118087,114118018,114118066,114118087,114118087,114121198,114121279,114121374,114121398</span></span>
<span><span class="co">## ENSG00000152583 88394955,88394955,88394955,88400730,88401672,88403712,88411545,88412030,88412842,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88415750,88416279,88416279,88416295,88416279,88418391,88420737,88420737,88426075,88449385,88449766,88449771,88450244,88450253,88450393,88450524,88450556,88450609,88450655,88450656,88451783,88452201,88452213</span></span>
<span><span class="co">## ENSG00000165995                                              18429785,18429785,18429785,18429785,18430144,18430144,18439904,18439904,18440198,18471036,18550246,18550246,18629906,18629906,18690029,18690972,18690972,18787406,18787722,18789877,18795476,18803298,18803459,18803970,18807345,18807897,18816626,18823156,18825125,18827294,18828653,18830038,18830056,18830798</span></span>
<span><span class="co">## ENSG00000163884                                                                                                                                                                                                                                                                                                              126062738,126071790,126071790,126073128,126076285</span></span>
<span><span class="co">## ENSG00000189221                                                                                           43515662,43515662,43515662,43517217,43542855,43542855,43552675,43552675,43571223,43571223,43571223,43572043,43572043,43587561,43587561,43590629,43590637,43591100,43592042,43595527,43599985,43601294,43603152,43603152,43603418,43603418,43603905,43604409,43606068</span></span>
<span><span class="co">## ENSG00000127954                                                                                                                                                                                                                                                                      87908943,87908943,87910394,87912483,87912483,87913586,87920329,87936195,87936203,87936206</span></span>
<span><span class="co">## ENSG00000162692                                                                                                                                                                                                101185480,101185480,101185480,101185480,101186121,101186307,101188896,101190446,101194938,101197074,101198240,101200324,101204595,101204169,101204594,101204601</span></span>
<span><span class="co">## ENSG00000101347                                                                                                             35521469,35526362,35526362,35526947,35532652,35532652,35533797,35533906,35539736,35540955,35540955,35545233,35545452,35547922,35555655,35555655,35559278,35559278,35563592,35563592,35563806,35569514,35569514,35575207,35575207,35580111,35580246</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                        11250012,11252427,11253831,11253831,11254716,11254716,11255235,11256038</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                        172196135,172196797,172197309,172198198</span></span>
<span><span class="co">##                                                                                                                                                              width</span></span>
<span><span class="co">## ENSG00000109906                                 290,158,146,38,78,191,436,534,1358,1358,157,97,98,98,177,87,87,2813,134,171,171,576,99,147,168,168,151,232,327,351</span></span>
<span><span class="co">## ENSG00000152583 469,456,386,149,149,137,121,119,73,1017,1017,230,230,223,182,159,66,147,147,163,115,109,65,65,102,108,60,65,48,57,197,328,360,413,459,460,87,72,84</span></span>
<span><span class="co">## ENSG00000165995                    180,120,110,65,258,36,93,93,387,301,549,57,261,241,288,120,120,123,439,137,77,134,20,62,81,59,110,152,96,186,495,1880,1898,2640</span></span>
<span><span class="co">## ENSG00000163884                                                                                                                              1261,1107,273,274,206</span></span>
<span><span class="co">## ENSG00000189221                                         196,115,113,108,95,95,138,138,105,105,105,92,92,142,142,142,150,160,97,54,58,98,202,112,63,63,292,796,2455</span></span>
<span><span class="co">## ENSG00000127954                                                                                                            3200,1601,165,1686,528,458,99,89,97,100</span></span>
<span><span class="co">## ENSG00000162692                                                                                    183,170,165,161,90,276,321,267,276,321,267,267,3865,491,916,923</span></span>
<span><span class="co">## ENSG00000101347                                                  2838,392,138,105,93,93,31,140,116,205,92,109,101,156,71,71,116,116,161,161,63,73,73,67,67,273,408</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                   615,101,195,148,199,134,325,1128</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                   1043,220,146,610</span></span>
<span><span class="co">##                                                                                        strand</span></span>
<span><span class="co">## ENSG00000109906                   +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000152583 -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000165995           +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000163884                                                                     -,-,-,-,-</span></span>
<span><span class="co">## ENSG00000189221                     +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000127954                                                           -,-,-,-,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000162692                                               +,+,+,+,+,+,+,+,+,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000101347                         -,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-,-</span></span>
<span><span class="co">## ENSG00000171819                                                               +,+,+,+,+,+,+,+</span></span>
<span><span class="co">## ENSG00000120129                                                                       -,-,-,-</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                          exon_id</span></span>
<span><span class="co">## ENSG00000109906                                                                360958,360959,360960,360961,360962,360963,360964,360965,360967,360966,360968,360969,360971,360970,360972,360974,360973,360975,360976,360977,360978,360979,360980,360981,360982,360983,360984,360985,360986,360987</span></span>
<span><span class="co">## ENSG00000152583 172689,172690,172691,172692,172693,172694,172695,172696,172697,172699,172698,172700,172701,172702,172703,172704,172705,172707,172706,172708,172709,172710,172711,172712,172713,172714,172715,172716,172717,172718,172719,172720,172721,172722,172723,172724,172725,172726,172727</span></span>
<span><span class="co">## ENSG00000165995                                    321791,321792,321793,321794,321795,321796,321797,321798,321799,321800,321801,321802,321803,321804,321805,321806,321807,321808,321809,321810,321811,321812,321813,321814,321815,321816,321817,321818,321819,321820,321821,321822,321823,321824</span></span>
<span><span class="co">## ENSG00000163884                                                                                                                                                                                                                                               145255,145256,145257,145258,145259</span></span>
<span><span class="co">## ENSG00000189221                                                                       653982,653983,653984,653985,653986,653987,653989,653988,653990,653991,653992,653993,653994,653995,653996,653997,653998,653999,654000,654001,654002,654003,654004,654005,654006,654007,654008,654009,654010</span></span>
<span><span class="co">## ENSG00000127954                                                                                                                                                                                                            263537,263538,263539,263540,263541,263542,263543,263544,263545,263546</span></span>
<span><span class="co">## ENSG00000162692                                                                                                                                                                                  14792,14793,14794,14795,14796,14797,14798,14799,14800,14801,14802,14803,14804,14805,14806,14807</span></span>
<span><span class="co">## ENSG00000101347                                                                                     626560,626561,626562,626563,626565,626564,626566,626567,626568,626569,626570,626571,626572,626573,626575,626574,626576,626577,626578,626579,626580,626581,626582,626584,626583,626585,626586</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                          2221,2222,2223,2224,2225,2226,2227,2228</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                      207637,207638,207639,207640</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       exon_name</span></span>
<span><span class="co">## ENSG00000109906                                                                                                                                                 ENSE00001344383,ENSE00002217296,ENSE00002239785,ENSE00002278969,ENSE00001777728,ENSE00002277816,ENSE00002264739,ENSE00002267450,ENSE00003532715,ENSE00003522546,ENSE00002223922,ENSE00002239280,ENSE00003606532,ENSE00003503866,ENSE00002304588,ENSE00003667676,ENSE00003652618,ENSE00002237247,ENSE00002308595,ENSE00003574983,ENSE00003648066,ENSE00002216844,ENSE00002254046,ENSE00002202712,ENSE00003533854,ENSE00003675344,ENSE00002251269,ENSE00002289800,ENSE00002238115,ENSE00001513888</span></span>
<span><span class="co">## ENSG00000152583 ENSE00001914146,ENSE00001135423,ENSE00002056886,ENSE00001175672,ENSE00001175678,ENSE00001175684,ENSE00001175691,ENSE00001006057,ENSE00001006053,ENSE00003651115,ENSE00003561541,ENSE00003487906,ENSE00003598972,ENSE00001786211,ENSE00002316459,ENSE00002028083,ENSE00001789990,ENSE00003602905,ENSE00003459209,ENSE00002234840,ENSE00001678790,ENSE00002045268,ENSE00003592507,ENSE00003613792,ENSE00001605148,ENSE00002310772,ENSE00002225747,ENSE00002211631,ENSE00002063937,ENSE00002061881,ENSE00002270604,ENSE00002252787,ENSE00002269371,ENSE00001175712,ENSE00001603831,ENSE00001942085,ENSE00002020252,ENSE00002083738,ENSE00002034086</span></span>
<span><span class="co">## ENSG00000165995                                                                                 ENSE00001552836,ENSE00001323316,ENSE00001912501,ENSE00001877551,ENSE00001563168,ENSE00001473561,ENSE00003566244,ENSE00003685211,ENSE00001824310,ENSE00001935997,ENSE00001473551,ENSE00001696833,ENSE00001402115,ENSE00001811759,ENSE00001473523,ENSE00003471380,ENSE00003510985,ENSE00001005446,ENSE00001886774,ENSE00001005448,ENSE00001005464,ENSE00001005450,ENSE00001367687,ENSE00001385143,ENSE00001098801,ENSE00001098797,ENSE00001098795,ENSE00001098805,ENSE00001737387,ENSE00001098799,ENSE00001266433,ENSE00001846919,ENSE00001548772,ENSE00001552865</span></span>
<span><span class="co">## ENSG00000163884                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001241585,ENSE00001079374,ENSE00002029422,ENSE00002086065,ENSE00001241575</span></span>
<span><span class="co">## ENSG00000189221                                                                                                                                                                 ENSE00001415055,ENSE00002218265,ENSE00001834008,ENSE00001836223,ENSE00003501661,ENSE00003656032,ENSE00003691177,ENSE00003550793,ENSE00003487993,ENSE00003504764,ENSE00003590154,ENSE00003495252,ENSE00003524833,ENSE00003526715,ENSE00003672501,ENSE00001811266,ENSE00001383783,ENSE00001382812,ENSE00001377890,ENSE00001373498,ENSE00001369566,ENSE00001677475,ENSE00001816354,ENSE00001614970,ENSE00003583432,ENSE00003609031,ENSE00001875207,ENSE00002311052,ENSE00001389556</span></span>
<span><span class="co">## ENSG00000127954                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001917465,ENSE00001277946,ENSE00001163797,ENSE00001753614,ENSE00001483677,ENSE00001120706,ENSE00001621319,ENSE00001673636,ENSE00001905312,ENSE00001871078</span></span>
<span><span class="co">## ENSG00000162692                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001167786,ENSE00001811760,ENSE00001825682,ENSE00001451872,ENSE00001451878,ENSE00001067757,ENSE00001067755,ENSE00001067758,ENSE00001067756,ENSE00001067753,ENSE00001067759,ENSE00001067751,ENSE00003565397,ENSE00001892498,ENSE00001451871,ENSE00001167778</span></span>
<span><span class="co">## ENSG00000101347                                                                                                                                                                                                 ENSE00000800455,ENSE00001930417,ENSE00000661812,ENSE00000661813,ENSE00003692177,ENSE00003551249,ENSE00001842453,ENSE00000661815,ENSE00001622813,ENSE00001604475,ENSE00001711142,ENSE00001793017,ENSE00001667406,ENSE00001752622,ENSE00003598833,ENSE00003555165,ENSE00003535434,ENSE00003608826,ENSE00003546823,ENSE00003559540,ENSE00002218059,ENSE00003462531,ENSE00003673643,ENSE00003619308,ENSE00003541183,ENSE00001957470,ENSE00001898623</span></span>
<span><span class="co">## ENSG00000171819                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001471825,ENSE00001168153,ENSE00001168146,ENSE00001948794,ENSE00001168140,ENSE00001823259,ENSE00001841505,ENSE00001471787</span></span>
<span><span class="co">## ENSG00000120129                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSE00001876542,ENSE00000769464,ENSE00000769465,ENSE00000812856</span></span>
<span><span class="co">##                     logFC   logCPM        LR        PValue           FDR</span></span>
<span><span class="co">## ENSG00000109906 -7.141823 4.143740 1293.2297 3.345918e-283 2.130580e-278</span></span>
<span><span class="co">## ENSG00000152583 -4.552568 5.528678  901.7878 4.010221e-198 1.276794e-193</span></span>
<span><span class="co">## ENSG00000165995 -3.274517 4.504348  749.6857 4.696169e-165 9.967933e-161</span></span>
<span><span class="co">## ENSG00000163884 -4.458891 4.685058  703.8821 4.280694e-155 6.814544e-151</span></span>
<span><span class="co">## ENSG00000189221 -3.335100 6.766776  647.7867 6.767998e-143 8.619316e-139</span></span>
<span><span class="co">## ENSG00000127954 -5.197020 3.652849  638.6801 6.471522e-141 6.868119e-137</span></span>
<span><span class="co">## ENSG00000162692  3.677380 4.599336  634.3726 5.595308e-140 5.089891e-136</span></span>
<span><span class="co">## ENSG00000101347 -3.751932 9.206504  631.1278 2.841479e-139 2.261711e-135</span></span>
<span><span class="co">## ENSG00000171819 -5.668260 3.500283  615.0520 8.911131e-136 6.304823e-132</span></span>
<span><span class="co">## ENSG00000120129 -2.932273 7.308257  603.5237 2.866583e-133 1.825354e-129</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># lrt to DESeqResults</span></span>
<span></span>
<span><span class="va">tbledger</span> <span class="op">&lt;-</span> <span class="va">lrt</span><span class="op">$</span><span class="va">table</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span> <span class="op">==</span> <span class="st">"PValue"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"pvalue"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span> <span class="op">==</span> <span class="st">"logFC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"log2FoldChange"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span> <span class="op">==</span> <span class="st">"logCPM"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"baseMean"</span></span>
<span><span class="co"># get from the logcpm to something more the baseMean for better</span></span>
<span><span class="va">tbledger</span><span class="op">$</span><span class="va">baseMean</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">tbledger</span><span class="op">$</span><span class="va">baseMean</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">dge_airway</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">lib.size</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1e6</span></span>
<span><span class="co"># use the constructor for DESeqResults</span></span>
<span><span class="va">edger_resu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html" class="external-link">DESeqResults</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="va">tbledger</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edger_resu</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">:::</span><span class="fu">pvalueAdjustment</span><span class="op">(</span><span class="va">edger_resu</span>,</span>
<span>  independentFiltering <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>, pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># cfr with res_airway</span></span>
<span><span class="co"># plot(res_airway$pvalue,edger_resu$pvalue)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">### with limma-voom</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge_airway</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span></span>
<span><span class="va">contr.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span><span class="va">dextrt</span> <span class="op">-</span> <span class="va">dexuntrt</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">vfit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">v</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">vfit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">vfit</span>, contrasts <span class="op">=</span> <span class="va">contr.matrix</span><span class="op">)</span></span>
<span><span class="va">efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">vfit</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tbllimma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span>, number <span class="op">=</span> <span class="cn">Inf</span>, sort.by <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span> <span class="op">==</span> <span class="st">"P.Value"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"pvalue"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span> <span class="op">==</span> <span class="st">"logFC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"log2FoldChange"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span> <span class="op">==</span> <span class="st">"AveExpr"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"baseMean"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span> <span class="op">==</span> <span class="st">"adj.P.Val"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"padj"</span></span>
<span><span class="co"># get from the logcpm to something more the baseMean for better</span></span>
<span><span class="va">tbllimma</span><span class="op">$</span><span class="va">baseMean</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">tbllimma</span><span class="op">$</span><span class="va">baseMean</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">dge_airway</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">lib.size</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1e6</span></span>
<span><span class="co"># use the constructor for DESeqResults</span></span>
<span><span class="va">limma_resu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html" class="external-link">DESeqResults</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="va">tbllimma</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># cfr with res_airway</span></span>
<span><span class="co"># plot(res_airway$pvalue,limma_resu$pvalue)</span></span></code></pre></div>
<p>After preparing the objects, you can launch the app with the
following command:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ideal.html">ideal</a></span><span class="op">(</span><span class="va">redo_dds_airway</span>, <span class="va">edger_resu</span><span class="op">)</span></span>
<span><span class="co"># or ...</span></span>
<span><span class="fu"><a href="../reference/ideal.html">ideal</a></span><span class="op">(</span><span class="va">redo_dds_airway</span>, <span class="va">limma_resu</span><span class="op">)</span></span></code></pre></div>
<p>Basically, you need a <code>data.frame</code>
(e.g. <code>myresults_df</code>) with 3 columns to create the
<code>DESeqResults</code> object to be passed to the app, with this
names:</p>
<ul>
<li>
<code>baseMean</code> as an average expression value</li>
<li>
<code>log2FoldChange</code> as an indicator for the effect size</li>
<li>
<code>pvalue</code> as a measure of significance</li>
</ul>
<p>Then, you just need to call
<code>DESeqResults(DataFrame(myresults_df))</code>. <code>padj</code>
can be computed either by <code>p.adjust</code> or using
<code>DESeq2:::pvalueAdjustment</code> as in the example above.</p>
</div>
</div>
<div class="section level2">
<h2 id="functions-exported-by-the-package-for-standalone-usage">Functions exported by the package for standalone usage<a class="anchor" aria-label="anchor" href="#functions-exported-by-the-package-for-standalone-usage"></a>
</h2>
<p>The functions exported by the <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
package can be also used in a standalone scenario, provided the required
objects are in the working environment. They are listed here for an
overview, but please refer to the documentation for additional details.
Where possible, for each function a code snippet will be provided for
its typical usage.</p>
<div class="section level3">
<h3 id="deseqresult2degenes-and-deseqresult2tbl">
<code>deseqresult2DEgenes</code> and
<code>deseqresult2tbl</code><a class="anchor" aria-label="anchor" href="#deseqresult2degenes-and-deseqresult2tbl"></a>
</h3>
<p><code>deseqresult2DEgenes</code> and <code>deseqresult2tbl</code>
generate a tidy table with the results of DESeq2, sorted by the values
in the <code>padj</code> column.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## out of 33469 with nonzero total read count</span></span>
<span><span class="co">## adjusted p-value &lt; 0.05</span></span>
<span><span class="co">## LFC &gt; 0 (up)       : 2211, 6.6%</span></span>
<span><span class="co">## LFC &lt; 0 (down)     : 1817, 5.4%</span></span>
<span><span class="co">## outliers [1]       : 0, 0%</span></span>
<span><span class="co">## low counts [2]     : 16687, 50%</span></span>
<span><span class="co">## (mean count &lt; 7)</span></span>
<span><span class="co">## [1] see 'cooksCutoff' argument of ?results</span></span>
<span><span class="co">## [2] see 'independentFiltering' argument of ?results</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_airway</span></span></code></pre></div>
<pre><code><span><span class="co">## log2 fold change (MLE): dex trt vs untrt </span></span>
<span><span class="co">## Wald test p-value: dex trt vs untrt </span></span>
<span><span class="co">## DataFrame with 63677 rows and 6 columns</span></span>
<span><span class="co">##                  baseMean log2FoldChange     lfcSE      stat      pvalue</span></span>
<span><span class="co">##                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## ENSG00000000003  708.6022     -0.3812540  0.100654 -3.787752 0.000152016</span></span>
<span><span class="co">## ENSG00000000005    0.0000             NA        NA        NA          NA</span></span>
<span><span class="co">## ENSG00000000419  520.2979      0.2068126  0.112219  1.842943 0.065337292</span></span>
<span><span class="co">## ENSG00000000457  237.1630      0.0379204  0.143445  0.264356 0.791505742</span></span>
<span><span class="co">## ENSG00000000460   57.9326     -0.0881682  0.287142 -0.307054 0.758801924</span></span>
<span><span class="co">## ...                   ...            ...       ...       ...         ...</span></span>
<span><span class="co">## ENSG00000273489  0.275899       1.483738   3.51395  0.422243    0.672848</span></span>
<span><span class="co">## ENSG00000273490  0.000000             NA        NA        NA          NA</span></span>
<span><span class="co">## ENSG00000273491  0.000000             NA        NA        NA          NA</span></span>
<span><span class="co">## ENSG00000273492  0.105978      -0.463679   3.52308 -0.131612    0.895291</span></span>
<span><span class="co">## ENSG00000273493  0.106142      -0.521355   3.53139 -0.147634    0.882631</span></span>
<span><span class="co">##                       padj</span></span>
<span><span class="co">##                  &lt;numeric&gt;</span></span>
<span><span class="co">## ENSG00000000003 0.00119659</span></span>
<span><span class="co">## ENSG00000000005         NA</span></span>
<span><span class="co">## ENSG00000000419 0.18625623</span></span>
<span><span class="co">## ENSG00000000457 0.90370473</span></span>
<span><span class="co">## ENSG00000000460 0.88616659</span></span>
<span><span class="co">## ...                    ...</span></span>
<span><span class="co">## ENSG00000273489         NA</span></span>
<span><span class="co">## ENSG00000273490         NA</span></span>
<span><span class="co">## ENSG00000273491         NA</span></span>
<span><span class="co">## ENSG00000273492         NA</span></span>
<span><span class="co">## ENSG00000273493         NA</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/deseqresult2tbl.html">deseqresult2tbl</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                              id   baseMean log2FoldChange     lfcSE     stat</span></span>
<span><span class="co">## ENSG00000152583 ENSG00000152583   997.4398       4.574919 0.1840561 24.85611</span></span>
<span><span class="co">## ENSG00000165995 ENSG00000165995   495.0929       3.291062 0.1331737 24.71255</span></span>
<span><span class="co">## ENSG00000120129 ENSG00000120129  3409.0294       2.947810 0.1214377 24.27426</span></span>
<span><span class="co">## ENSG00000101347 ENSG00000101347 12703.3871       3.766995 0.1554380 24.23472</span></span>
<span><span class="co">## ENSG00000189221 ENSG00000189221  2341.7673       3.353580 0.1417823 23.65301</span></span>
<span><span class="co">## ENSG00000211445 ENSG00000211445 12285.6151       3.730403 0.1658306 22.49526</span></span>
<span><span class="co">##                        pvalue          padj</span></span>
<span><span class="co">## ENSG00000152583 2.220933e-136 3.727170e-132</span></span>
<span><span class="co">## ENSG00000165995 7.839410e-135 6.578049e-131</span></span>
<span><span class="co">## ENSG00000120129 3.666925e-130 2.051278e-126</span></span>
<span><span class="co">## ENSG00000101347 9.583815e-130 4.020890e-126</span></span>
<span><span class="co">## ENSG00000189221 1.098955e-123 3.688532e-120</span></span>
<span><span class="co">## ENSG00000211445 4.618318e-112 1.291744e-108</span></span></code></pre>
<p>In particular, <code>deseqresult2DEgenes</code> only includes genes
detected as DE.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span><span class="op">(</span><span class="va">res_airway</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                              id   baseMean log2FoldChange     lfcSE     stat</span></span>
<span><span class="co">## ENSG00000152583 ENSG00000152583   997.4398       4.574919 0.1840561 24.85611</span></span>
<span><span class="co">## ENSG00000165995 ENSG00000165995   495.0929       3.291062 0.1331737 24.71255</span></span>
<span><span class="co">## ENSG00000120129 ENSG00000120129  3409.0294       2.947810 0.1214377 24.27426</span></span>
<span><span class="co">## ENSG00000101347 ENSG00000101347 12703.3871       3.766995 0.1554380 24.23472</span></span>
<span><span class="co">## ENSG00000189221 ENSG00000189221  2341.7673       3.353580 0.1417823 23.65301</span></span>
<span><span class="co">## ENSG00000211445 ENSG00000211445 12285.6151       3.730403 0.1658306 22.49526</span></span>
<span><span class="co">##                        pvalue          padj</span></span>
<span><span class="co">## ENSG00000152583 2.220933e-136 3.727170e-132</span></span>
<span><span class="co">## ENSG00000165995 7.839410e-135 6.578049e-131</span></span>
<span><span class="co">## ENSG00000120129 3.666925e-130 2.051278e-126</span></span>
<span><span class="co">## ENSG00000101347 9.583815e-130 4.020890e-126</span></span>
<span><span class="co">## ENSG00000189221 1.098955e-123 3.688532e-120</span></span>
<span><span class="co">## ENSG00000211445 4.618318e-112 1.291744e-108</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the output in the first lines is the same, but</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 63677     6</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4028    7</span></span></code></pre>
<p>This tables can be enhanced with clickable links to the ENSEMBL and
NCBI gene databases by the following commands:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myde</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span><span class="op">(</span><span class="va">res_airway</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">myde</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">myde</span><span class="op">$</span><span class="va">id</span><span class="op">)</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span>, keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myde_enhanced</span> <span class="op">&lt;-</span> <span class="va">myde</span></span>
<span><span class="va">myde_enhanced</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu">ideal</span><span class="fu">:::</span><span class="fu">createLinkENS</span><span class="op">(</span><span class="va">myde_enhanced</span><span class="op">$</span><span class="va">id</span>, species <span class="op">=</span> <span class="st">"Homo_sapiens"</span><span class="op">)</span></span>
<span><span class="va">myde_enhanced</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="fu">ideal</span><span class="fu">:::</span><span class="fu">createLinkGeneSymbol</span><span class="op">(</span><span class="va">myde_enhanced</span><span class="op">$</span><span class="va">symbol</span><span class="op">)</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">myde_enhanced</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span>, escape <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["ENSG00000152583","ENSG00000165995","ENSG00000120129","ENSG00000101347","ENSG00000189221","ENSG00000211445","ENSG00000157214","ENSG00000162614","ENSG00000125148","ENSG00000154734","ENSG00000139132","ENSG00000162493","ENSG00000134243","ENSG00000179094","ENSG00000162692","ENSG00000163884","ENSG00000178695","ENSG00000198624","ENSG00000107562","ENSG00000148848","ENSG00000146250","ENSG00000106976","ENSG00000154930","ENSG00000163394","ENSG00000253368","ENSG00000163661","ENSG00000162616","ENSG00000148175","ENSG00000124766","ENSG00000197381","ENSG00000172403","ENSG00000116285","ENSG00000163083","ENSG00000166741","ENSG00000070404","ENSG00000105989","ENSG00000122035","ENSG00000119139","ENSG00000136478","ENSG00000196569","ENSG00000108821","ENSG00000221866","ENSG00000100242","ENSG00000145685","ENSG00000134686","ENSG00000172986","ENSG00000102804","ENSG00000060718","ENSG00000099337","ENSG00000126803","ENSG00000157150","ENSG00000133142","ENSG00000138316","ENSG00000103196","ENSG00000131389","ENSG00000152268","ENSG00000168309","ENSG00000163171","ENSG00000124440","ENSG00000141401","ENSG00000103742","ENSG00000112837","ENSG00000164647","ENSG00000068383","ENSG00000156675","ENSG00000136999","ENSG00000206538","ENSG00000227051","ENSG00000134363","ENSG00000196517","ENSG00000124762","ENSG00000149218","ENSG00000109906","ENSG00000185432","ENSG00000163378","ENSG00000167641","ENSG00000196975","ENSG00000106484","ENSG00000126016","ENSG00000151726","ENSG00000150907","ENSG00000130513","ENSG00000221869","ENSG00000169715","ENSG00000147027","ENSG00000145390","ENSG00000185950","ENSG00000013293","ENSG00000131831","ENSG00000146122","ENSG00000137672","ENSG00000020577","ENSG00000119138","ENSG00000138735","ENSG00000163110","ENSG00000071282","ENSG00000101825","ENSG00000162407","ENSG00000163513","ENSG00000246430"],["<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000152583\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000152583<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000165995\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000165995<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000120129\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000120129<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000101347\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000101347<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000189221\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000189221<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000211445\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000211445<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157214\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000157214<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162614\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162614<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000125148\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000125148<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000154734\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000154734<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000139132\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000139132<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162493\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162493<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134243\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134243<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000179094\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000179094<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162692\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162692<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163884\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163884<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000178695\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000178695<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000198624\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000198624<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000107562\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000107562<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000148848\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000148848<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000146250\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000146250<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000106976\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000106976<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000154930\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000154930<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163394\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163394<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000253368\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000253368<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163661\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163661<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162616\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162616<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000148175\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000148175<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124766\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124766<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000197381\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000197381<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000172403\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000172403<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000116285\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000116285<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163083\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163083<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000166741\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000166741<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000070404\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000070404<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000105989\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000105989<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000122035\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000122035<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000119139\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000119139<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000136478\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000136478<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196569\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196569<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000108821\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000108821<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000221866\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000221866<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000100242\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000100242<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000145685\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000145685<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134686\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134686<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000172986\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000172986<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000102804\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000102804<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000060718\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000060718<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000099337\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000099337<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000126803\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000126803<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000157150\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000157150<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000133142\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000133142<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000138316\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000138316<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000103196\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000103196<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000131389\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000131389<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000152268\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000152268<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000168309\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000168309<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163171\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163171<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124440\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124440<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000141401\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000141401<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000103742\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000103742<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000112837\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000112837<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000164647\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000164647<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000068383\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000068383<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000156675\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000156675<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000136999\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000136999<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000206538\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000206538<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000227051\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000227051<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000134363\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000134363<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196517\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196517<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000124762\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000124762<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000149218\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000149218<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000109906\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000109906<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000185432\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000185432<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163378\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163378<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000167641\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000167641<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000196975\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000196975<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000106484\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000106484<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000126016\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000126016<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000151726\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000151726<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000150907\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000150907<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000130513\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000130513<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000221869\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000221869<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000169715\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000169715<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000147027\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000147027<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000145390\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000145390<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000185950\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000185950<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000013293\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000013293<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000131831\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000131831<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000146122\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000146122<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000137672\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000137672<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000020577\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000020577<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000119138\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000119138<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000138735\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000138735<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163110\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163110<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000071282\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000071282<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000101825\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000101825<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000162407\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000162407<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163513\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000163513<\/a>","<a href=\"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000246430\" target=\"_blank\" class=\"btn btn-primary\">ENSG00000246430<\/a>"],[997.439773207048,495.0929066985461,3409.029375309564,12703.38706218211,2341.767252755909,12285.6151498691,3009.263222957941,5393.101683492233,3656.252782246999,30315.13547110783,1223.466140811016,1099.625865946361,5510.958255979163,776.596665468835,508.1702324397553,561.1071661290711,2649.850148958817,2057.19173059348,25136.3075719089,1365.173985907024,313.6899778001127,884.7458829930462,670.0467035304332,1216.602064987596,1040.176490349064,8565.701164266109,1467.635827039921,11193.71877388702,1281.468658015947,1969.048356310049,8014.781650838901,3232.635437121343,418.9214765558838,7432.936351634205,2128.114024809806,333.2146881638375,943.3405742311438,5029.851420770923,3701.528597671657,17178.42604309081,100990.6943515149,847.4939460219118,2236.238321237396,2018.163688242023,2737.98198594824,756.366829107728,5397.336845219525,363.7622940043581,530.0604850219261,449.1768481733303,167.6155056830764,3055.973848054261,799.5685841064611,3082.06913512982,740.4108031269425,1597.83892249109,159.5269197640233,1357.57154103081,197.0436228038409,627.5272330351917,370.8298225527897,2754.968128065713,1920.710129862853,1247.241851871935,1631.877858083132,435.7127167221307,12754.94768128105,4615.740067532874,2906.341053088197,229.6407427404245,8020.528816804033,2027.185731346813,385.0710288625155,3885.649240401704,1805.289310021479,231.671616289766,4908.818151642603,677.213238740759,386.4151621032122,1460.178002980321,678.868174958036,451.0047652992933,1963.574298620355,554.1946412541968,12593.32959101075,8442.053719625275,1974.883079170174,244.4973261848038,1017.175972499539,1942.414433516911,655.4494367844393,3105.005556376751,1683.268207727695,6850.269516931819,1314.184325320424,3659.868842963534,4767.14944686345,25853.66201739551,13216.64760190749,2162.538814709049],[4.574918738434916,3.291061764033147,2.947809863919499,3.766995431587298,3.353579943787815,3.730402666336068,1.976772518578735,2.03566491652844,2.210978555837228,2.3456036828563,2.228902981872019,1.891217015274022,2.195711631104428,3.191749908544629,-3.692660580825031,4.459128208667638,-2.528174560603742,2.918435724477911,-1.911669836411317,-1.814543145333749,-2.756242784061944,-1.756256440690706,1.898529923501016,-2.188504177760467,1.989058831967787,2.260266820559857,1.519118066996143,1.447873824057865,-2.411990128665098,2.321090813886851,1.744357526232243,2.424512991778049,3.44998558648886,2.187373187804034,1.815524354648702,-3.041840702728003,2.288870596121995,1.927936864387675,1.465830774082189,2.03064228241387,-1.304946442022902,2.753151919191844,1.859046079604784,-1.239943635869887,1.387548863723746,-1.572903674267433,1.201387993683162,2.427342087912566,2.718277713005477,2.458712576459318,2.945320211552584,1.539090675264068,-2.507192760179377,2.63126675642951,-2.124916145609236,1.912956949199613,4.835847401876525,1.50562831455222,2.529887056211458,2.326868076046045,-2.197342478247093,-1.833562444714175,1.56739405864982,1.381124133937094,1.547251881576838,-2.552480792583466,1.366220326307899,-1.151117902147002,-1.807542787660367,-2.229245441318286,-1.007978472919033,1.926430978447248,7.352625808980627,2.249366170788731,1.261851283858287,2.896522095825831,1.185074020212085,-1.863765086024298,-2.374282088384736,1.602490233568768,2.227845028324551,-1.868533375359122,1.499829096984271,2.225180242902452,1.164749521777263,2.147508444170347,2.124238233117072,-2.882906886746321,1.428473889366713,1.68143695109109,1.954307268380026,1.203307975794107,2.082919349813931,-1.178827884383419,1.605456475080747,1.605270350767553,-1.323768200495953,-1.001459940546598,1.458381805415252,2.051206972177098],[0.1840560920940141,0.1331736959281713,0.1214376897788147,0.1554379897234309,0.141782348968745,0.165830584142871,0.08998232436857162,0.09418230724133246,0.1056407805937341,0.115808058684838,0.112839437391301,0.09577959302360103,0.1120076816871916,0.1639629204576108,0.1901858982034105,0.23486368758588,0.1344301237509221,0.1612963417242065,0.105748095217681,0.1004836488658721,0.1548884573700105,0.09970926690862285,0.1086809788148181,0.1254269943908341,0.114476718104887,0.1307842425627709,0.08795653654371176,0.08434154998999992,0.1408057976265046,0.1357544671102961,0.1032920129831144,0.1442773223450313,0.2088101706640176,0.1344678665949618,0.1133767966113206,0.1913709590526709,0.1448732545159464,0.1222394109538642,0.09313763209325222,0.132700384998173,0.08547599914730807,0.1818605646011955,0.1229560270760433,0.08204202338198313,0.0918342407669082,0.1042774451426527,0.07984595349858406,0.1617193757844728,0.1812773963460193,0.1644376350257268,0.1972841126659491,0.1036537901906494,0.1690594594941362,0.1796502106525131,0.1452333692887273,0.1310717645164683,0.3317017111744149,0.1033087400782958,0.1742710078523639,0.1608040442183412,0.1521383963209018,0.1273420510102338,0.1093882299628378,0.09704344648912419,0.1094458412860098,0.1827621563588516,0.09796194076983809,0.08276356158091056,0.1307875976673001,0.1613270836829059,0.07311772866428656,0.1402561578428176,0.5363887085116283,0.1643695698712802,0.09243773321108903,0.212578531690494,0.08716509871974336,0.137305943233995,0.1751746741296271,0.1183812041170814,0.1648916655365215,0.1385684998037007,0.1112434661773584,0.1667652507475587,0.08734118218684346,0.1614240201506207,0.1599903806286851,0.2184899772665556,0.1087276278803911,0.1289263450504798,0.1499998160087051,0.09247201055427356,0.1605224643189063,0.09092570825428338,0.1239216265236845,0.1246534474108873,0.1031952852033454,0.07826277285526721,0.1146872097490929,0.1614564662281606],[24.8561114516008,24.71255108672674,24.27425842247665,24.23471532467624,23.65301441385409,22.49526337748499,21.96845360964212,21.61409054582045,20.9292144890528,20.25423540895082,19.75287216421242,19.74551107988115,19.60322361850575,19.46629091282739,-19.41605879146519,18.9860265522618,-18.80660740362072,18.09362626133217,-18.07758175195658,-18.05809368801728,-17.7950173361053,-17.61377347503921,17.46883349970494,-17.44843036691947,17.37522585287037,17.28240938104623,17.2712356203474,17.16679174415854,-17.1299063626843,17.09771223956145,16.88763221719187,16.80453277320991,16.52211468204775,16.26688400131121,16.01319148990158,-15.89499638704741,15.79912457802938,15.77181082061431,15.73832983658587,15.30245961563584,-15.26681706023672,15.13880661939695,15.11960107864448,-15.11351847207348,15.10927571389842,-15.08383401718064,15.04632283844498,15.00959347720672,14.99512773129681,14.95224968465793,14.92933298962467,14.84837816767949,-14.830242375561,14.64661102746501,-14.63104626723115,14.59472950758409,14.57890399405792,14.57406520891777,14.51697036350812,14.47020867762879,-14.44305008718701,-14.39871927747435,14.32872676687709,14.23201858449945,14.13714640406916,-13.96613414634755,13.94643997017003,-13.90851094562499,-13.820444903793,-13.81817231445122,-13.78569180597875,13.73509019551329,13.7076446470745,13.68480901027019,13.65082461484368,13.62565670574418,13.59574001083142,-13.5738121899654,-13.55379766043005,13.5366948285462,13.51096200693728,-13.48454647344909,13.48240169533242,13.34318890133068,13.33562808075563,13.30352473049897,13.2772872017045,-13.19468711019729,13.13809486341546,13.0418414516656,13.02873110368755,13.01267235979327,12.97587448991465,-12.96473689362641,12.95541803410645,12.87786566765539,-12.82779729604389,-12.79612137431697,12.71616781510186,12.70439654784996],[2.220932851917361e-136,7.839409795597029e-135,3.666924609466664e-130,9.58381484422359e-130,1.098954841979815e-123,4.61831818401843e-112,5.769974564957389e-107,1.323849134881932e-103,2.902397283549528e-97,3.259713239700498e-91,7.578283590491733e-87,8.767336381538136e-87,1.451320598901691e-85,2.120834051929248e-84,5.646021815867905e-84,2.225441940725254e-80,6.667027282906713e-79,3.577533078248825e-73,4.786167184694882e-73,6.813522181883761e-73,7.724347685740672e-71,1.931299669200472e-69,2.475034052982922e-68,3.538228049632937e-68,1.271048802629488e-67,6.3825213222146e-67,7.746590236700797e-67,4.707295814220601e-66,8.879751523977087e-66,1.543451453777832e-65,5.548546075762554e-64,2.260869848270583e-63,2.54322372376952e-61,1.695700829785033e-59,1.033687746035297e-57,6.863241826185704e-57,3.154669537227556e-56,4.86351232159387e-56,8.259503202505441e-56,7.362091249703313e-53,1.272336377027588e-52,8.982896832677114e-52,1.202689673859649e-51,1.319044304411795e-51,1.406774963522209e-51,2.068973490039437e-51,3.649645009786684e-51,6.353590579871306e-51,7.901054904536934e-51,1.505751760289628e-50,2.123798499887231e-50,7.127212380081771e-50,9.339513124936186e-50,1.416045958204927e-48,1.780282147526719e-48,3.034137213794536e-48,3.826099679866781e-48,4.107059431864551e-48,9.459981196965308e-48,1.869042643263274e-47,2.772942649005625e-47,5.271124800427369e-47,1.447500769353662e-46,5.798357177406951e-46,2.241926057642992e-45,2.508796422536356e-44,3.307053802946358e-44,5.623809073287414e-44,1.918802534703275e-43,1.980342276269985e-43,3.107730497213099e-43,6.257809156721593e-43,9.13762134003465e-43,1.251355103492965e-42,1.996069622925683e-42,2.818633303469869e-42,4.244458069160934e-42,5.726450013396624e-42,7.523481011794151e-42,9.496660802899337e-42,1.347490441138912e-41,1.928476715711054e-41,1.985373084815742e-41,1.297865966693351e-40,1.436398343638809e-40,2.208085079936485e-40,3.135490104838739e-40,9.414495183121697e-40,1.991784801614773e-39,7.07279009918611e-39,8.399283752425501e-39,1.036527974992288e-38,1.676716621396832e-38,1.938945559579057e-38,2.189401385377157e-38,5.997114779397339e-38,1.14575459559318e-37,1.723465696710467e-37,4.80871289174999e-37,5.58989544171519e-37],[3.727169512087715e-132,6.578048759485467e-131,2.051277626535652e-126,4.020889517894007e-126,3.688532031621052e-120,1.291743596069955e-108,1.383310187844499e-103,2.777104522698573e-100,5.412003468058687e-94,5.470450758865376e-88,1.156170501960293e-83,1.226111992958108e-83,1.873543253136013e-82,2.542274075676903e-81,6.316769207593012e-81,2.334210415578201e-77,6.58153246245532e-76,3.335453339953988e-70,4.227445141765764e-70,5.717226462818664e-70,6.172857279147617e-68,1.47323050220556e-66,1.8059139772678e-65,2.474105963705831e-65,8.532296402291227e-65,4.119672031900209e-64,4.814936198233806e-64,2.821351369794648e-63,5.138620347427017e-63,8.634067432433189e-63,3.003732265917651e-61,1.185684931052404e-60,1.293344864615154e-58,8.369779801603655e-57,4.956385072561246e-55,3.199414564640236e-54,1.430855788479807e-53,2.147880625815483e-53,3.554127762678111e-53,3.088765383813025e-50,5.207890019335848e-50,3.589308920142555e-49,4.69384607132852e-49,5.030954890145171e-49,5.246332763962159e-49,7.548155023878659e-49,1.303156224558301e-48,2.221374106487505e-48,2.70603068179467e-48,5.053905208236108e-48,6.988546357864219e-48,2.300170733894851e-47,2.957277533258096e-47,4.400756161221311e-46,5.4321263635988e-46,9.092659057482125e-46,1.126484295219725e-45,1.188356403199153e-45,2.690803465211387e-45,5.227712273207377e-45,7.628774350100392e-45,1.42677445807697e-44,3.855866335125897e-44,1.520437971113179e-43,5.788308169133029e-43,6.379185085303809e-42,8.283429391200862e-42,1.387922998057491e-41,4.666861469187008e-41,4.747729154337554e-41,7.345624394962004e-41,1.458591017612525e-40,2.100651525047418e-40,2.837870452272828e-40,4.466405388258507e-40,6.223987381425178e-40,9.250713677488156e-40,1.232067745190027e-39,1.598215928353537e-39,1.992162019928208e-39,2.791800565826323e-39,3.946792224763768e-39,4.01428085655154e-39,2.592950792029502e-38,2.835957294464294e-38,4.308846954824894e-38,6.048252291885485e-38,1.795387024581231e-37,3.755745229292035e-37,1.318839593828237e-36,1.548975603661591e-36,1.89076222568702e-36,3.025662187127056e-36,3.461636636261249e-36,3.867635163094678e-36,1.048370627373397e-35,1.982273569406676e-35,2.951347073693374e-35,8.151496944378619e-35,9.380962530286432e-35],["<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SPARCL1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SPARCL1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CACNB2[sym]\" target=\"_blank\" class=\"btn btn-primary\">CACNB2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DUSP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">DUSP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SAMHD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SAMHD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MAOA[sym]\" target=\"_blank\" class=\"btn btn-primary\">MAOA<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GPX3[sym]\" target=\"_blank\" class=\"btn btn-primary\">GPX3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STEAP2[sym]\" target=\"_blank\" class=\"btn btn-primary\">STEAP2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NEXN[sym]\" target=\"_blank\" class=\"btn btn-primary\">NEXN<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MT2A[sym]\" target=\"_blank\" class=\"btn btn-primary\">MT2A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAMTS1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAMTS1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FGD4[sym]\" target=\"_blank\" class=\"btn btn-primary\">FGD4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDPN[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDPN<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SORT1[sym]\" target=\"_blank\" class=\"btn btn-primary\">SORT1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PER1[sym]\" target=\"_blank\" class=\"btn btn-primary\">PER1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=VCAM1[sym]\" target=\"_blank\" class=\"btn btn-primary\">VCAM1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KLF15[sym]\" target=\"_blank\" class=\"btn btn-primary\">KLF15<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KCTD12[sym]\" target=\"_blank\" class=\"btn btn-primary\">KCTD12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCDC69[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCDC69<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CXCL12[sym]\" target=\"_blank\" class=\"btn btn-primary\">CXCL12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAM12[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAM12<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PRSS35[sym]\" target=\"_blank\" class=\"btn btn-primary\">PRSS35<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DNM1[sym]\" target=\"_blank\" class=\"btn btn-primary\">DNM1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ACSS1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ACSS1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCKAR[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCKAR<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TRNP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">TRNP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PTX3[sym]\" target=\"_blank\" class=\"btn btn-primary\">PTX3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DNAJB4[sym]\" target=\"_blank\" class=\"btn btn-primary\">DNAJB4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STOM[sym]\" target=\"_blank\" class=\"btn btn-primary\">STOM<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SOX4[sym]\" target=\"_blank\" class=\"btn btn-primary\">SOX4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADARB1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADARB1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SYNPO2[sym]\" target=\"_blank\" class=\"btn btn-primary\">SYNPO2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ERRFI1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ERRFI1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=INHBB[sym]\" target=\"_blank\" class=\"btn btn-primary\">INHBB<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NNMT[sym]\" target=\"_blank\" class=\"btn btn-primary\">NNMT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FSTL3[sym]\" target=\"_blank\" class=\"btn btn-primary\">FSTL3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=WNT2[sym]\" target=\"_blank\" class=\"btn btn-primary\">WNT2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RASL11A[sym]\" target=\"_blank\" class=\"btn btn-primary\">RASL11A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TJP2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TJP2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TEX2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TEX2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LAMA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">LAMA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=COL1A1[sym]\" target=\"_blank\" class=\"btn btn-primary\">COL1A1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PLXNA4[sym]\" target=\"_blank\" class=\"btn btn-primary\">PLXNA4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SUN2[sym]\" target=\"_blank\" class=\"btn btn-primary\">SUN2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LHFPL2[sym]\" target=\"_blank\" class=\"btn btn-primary\">LHFPL2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PHC2[sym]\" target=\"_blank\" class=\"btn btn-primary\">PHC2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GXYLT2[sym]\" target=\"_blank\" class=\"btn btn-primary\">GXYLT2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TSC22D1[sym]\" target=\"_blank\" class=\"btn btn-primary\">TSC22D1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=COL11A1[sym]\" target=\"_blank\" class=\"btn btn-primary\">COL11A1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KCNK6[sym]\" target=\"_blank\" class=\"btn btn-primary\">KCNK6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=HSPA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">HSPA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TIMP4[sym]\" target=\"_blank\" class=\"btn btn-primary\">TIMP4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TCEAL4[sym]\" target=\"_blank\" class=\"btn btn-primary\">TCEAL4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ADAMTS14[sym]\" target=\"_blank\" class=\"btn btn-primary\">ADAMTS14<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CRISPLD2[sym]\" target=\"_blank\" class=\"btn btn-primary\">CRISPLD2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC6A6[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC6A6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=NA[sym]\" target=\"_blank\" class=\"btn btn-primary\">NA<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FAM107A[sym]\" target=\"_blank\" class=\"btn btn-primary\">FAM107A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CDC42EP3[sym]\" target=\"_blank\" class=\"btn btn-primary\">CDC42EP3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=HIF3A[sym]\" target=\"_blank\" class=\"btn btn-primary\">HIF3A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IMPA2[sym]\" target=\"_blank\" class=\"btn btn-primary\">IMPA2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IGDCC4[sym]\" target=\"_blank\" class=\"btn btn-primary\">IGDCC4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TBX18[sym]\" target=\"_blank\" class=\"btn btn-primary\">TBX18<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=STEAP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">STEAP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=INPP5A[sym]\" target=\"_blank\" class=\"btn btn-primary\">INPP5A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RAB11FIP1[sym]\" target=\"_blank\" class=\"btn btn-primary\">RAB11FIP1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CCN3[sym]\" target=\"_blank\" class=\"btn btn-primary\">CCN3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=VGLL3[sym]\" target=\"_blank\" class=\"btn btn-primary\">VGLL3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=C14orf132[sym]\" target=\"_blank\" class=\"btn btn-primary\">C14orf132<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FST[sym]\" target=\"_blank\" class=\"btn btn-primary\">FST<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC6A9[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC6A9<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CDKN1A[sym]\" target=\"_blank\" class=\"btn btn-primary\">CDKN1A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ENDOD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ENDOD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ZBTB16[sym]\" target=\"_blank\" class=\"btn btn-primary\">ZBTB16<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TMT1A[sym]\" target=\"_blank\" class=\"btn btn-primary\">TMT1A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=EOGT[sym]\" target=\"_blank\" class=\"btn btn-primary\">EOGT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PPP1R14A[sym]\" target=\"_blank\" class=\"btn btn-primary\">PPP1R14A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ANXA4[sym]\" target=\"_blank\" class=\"btn btn-primary\">ANXA4<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MEST[sym]\" target=\"_blank\" class=\"btn btn-primary\">MEST<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=AMOT[sym]\" target=\"_blank\" class=\"btn btn-primary\">AMOT<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=ACSL1[sym]\" target=\"_blank\" class=\"btn btn-primary\">ACSL1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=FOXO1[sym]\" target=\"_blank\" class=\"btn btn-primary\">FOXO1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=GDF15[sym]\" target=\"_blank\" class=\"btn btn-primary\">GDF15<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=CEBPD[sym]\" target=\"_blank\" class=\"btn btn-primary\">CEBPD<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MT1E[sym]\" target=\"_blank\" class=\"btn btn-primary\">MT1E<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TMEM47[sym]\" target=\"_blank\" class=\"btn btn-primary\">TMEM47<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=USP53[sym]\" target=\"_blank\" class=\"btn btn-primary\">USP53<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=IRS2[sym]\" target=\"_blank\" class=\"btn btn-primary\">IRS2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SLC7A14[sym]\" target=\"_blank\" class=\"btn btn-primary\">SLC7A14<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=RAI2[sym]\" target=\"_blank\" class=\"btn btn-primary\">RAI2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=DAAM2[sym]\" target=\"_blank\" class=\"btn btn-primary\">DAAM2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TRPC6[sym]\" target=\"_blank\" class=\"btn btn-primary\">TRPC6<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=SAMD4A[sym]\" target=\"_blank\" class=\"btn btn-primary\">SAMD4A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=KLF9[sym]\" target=\"_blank\" class=\"btn btn-primary\">KLF9<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDE5A[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDE5A<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PDLIM5[sym]\" target=\"_blank\" class=\"btn btn-primary\">PDLIM5<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LMCD1[sym]\" target=\"_blank\" class=\"btn btn-primary\">LMCD1<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=MXRA5[sym]\" target=\"_blank\" class=\"btn btn-primary\">MXRA5<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=PLPP3[sym]\" target=\"_blank\" class=\"btn btn-primary\">PLPP3<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=TGFBR2[sym]\" target=\"_blank\" class=\"btn btn-primary\">TGFBR2<\/a>","<a href=\"http://www.ncbi.nlm.nih.gov/gene/?term=LINC00968[sym]\" target=\"_blank\" class=\"btn btn-primary\">LINC00968<\/a>"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>id<\/th>\n      <th>baseMean<\/th>\n      <th>log2FoldChange<\/th>\n      <th>lfcSE<\/th>\n      <th>stat<\/th>\n      <th>pvalue<\/th>\n      <th>padj<\/th>\n      <th>symbol<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"id","targets":1},{"name":"baseMean","targets":2},{"name":"log2FoldChange","targets":3},{"name":"lfcSE","targets":4},{"name":"stat","targets":5},{"name":"pvalue","targets":6},{"name":"padj","targets":7},{"name":"symbol","targets":8}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="ggplotcounts">
<code>ggplotCounts</code><a class="anchor" aria-label="anchor" href="#ggplotcounts"></a>
</h3>
<p><code>ggplotCounts</code> extends the functionality of the
<code>plotCounts</code> function of <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>,
and plots the normalized counts of a single gene as a boxplot, with
jittered points superimposed.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggplotCounts.html">ggplotCounts</a></span><span class="op">(</span></span>
<span>  dds <span class="op">=</span> <span class="va">dds_airway</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">"ENSG00000103196"</span>, <span class="co"># CRISPLD2 in the original publication</span></span>
<span>  intgroup <span class="op">=</span> <span class="st">"dex"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ideal-usersguide_files/figure-html/ggpc1-1.png" width="700"></p>
<p>If an <code>annotation_obj</code> is provided, their gene name can
also be included in the title.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggplotCounts.html">ggplotCounts</a></span><span class="op">(</span></span>
<span>  dds <span class="op">=</span> <span class="va">dds_airway</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">"ENSG00000103196"</span>, <span class="co"># CRISPLD2 in the original publication</span></span>
<span>  intgroup <span class="op">=</span> <span class="st">"dex"</span>,</span>
<span>  annotation_obj <span class="op">=</span> <span class="va">annotation_airway</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ideal-usersguide_files/figure-html/ggpc2-1.png" width="700"></p>
<p>When used in the context of the app, it is possible to seamless
search for genes also by their gene name, making exploration even more
immediate.</p>
</div>
<div class="section level3">
<h3 id="goseqtable">
<code>goseqTable</code><a class="anchor" aria-label="anchor" href="#goseqtable"></a>
</h3>
<p><code>goseqTable</code> is a wrapper to extract the functional GO
terms enriched in in a list of (DE) genes, based on the algorithm and
the implementation in the <em><a href="https://bioconductor.org/packages/3.19/goseq" class="external-link">goseq</a></em>
package.</p>
<p>Its counterpart, based on the <em><a href="https://bioconductor.org/packages/3.19/topGO" class="external-link">topGO</a></em>
package, can be found in the <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>
package.</p>
<p><em>Please note that the following code chunks are not run only for
reducing compilation time. The reader is invited to try out the
functionality by simply executing these commands.</em></p>
<!-- maybe eval=FALSE here? -->
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span> <span class="co"># taking only a subset of the DE genes</span></span>
<span><span class="va">myde</span> <span class="op">&lt;-</span> <span class="va">res_subset</span><span class="op">$</span><span class="va">id</span></span>
<span><span class="va">myassayed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span></span>
<span><span class="va">mygo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/goseqTable.html">goseqTable</a></span><span class="op">(</span></span>
<span>  de.genes <span class="op">=</span> <span class="va">myde</span>,</span>
<span>  assayed.genes <span class="op">=</span> <span class="va">myassayed</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  id <span class="op">=</span> <span class="st">"ensGene"</span>,</span>
<span>  testCats <span class="op">=</span> <span class="st">"GO:BP"</span>,</span>
<span>  addGeneToTerms <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mygo</span><span class="op">)</span></span></code></pre></div>
<p>As for the results, this table can be enhanced by adding the links
for each category to the AmiGO database.</p>
<!-- maybe eval=FALSE here? -->
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygo_enhanced</span> <span class="op">&lt;-</span> <span class="va">mygo</span></span>
<span><span class="co"># using the unexported function to link the GO term to the entry in the AmiGO db</span></span>
<span><span class="va">mygo_enhanced</span><span class="op">$</span><span class="va">category</span> <span class="op">&lt;-</span> <span class="fu">ideal</span><span class="fu">:::</span><span class="fu">createLinkGO</span><span class="op">(</span><span class="va">mygo_enhanced</span><span class="op">$</span><span class="va">category</span><span class="op">)</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">mygo_enhanced</span>, escape <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot_ma">
<code>plot_ma</code><a class="anchor" aria-label="anchor" href="#plot_ma"></a>
</h3>
<p>The MA plot provided by <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em>
displays the gene-wise log2-fold changes (logFCs) versus the average
expression value. As a main input parameter, a <code>DESeqResults</code>
object is required. Control on the appearance of the plot can be applied
by selecting the False Discovery Rate (<code>FDR</code>), the point
transparency (<code>point_alpha</code>), adding horizontal lines at
particular logFC values (<code>hlines</code>). The user can also decide
to add rug plots in the margins (setting <code>add_rug</code> to
<code>TRUE</code>).</p>
<p>To facilitate the inspection of a particular gene or gene set,
<code>intgenes</code> can assume the value of a vector of genes (either
the IDs or the gene symbols if <code>symbol</code> column is provided in
<code>res_obj</code>. Labels can be added via
<code>labels_intgenes</code>, while classical labels/title can be also
edited as preferred (see <code>plot_ma</code> for all details).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span><span class="op">(</span><span class="va">res_airway</span>,</span>
<span>  FDR <span class="op">=</span> <span class="fl">0.05</span>, hlines <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Adding horizontal lines"</span>, add_rug <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span><span class="op">(</span><span class="va">res_airway</span>,</span>
<span>  FDR <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  intgenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"ENSG00000103196"</span>, <span class="co"># CRISPLD2</span></span>
<span>    <span class="st">"ENSG00000120129"</span>, <span class="co"># DUSP1</span></span>
<span>    <span class="st">"ENSG00000163884"</span>, <span class="co"># KLF15</span></span>
<span>    <span class="st">"ENSG00000179094"</span></span>
<span>  <span class="op">)</span>, <span class="co"># PER1</span></span>
<span>  title <span class="op">=</span> <span class="st">"Providing a shortlist of genes"</span>,</span>
<span>  add_rug <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_airway2</span> <span class="op">&lt;-</span> <span class="va">res_airway</span></span>
<span><span class="va">res_airway2</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_airway2</span><span class="op">)</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span>, keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ma.html">plot_ma</a></span><span class="op">(</span><span class="va">res_airway2</span>,</span>
<span>  FDR <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  intgenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CRISPLD2"</span>, <span class="co"># CRISPLD2</span></span>
<span>    <span class="st">"DUSP1"</span>, <span class="co"># DUSP1</span></span>
<span>    <span class="st">"KLF15"</span>, <span class="co"># KLF15</span></span>
<span>    <span class="st">"PER1"</span></span>
<span>  <span class="op">)</span>, <span class="co"># PER1</span></span>
<span>  annotation_obj <span class="op">=</span> <span class="va">annotation_airway</span>,</span>
<span>  hlines <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  add_rug <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Putting gene names..."</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ideal-usersguide_files/figure-html/plotma2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plot_volcano">
<code>plot_volcano</code><a class="anchor" aria-label="anchor" href="#plot_volcano"></a>
</h3>
<p>The volcano plot gives a different flavor for the gene overview,
displaying log2-fold changes and log p-values.</p>
<p>In a way similar to <code>plot_ma</code>, genes can be annotated with
<code>intgenes</code>, and vertical lines can be added via
<code>vlines</code>. <code>ylim_up</code> controls the y axis upper
limit to visualize better the bulk of genes - keep in mind that very
small p-values due to robust differences/large effect sizes can be “cut
out”.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_airway</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_airway2</span>,</span>
<span>  FDR <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  intgenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CRISPLD2"</span>, <span class="co"># CRISPLD2</span></span>
<span>    <span class="st">"DUSP1"</span>, <span class="co"># DUSP1</span></span>
<span>    <span class="st">"KLF15"</span>, <span class="co"># KLF15</span></span>
<span>    <span class="st">"PER1"</span></span>
<span>  <span class="op">)</span>, <span class="co"># PER1</span></span>
<span>  title <span class="op">=</span> <span class="st">"Selecting the handful of genes - and displaying the full range for the -log10(pvalue)"</span>,</span>
<span>  ylim_up <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">res_airway2</span><span class="op">$</span><span class="va">pvalue</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ideal-usersguide_files/figure-html/plotvolcano2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="sepguesser">
<code>sepguesser</code><a class="anchor" aria-label="anchor" href="#sepguesser"></a>
</h3>
<p><code>sepguesser</code> makes an educated guess on the separator
character for the input text file (<code>file</code>). The separator
list can be provided as a vector in <code>sep_list</code> (defaults to
comma, tab, semicolon, and whitespace - which ideally could cover most
of the cases). The heuristics is based on the number of occurrencies of
each separator in each line.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/design_commas.txt"</span>, package <span class="op">=</span> <span class="st">"ideal"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] ","</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/design_semicolons.txt"</span>, package <span class="op">=</span> <span class="st">"ideal"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] ";"</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/design_spaces.txt"</span>, package <span class="op">=</span> <span class="st">"ideal"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] " "</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anyfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/design_tabs.txt"</span>, package <span class="op">=</span> <span class="st">"ideal"</span><span class="op">)</span> <span class="co"># we know it is going to be TAB</span></span>
<span><span class="va">guessed_sep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sepguesser.html">sepguesser</a></span><span class="op">(</span><span class="va">anyfile</span><span class="op">)</span></span>
<span><span class="va">guessed_sep</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "\t"</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># to be used for reading in the same file, without having to specify the sep</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">anyfile</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>, as.is <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sep <span class="op">=</span> <span class="va">guessed_sep</span>,</span>
<span>  quote <span class="op">=</span> <span class="st">""</span>, row.names <span class="op">=</span> <span class="fl">1</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            SampleName    cell   dex albut        Run avgLength Experiment</span></span>
<span><span class="co">## SRR1039508 GSM1275862  N61311 untrt untrt SRR1039508       126  SRX384345</span></span>
<span><span class="co">## SRR1039509 GSM1275863  N61311   trt untrt SRR1039509       126  SRX384346</span></span>
<span><span class="co">## SRR1039512 GSM1275866 N052611 untrt untrt SRR1039512       126  SRX384349</span></span>
<span><span class="co">## SRR1039513 GSM1275867 N052611   trt untrt SRR1039513        87  SRX384350</span></span>
<span><span class="co">## SRR1039516 GSM1275870 N080611 untrt untrt SRR1039516       120  SRX384353</span></span>
<span><span class="co">## SRR1039517 GSM1275871 N080611   trt untrt SRR1039517       126  SRX384354</span></span>
<span><span class="co">## SRR1039520 GSM1275874 N061011 untrt untrt SRR1039520       101  SRX384357</span></span>
<span><span class="co">## SRR1039521 GSM1275875 N061011   trt untrt SRR1039521        98  SRX384358</span></span>
<span><span class="co">##               Sample    BioSample sizeFactor</span></span>
<span><span class="co">## SRR1039508 SRS508568 SAMN02422669  1.0236476</span></span>
<span><span class="co">## SRR1039509 SRS508567 SAMN02422675  0.8961667</span></span>
<span><span class="co">## SRR1039512 SRS508571 SAMN02422678  1.1794861</span></span>
<span><span class="co">## SRR1039513 SRS508572 SAMN02422670  0.6700538</span></span>
<span><span class="co">## SRR1039516 SRS508575 SAMN02422682  1.1776714</span></span>
<span><span class="co">## SRR1039517 SRS508576 SAMN02422673  1.3990365</span></span>
<span><span class="co">## SRR1039520 SRS508579 SAMN02422683  0.9207787</span></span>
<span><span class="co">## SRR1039521 SRS508580 SAMN02422677  0.9445141</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="creating-and-sharing-output-objects">Creating and sharing output objects<a class="anchor" aria-label="anchor" href="#creating-and-sharing-output-objects"></a>
</h2>
<p>While running the app, the user can</p>
<ul>
<li>generate and save graphics</li>
<li>create and export tables</li>
<li>generate, preview, download/export an HTML report</li>
<li>save the values of the <code>reactiveValues</code> in an
environment, or in binary format</li>
</ul>
<p>This functionality to retrieve and share the output is provided by
action buttons that are placed close to each element of interest.</p>
<div class="section level3">
<h3 id="expanding-the-analysis-after-ideal">Expanding the analysis after <code>ideal</code><a class="anchor" aria-label="anchor" href="#expanding-the-analysis-after-ideal"></a>
</h3>
<p>After exporting the objects to the R environment, via the button
“Exit ideal and save”, it is possible to continue the analyses
e.g. using alternative methods for functional enrichment analysis, with
Functional Class Scoring methods (such as GSEA, as implemented in <em><a href="https://bioconductor.org/packages/3.19/fgsea" class="external-link">fgsea</a></em>), or
additional methods for Overrepresentation Analysis (e.g. <em><a href="https://CRAN.R-project.org/package=gprofiler2" class="external-link">gprofiler2</a></em>).<br>
In the example below, we show an example of the code required to perform
<code>fgsea</code> on the set of objects exported from
<code><a href="../reference/ideal.html">ideal()</a></code>.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"fgsea"</span><span class="op">)</span></span>
<span><span class="co"># selecting the result object from the exported environment</span></span>
<span><span class="va">exported_res</span> <span class="op">&lt;-</span> <span class="va">ideal_env</span><span class="op">$</span><span class="va">ideal_values_20200926_135052</span><span class="op">$</span><span class="va">res_obj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">exported_res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for processing of intermediate steps</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tibble.tidyverse.org/" class="external-link">"tibble"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extracting the rankes for pre-ranked gsea</span></span>
<span><span class="va">de_ranks</span> <span class="op">&lt;-</span> <span class="va">exported_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">stat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">deframe</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">de_ranks</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># loading the signatures from the gmt file - to be downloaded locally</span></span>
<span><span class="va">pathways_gmtfile</span> <span class="op">&lt;-</span> <span class="fu">gmtPathways</span><span class="op">(</span><span class="st">"path_to/msigdb_v7.0_GMTs/h.all.v6.2.symbols.gmt"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># running fgsea</span></span>
<span><span class="va">fgsea_result</span> <span class="op">&lt;-</span> <span class="fu">fgsea</span><span class="op">(</span></span>
<span>  pathways <span class="op">=</span> <span class="va">pathways_gmtfile</span>,</span>
<span>  stats <span class="op">=</span> <span class="va">de_ranks</span>,</span>
<span>  nperm <span class="op">=</span> <span class="fl">100000</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fgsea_result</span> <span class="op">&lt;-</span> <span class="va">fgsea_result</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">NES</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># displaying the results</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">fgsea_result</span><span class="op">)</span></span></code></pre></div>
<p>The following chunk shows an example of how to perform the functional
enrichment analysis with <code>g:Profiler</code>’s tool
<code>g:GOST</code>, using the interface in <em><a href="https://CRAN.R-project.org/package=gprofiler2" class="external-link">gprofiler2</a></em></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"gprofiler2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># selecting the result object from the exported environment</span></span>
<span><span class="va">exported_res</span> <span class="op">&lt;-</span> <span class="va">ideal_env</span><span class="op">$</span><span class="va">ideal_values_20200926_135052</span><span class="op">$</span><span class="va">res_obj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">exported_res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extracting the DE genes vector</span></span>
<span><span class="va">degenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseqresult2DEgenes.html">deseqresult2DEgenes</a></span><span class="op">(</span><span class="va">exported_res</span>, FDR <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span></span>
<span></span>
<span><span class="co"># running the test</span></span>
<span><span class="va">gost_res</span> <span class="op">&lt;-</span> <span class="fu">gost</span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="va">degenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"hsapiens"</span>,</span>
<span>  ordered_query <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  multi_query <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  significant <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude_iea <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  measure_underrepresentation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  evcodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  user_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  correction_method <span class="op">=</span> <span class="st">"g_SCS"</span>,</span>
<span>  domain_scope <span class="op">=</span> <span class="st">"annotated"</span>,</span>
<span>  numeric_ns <span class="op">=</span> <span class="st">""</span>,</span>
<span>  sources <span class="op">=</span> <span class="st">"GO:BP"</span>,</span>
<span>  as_short_link <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># displaying the results</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">gost_res</span><span class="op">$</span><span class="va">result</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="enhancing-ideal">Enhancing <code>ideal</code><a class="anchor" aria-label="anchor" href="#enhancing-ideal"></a>
</h2>
<p>The <code>annotation_obj</code> is a quick helper to make your data
and results easier to read. For creating it, you can exploit the
corresponding <code>org.XX.eg.db</code> packages, available in
Bioconductor.</p>
<p>Currently available are the following (in alphabetical order):</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Ag.eg.db" class="external-link">org.Ag.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.At.tair.db" class="external-link">org.At.tair.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Bt.eg.db" class="external-link">org.Bt.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Ce.eg.db" class="external-link">org.Ce.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Cf.eg.db" class="external-link">org.Cf.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Dm.eg.db" class="external-link">org.Dm.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Dr.eg.db" class="external-link">org.Dr.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.EcK12.eg.db" class="external-link">org.EcK12.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.EcSakai.eg.db" class="external-link">org.EcSakai.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Gg.eg.db" class="external-link">org.Gg.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Mm.eg.db" class="external-link">org.Mm.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Mmu.eg.db" class="external-link">org.Mmu.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Pf.plasmo.db" class="external-link">org.Pf.plasmo.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Pt.eg.db" class="external-link">org.Pt.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Rn.eg.db" class="external-link">org.Rn.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Sc.sgd.db" class="external-link">org.Sc.sgd.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Sco.eg.db" class="external-link">org.Sco.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Ss.eg.db" class="external-link">org.Ss.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Tgondii.eg.db" class="external-link">org.Tgondii.eg.db</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/org.Xl.eg.db" class="external-link">org.Xl.eg.db</a></em></li>
</ul>
<p>They can all be easily installed with
<code>BiocManager::install("orgdb_packagename")</code>.</p>
<p>In a similar way, for using at best the <code>goseq</code> package,
the gene length information can also be computed if the corresponding
<code>TxDb</code> packages are installed. Currently, following
<code>TxDb</code> packages are available:</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Athaliana.BioMart.plantsmart22" class="external-link">TxDb.Athaliana.BioMart.plantsmart22</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Athaliana.BioMart.plantsmart25" class="external-link">TxDb.Athaliana.BioMart.plantsmart25</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Athaliana.BioMart.plantsmart28" class="external-link">TxDb.Athaliana.BioMart.plantsmart28</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Btaurus.UCSC.bosTau8.refGene" class="external-link">TxDb.Btaurus.UCSC.bosTau8.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Celegans.UCSC.ce11.refGene" class="external-link">TxDb.Celegans.UCSC.ce11.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Celegans.UCSC.ce6.ensGene" class="external-link">TxDb.Celegans.UCSC.ce6.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Cfamiliaris.UCSC.canFam3.refGene" class="external-link">TxDb.Cfamiliaris.UCSC.canFam3.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Dmelanogaster.UCSC.dm3.ensGene" class="external-link">TxDb.Dmelanogaster.UCSC.dm3.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Dmelanogaster.UCSC.dm6.ensGene" class="external-link">TxDb.Dmelanogaster.UCSC.dm6.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Drerio.UCSC.danRer10.refGene" class="external-link">TxDb.Drerio.UCSC.danRer10.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Ggallus.UCSC.galGal4.refGene" class="external-link">TxDb.Ggallus.UCSC.galGal4.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg18.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg18.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg19.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg19.lincRNAsTranscripts" class="external-link">TxDb.Hsapiens.UCSC.hg19.lincRNAsTranscripts</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Mmulatta.UCSC.rheMac3.refGene" class="external-link">TxDb.Mmulatta.UCSC.rheMac3.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Mmulatta.UCSC.rheMac8.refGene" class="external-link">TxDb.Mmulatta.UCSC.rheMac8.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Mmusculus.UCSC.mm10.ensGene" class="external-link">TxDb.Mmusculus.UCSC.mm10.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Mmusculus.UCSC.mm10.knownGene" class="external-link">TxDb.Mmusculus.UCSC.mm10.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Mmusculus.UCSC.mm9.knownGene" class="external-link">TxDb.Mmusculus.UCSC.mm9.knownGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Ptroglodytes.UCSC.panTro4.refGene" class="external-link">TxDb.Ptroglodytes.UCSC.panTro4.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Rnorvegicus.UCSC.rn4.ensGene" class="external-link">TxDb.Rnorvegicus.UCSC.rn4.ensGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Rnorvegicus.UCSC.rn5.refGene" class="external-link">TxDb.Rnorvegicus.UCSC.rn5.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Rnorvegicus.UCSC.rn6.refGene" class="external-link">TxDb.Rnorvegicus.UCSC.rn6.refGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Scerevisiae.UCSC.sacCer2.sgdGene" class="external-link">TxDb.Scerevisiae.UCSC.sacCer2.sgdGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Scerevisiae.UCSC.sacCer3.sgdGene" class="external-link">TxDb.Scerevisiae.UCSC.sacCer3.sgdGene</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.19/TxDb.Sscrofa.UCSC.susScr3.refGene" class="external-link">TxDb.Sscrofa.UCSC.susScr3.refGene</a></em></li>
</ul>
<p>As for the <code>org.XX.eg.db</code> above, these can all be quickly
installed with
<code>BiocManager::install("txdb_packagename")</code>.</p>
</div>
<div class="section level2">
<h2 id="further-development">Further development<a class="anchor" aria-label="anchor" href="#further-development"></a>
</h2>
<p>Additional functionality for the <em><a href="https://bioconductor.org/packages/3.19/ideal" class="external-link">ideal</a></em> will
be added in the future, as it is tightly related to a topic under
current development research.</p>
<p>Improvements, suggestions, bugs, issues and feedback of any type can
be sent to <a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-03-27 r86211)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Monterey 12.7.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] edgeR_4.1.19                limma_3.59.6               </span></span>
<span><span class="co">##  [3] DEFormats_1.31.0            org.Hs.eg.db_3.18.0        </span></span>
<span><span class="co">##  [5] DESeq2_1.43.4               airway_1.23.0              </span></span>
<span><span class="co">##  [7] SummarizedExperiment_1.33.3 GenomicRanges_1.55.4       </span></span>
<span><span class="co">##  [9] GenomeInfoDb_1.39.9         MatrixGenerics_1.15.0      </span></span>
<span><span class="co">## [11] matrixStats_1.2.0           ideal_1.27.0               </span></span>
<span><span class="co">## [13] topGO_2.55.0                SparseM_1.81               </span></span>
<span><span class="co">## [15] GO.db_3.18.0                AnnotationDbi_1.65.2       </span></span>
<span><span class="co">## [17] IRanges_2.37.1              S4Vectors_0.41.5           </span></span>
<span><span class="co">## [19] Biobase_2.63.0              graph_1.81.0               </span></span>
<span><span class="co">## [21] BiocGenerics_0.49.1         BiocStyle_2.31.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.3                 bitops_1.0-7             httr_1.4.7              </span></span>
<span><span class="co">##   [4] webshot_0.5.5            RColorBrewer_1.1-3       doParallel_1.0.17       </span></span>
<span><span class="co">##   [7] Rgraphviz_2.47.0         backports_1.4.1          tools_4.5.0             </span></span>
<span><span class="co">##  [10] utf8_1.2.4               R6_2.5.1                 DT_0.32                 </span></span>
<span><span class="co">##  [13] lazyeval_0.2.2           mgcv_1.9-1               withr_3.0.0             </span></span>
<span><span class="co">##  [16] prettyunits_1.2.0        gridExtra_2.3            fdrtool_1.2.17          </span></span>
<span><span class="co">##  [19] cli_3.6.2                textshaping_0.3.7        TSP_1.2-4               </span></span>
<span><span class="co">##  [22] labeling_0.4.3           slam_0.1-50              sass_0.4.9              </span></span>
<span><span class="co">##  [25] genefilter_1.85.1        goseq_1.55.0             pkgdown_2.0.7.9000      </span></span>
<span><span class="co">##  [28] Rsamtools_2.19.4         systemfonts_1.0.6        rentrez_1.2.3           </span></span>
<span><span class="co">##  [31] AnnotationForge_1.45.0   RSQLite_2.3.5            generics_0.1.3          </span></span>
<span><span class="co">##  [34] GOstats_2.69.0           BiocIO_1.13.0            gtools_3.9.5            </span></span>
<span><span class="co">##  [37] crosstalk_1.2.1          dplyr_1.1.4              dendextend_1.17.1       </span></span>
<span><span class="co">##  [40] Matrix_1.7-0             fansi_1.0.6              abind_1.4-5             </span></span>
<span><span class="co">##  [43] lifecycle_1.0.4          yaml_2.3.8               gplots_3.1.3.1          </span></span>
<span><span class="co">##  [46] SparseArray_1.3.4        BiocFileCache_2.11.1     grid_4.5.0              </span></span>
<span><span class="co">##  [49] blob_1.2.4               promises_1.2.1           crayon_1.5.2            </span></span>
<span><span class="co">##  [52] shinydashboard_0.7.2     lattice_0.22-6           GenomicFeatures_1.55.4  </span></span>
<span><span class="co">##  [55] annotate_1.81.2          KEGGREST_1.43.0          pillar_1.9.0            </span></span>
<span><span class="co">##  [58] knitr_1.45               rjson_0.2.21             codetools_0.2-19        </span></span>
<span><span class="co">##  [61] glue_1.7.0               data.table_1.15.2        vctrs_0.6.5             </span></span>
<span><span class="co">##  [64] png_0.1-8                gtable_0.3.4             assertthat_0.2.1        </span></span>
<span><span class="co">##  [67] cachem_1.0.8             xfun_0.43                pcaExplorer_2.29.0      </span></span>
<span><span class="co">##  [70] S4Arrays_1.3.6           mime_0.12                survival_3.5-8          </span></span>
<span><span class="co">##  [73] pheatmap_1.0.12          seriation_1.5.4          iterators_1.0.14        </span></span>
<span><span class="co">##  [76] statmod_1.5.0            nlme_3.1-164             Category_2.69.0         </span></span>
<span><span class="co">##  [79] bit64_4.0.5              threejs_0.3.3            progress_1.2.3          </span></span>
<span><span class="co">##  [82] filelock_1.0.3           UpSetR_1.4.0             bslib_0.6.2             </span></span>
<span><span class="co">##  [85] KernSmooth_2.23-22       colorspace_2.1-0         DBI_1.2.2               </span></span>
<span><span class="co">##  [88] tidyselect_1.2.1         bit_4.0.5                compiler_4.5.0          </span></span>
<span><span class="co">##  [91] curl_5.2.1               httr2_1.0.0              BiasedUrn_2.0.11        </span></span>
<span><span class="co">##  [94] xml2_1.3.6               desc_1.4.3               DelayedArray_0.29.9     </span></span>
<span><span class="co">##  [97] plotly_4.10.4            bookdown_0.38            rtracklayer_1.63.1      </span></span>
<span><span class="co">## [100] checkmate_2.3.1          scales_1.3.0             caTools_1.18.2          </span></span>
<span><span class="co">## [103] RBGL_1.79.0              NMF_0.27                 rappdirs_0.3.3          </span></span>
<span><span class="co">## [106] stringr_1.5.1            digest_0.6.35            shinyBS_0.61.1          </span></span>
<span><span class="co">## [109] rmarkdown_2.26           ca_0.71.1                XVector_0.43.1          </span></span>
<span><span class="co">## [112] htmltools_0.5.8          pkgconfig_2.0.3          base64enc_0.1-3         </span></span>
<span><span class="co">## [115] lpsymphony_1.31.0        highr_0.10               dbplyr_2.5.0            </span></span>
<span><span class="co">## [118] fastmap_1.1.1            rlang_1.1.3              htmlwidgets_1.6.4       </span></span>
<span><span class="co">## [121] shiny_1.8.1              farver_2.1.1             jquerylib_0.1.4         </span></span>
<span><span class="co">## [124] IHW_1.31.0               jsonlite_1.8.8           BiocParallel_1.37.1     </span></span>
<span><span class="co">## [127] RCurl_1.98-1.14          magrittr_2.0.3           GenomeInfoDbData_1.2.11 </span></span>
<span><span class="co">## [130] munsell_0.5.0            Rcpp_1.0.12              viridis_0.6.5           </span></span>
<span><span class="co">## [133] stringi_1.8.3            rintrojs_0.3.4           zlibbioc_1.49.3         </span></span>
<span><span class="co">## [136] plyr_1.8.9               parallel_4.5.0           ggrepel_0.9.5           </span></span>
<span><span class="co">## [139] Biostrings_2.71.5        splines_4.5.0            hms_1.1.3               </span></span>
<span><span class="co">## [142] geneLenDataBase_1.39.0   locfit_1.5-9.9           igraph_2.0.3            </span></span>
<span><span class="co">## [145] rngtools_1.5.2           reshape2_1.4.4           biomaRt_2.59.1          </span></span>
<span><span class="co">## [148] XML_3.99-0.16.1          evaluate_0.23            BiocManager_1.30.22     </span></span>
<span><span class="co">## [151] foreach_1.5.2            httpuv_1.6.15            tidyr_1.3.1             </span></span>
<span><span class="co">## [154] purrr_1.0.2              heatmaply_1.5.0          ggplot2_3.5.0           </span></span>
<span><span class="co">## [157] gridBase_0.4-7           xtable_1.8-4             restfulr_0.0.15         </span></span>
<span><span class="co">## [160] later_1.3.2              viridisLite_0.4.2        ragg_1.3.0              </span></span>
<span><span class="co">## [163] tibble_3.2.1             memoise_2.0.1            registry_0.5-1          </span></span>
<span><span class="co">## [166] GenomicAlignments_1.39.4 cluster_2.1.6            shinyAce_0.4.2          </span></span>
<span><span class="co">## [169] GSEABase_1.65.1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://federicomarini.github.io" class="external-link">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
